// uu5g04/dist-esm/bricks/blockquote.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import Footer from './blockquote-footer.js';

import './blockquote.less';

export const Blockquote = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.ColorSchemaMixin, UU5.Common.ContentMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Blockquote"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: ns.css("blockquote"),
      bg: ns.css("blockquote-bg"),
      right: 'blockquote-reverse'
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    background: PropTypes.bool,
    alignment: PropTypes.oneOf(['left', 'right']),
    footer: PropTypes.any,
    footerAlignment: PropTypes.oneOf(['left', 'right'])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      background: false,
      alignment: 'left',
      footer: null,
      footerAlignment: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _buildMainAttrs: function () {
    var mainAttrs = this.getMainAttrs();
    this.props.background && (mainAttrs.className += ' ' + this.getClassName().bg);
    this.props.alignment === 'right' && (mainAttrs.className += ' ' + this.getClassName().right);
    return mainAttrs;
  },

  _getFooterAlignment: function () {
    return this.props.footerAlignment || this.props.alignment;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return this.getNestingLevel() ? <blockquote {...this._buildMainAttrs()}>
            {this.getChildren()}
            {this.props.footer && <Footer content={this.props.footer} alignment={this._getFooterAlignment()} />}
            {this.getDisabledCover()}
          </blockquote> : null;
  }
  //@@viewOff:render
});

export default Blockquote;
