// uu5g04/dist-esm/bricks/cookie-bar.js
import React from 'react';
import createReactClass from '../create-react-class.js';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import { Div } from './factory.js';
import Span from './span.js';
import Link from './link.js';
import Button from './button.js';

import './cookie-bar.less';

export const CookieBar = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.ColorSchemaMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("CookieBar"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: ns.css("cookie-bar uu5-common-bg"),
      top: ns.css("cookie-bar-top"),
      bottom: ns.css("cookie-bar-bottom"),
      button: ns.css("cookie-bar-button", "button-inverted"),
      link: ns.css("cookie-bar-link")
    },
    defaults: {
      content: 'Cookies help us to provide, protect and improve our services. By viewing this site, you agree to their use.',
      expireDays: 10 * 365.25
    },
    opt: {
      nestingLevelRoot: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    agreedText: PropTypes.any,
    infoText: PropTypes.any,
    infoHref: PropTypes.string,
    infoTarget: PropTypes.oneOf(['_blank', '_parent', '_top', '_self']),
    fixed: PropTypes.oneOf(['top', 'bottom']),
    onAgree: PropTypes.func,
    expireDays: PropTypes.number,
    cookieKey: PropTypes.string,
    cookieValue: PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      agreedText: 'I Understand',
      infoText: null,
      infoHref: null,
      infoTarget: '_blank',
      fixed: 'bottom',
      onAgree: null,
      expireDays: null,
      cookieKey: 'uu5-cookies',
      cookieValue: 'yes'
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function () {
    this.props.expireDays ? this._checkCookies() : this._checkLocalStorageItem();
  },

  componentWillReceiveProps: function (nextProps) {
    this.props.expireDays ? this._checkCookies() : this._checkLocalStorageItem();
  },
  //@@viewOff:standardComponentLifeCycle

  // Interface

  // Overriding Functions

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _checkCookies: function () {
    if (UU5.Common.Tools.getCookie(this.props.cookieKey) === this.props.cookieValue && !this._isAgreed) {
      this._isAgreed = true;
      this.setState({ hidden: true }, this.props.onAgree);
    }
    return this;
  },

  _checkLocalStorageItem: function () {
    var item = localStorage.getItem(this.props.cookieKey);
    if (item === this.props.cookieValue) {
      this.setState({ hidden: true }, this.props.onAgree);
    }
    return this;
  },

  _setLocalStorageItem: function () {
    localStorage.setItem(this.props.cookieKey, this.props.cookieValue);
    this.hide(this.props.onAgree);
  },

  _confirm: function () {
    UU5.Common.Tools.setCookie(this.props.cookieKey, this.props.cookieValue, this.props.expireDays || this.getDefault().expireDays);
    this.hide(this.props.onAgree);
  },

  _getText: function () {
    var content = this.getContent();

    if (!content && !this.props.children) {
      content = this.getDefault().content;
    }

    return <Span content={content}>{this.props.children && React.Children.toArray(this.props.children)}</Span>;
  },

  _getButton: function () {
    var buttonProps = {
      content: this.props.agreedText,
      onClick: this.props.expireDays ? this._confirm : this._setLocalStorageItem,
      className: this.getClassName().button,
      bsStyle: 'inverted'
    };

    return <Button {...buttonProps} />;
  },

  _getLink: function () {
    var link = null;

    if (this.props.infoText) {
      // href -> linkHref || undefined -> if undefined, default prop of link is set - other way, null is set
      link = <Link content={this.props.infoText} href={this.props.infoHref || undefined} className={this.getClassName().link} target={this.props.infoTarget} />;
    }

    return link;
  },

  _getMainProps: function () {
    var mainProps = this.getMainPropsToPass(["UU5.Common.BaseMixin", "UU5.Common.ElementaryMixin"]);
    this.props.fixed && (mainProps.className += ' ' + this.getClassName(this.props.fixed));

    mainProps.nestingLevel = this.getNestingLevel();

    return mainProps;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return <Div {...this._getMainProps()}>
        {this._getText()}
        {this._getButton()}
        {this._getLink()}
      </Div>;
  }
  //@@viewOff:render
});

export default CookieBar;
