// uu5g04/dist-esm/bricks/modal.js
import React from 'react';
import createReactClass from '../create-react-class.js';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import Header from './modal-header.js';
import Body from './modal-body.js';
import Footer from './modal-footer.js';

import './modal.less';

export const Modal = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin, UU5.Common.NestingLevelMixin, UU5.Common.CcrReaderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Modal"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: ns.css("modal"),
      dialog: ns.css("modal-dialog", "modal-"),
      isFooter: ns.css("modal-isfooter")
    },
    defaults: {
      header: 'noHeader',
      body: 'noBody',
      animationDuration: 150, // ms
      closeTypes: {
        closedButton: 'closedButton',
        blur: 'blur',
        ifc: 'interface'
      }
    },
    opt: {
      nestingLevelRoot: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    size: PropTypes.oneOf(['s', 'm', 'l']),
    shown: PropTypes.bool,
    sticky: PropTypes.bool,
    stickyBackground: PropTypes.bool,
    scrollableBackground: PropTypes.bool,
    forceRender: PropTypes.bool,
    onClose: PropTypes.func
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      size: 'm',
      shown: false,
      sticky: false,
      stickyBackground: false,
      scrollableBackground: false,
      forceRender: false,
      onClose: null
    };
  },

  getInitialState() {
    return {
      header: this.getHeader(),
      content: this.getContent() || this.props.children,
      footer: this.getFooter(),
      className: null,
      size: this.props.size,
      sticky: this.props.sticky,
      stickyBackground: this.props.stickyBackground,
      scrollableBackground: this.props.scrollableBackground,
      onClose: this.props.onClose
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount() {
    this.props.shown ? this.open() : this.hide();
  },

  componentDidMount() {
    if (!this.isSticky()) {
      UU5.Environment.EventListener.addWindowEvent('keyup', this.getId(), this._onCloseESC);
    }
  },

  componentWillReceiveProps(nextProps) {
    if (nextProps.controlled) {
      this.setState(function (state) {
        let newState = {};

        if (nextProps.shown && state.hidden) {
          newState.hidden = false;
        } else if (!nextProps.shown && !state.hidden) {
          newState.hidden = true;
        }

        newState.header = nextProps.header;
        newState.footer = nextProps.footer;
        newState.content = nextProps.content || nextProps.children || this.state.content;
        newState.size = nextProps.size;
        newState.sticky = nextProps.sticky;
        newState.stickyBackground = nextProps.stickyBackground;
        newState.scrollableBackground = nextProps.scrollableBackground;
        newState.onClose = nextProps.onClose;

        return newState;
      });
    }
  },

  componentWillUnmount() {
    UU5.Environment.EventListener.removeWindowEvent('keyup', this.getId(), this._onCloseESC);
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isModal() {
    return true;
  },

  open: function (openProps, setStateCallback) {
    let page = this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE);
    if (!this.props.forceRender && page && page.getModal() && page.getModal().getId() !== this.getId()) {
      let centralModal = page.getModal();
      let newProps = UU5.Common.Tools.merge(this.props, openProps);
      centralModal.open(newProps, setStateCallback);
    } else {
      let newState = {
        hidden: false
      };

      openProps = openProps || {};
      newState.header = openProps.header === undefined ? this.props.header : openProps.header;
      newState.footer = openProps.footer === undefined ? this.props.footer : openProps.footer;
      newState.content = openProps.content === undefined ? this.getContent() || this.props.children : openProps.content;
      newState.className = openProps.className === undefined ? this.props.className : openProps.className;
      newState.size = openProps.size === undefined ? this.props.size : openProps.size;
      newState.sticky = openProps.sticky === undefined ? this.props.sticky : openProps.sticky;
      newState.stickyBackground = openProps.stickyBackground === undefined ? this.props.stickyBackground : openProps.stickyBackground;
      newState.scrollableBackground = openProps.scrollableBackground === undefined ? this.props.scrollableBackground : openProps.scrollableBackground;
      newState.onClose = openProps.onClose === undefined ? this.props.onClose : openProps.onClose;

      this._stopScroll(newState.scrollableBackground);
      this.setState(newState, setStateCallback);
    }
    return this;
  },

  close: function (shouldOnClose, setStateCallback) {
    let page = this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE);

    if (!this.props.forceRender && page && page.getModal() && page.getModal().getId() !== this.getId()) {
      let centralModal = page.getModal();
      centralModal.close(shouldOnClose, setStateCallback);
    } else if (typeof this.state.onClose === 'function' && shouldOnClose !== false) {
      this.state.onClose({ component: this, closeType: this.getDefault().closeTypes.ifc, callback: setStateCallback });
    } else {
      this._close(setStateCallback);
    }
    return this;
  },

  toggle(setStateCallback) {
    let modal = this;
    this.setState(state => {
      state.hidden ? modal._stopScroll(state.scrollableBackground) : modal._startScroll();
      return { hidden: !state.hidden };
    }, setStateCallback);
    return this;
  },

  isSticky() {
    return this.state.sticky;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  buildHeaderChild_(headerTypes) {
    let headerType = this.getHeaderType(headerTypes);

    let headerChild;
    if (headerType === 'contentOfStandardHeader') {
      headerChild = <Header content={headerTypes.header} />;
      headerChild = this.cloneChild(headerChild, this.expandHeaderProps(headerChild));
    }

    return headerChild;
  },

  expandHeaderProps_(headerChild) {
    let extendedHeaderProps = this._extendPartProps(headerChild.props, 'header');
    if (extendedHeaderProps) {
      extendedHeaderProps._sticky = this.state.sticky;
      extendedHeaderProps._onClose = this._onCloseHandler;
    }
    return extendedHeaderProps;
  },

  buildFooterChild_(footerTypes) {
    let footerType = this.getFooterType(footerTypes);

    let footerChild;
    if (footerType === 'contentOfStandardFooter') {
      footerChild = <Footer content={footerTypes.footer} />;
      footerChild = this.cloneChild(footerChild, this.expandFooterProps(footerChild));
    }

    return footerChild;
  },

  expandFooterProps_(footerChild) {
    return this._extendPartProps(footerChild.props, 'footer');
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onCloseESC(e) {
    e.which === 27 && !this.isHidden() && !this.isSticky() && this._blur(e);
    return this;
  },

  _onBlurHandler(event) {
    event.target.id === this.getId() && this._blur(event);
    return this;
  },

  _onCloseHandler(e) {
    if (typeof this.state.onClose === 'function') {
      this.state.onClose({ component: this, event: e, closeType: this.getDefault().closeTypes.closedButton });
    } else {
      this._close();
    }
    return this;
  },

  _blur(e) {
    if (typeof this.state.onClose === 'function') {
      this.state.onClose({ component: this, event: e, closeType: this.getDefault().closeTypes.blur });
    } else {
      this._close();
    }
    return this;
  },

  _close(callback) {
    this._startScroll();
    this.hide(callback);
    return this;
  },

  _getScrollbarWidth() {
    let width = 0;

    // if scroll bar is visible
    if (UU5.Common.Tools.getDocumentHeight() > window.innerHeight) {
      let div = document.createElement("div");
      div.style.overflow = "scroll";
      div.style.visibility = "hidden";
      div.style.position = 'absolute';
      div.style.width = '100px';
      div.style.height = '100px';

      // temporarily creates a div into DOM
      document.body.appendChild(div);
      try {
        width = div.offsetWidth - div.clientWidth;
      } finally {
        document.body.removeChild(div);
      }
    }

    return width;
  },

  _startScroll() {
    let modal = this;
    // TODO: wrong, but not found better solution
    setTimeout(function () {
      !modal.props.scrollableBackground && (document.body.style.overflow = '');
      document.body.style.paddingRight = '';
      !document.body.style.length && document.body.removeAttribute("style");
    }, this.getDefault().animationDuration);
  },

  _stopScroll(scrollableBackground) {
    // TODO: wrong, but not found better solution
    if (!scrollableBackground) {
      document.body.style.overflow = 'hidden';

      let paddingRight = this._getScrollbarWidth();
      paddingRight && (document.body.style.paddingRight = paddingRight + 'px');
    }
  },

  _getMainAttrs() {
    let mainAttrs = this.getMainAttrs();

    // id because of checking backdrop on click in _onBlurHandler function
    mainAttrs.id = this.getId();
    this.state.footer && (mainAttrs.className += ' ' + this.getClassName("isFooter"));
    this.state.className && (mainAttrs.className += ' ' + this.state.className);
    !this.state.sticky && !this.state.stickyBackground && (mainAttrs.onClick = this._onBlurHandler);

    let sec = this.getDefault().animationDuration / 1000 + 's';
    mainAttrs.style = mainAttrs.style || {};
    mainAttrs.style.WebkitTransitionDuration = sec;
    mainAttrs.style.MozTransitionDuration = sec;
    mainAttrs.style.OTransitionDuration = sec;
    mainAttrs.style.transitionDuration = sec;

    return mainAttrs;
  },

  _extendPartProps(partProps, part) {
    let newProps = {};

    // default values is used if child is set as react element so null or undefined will not set!!!
    for (let key in partProps) {
      partProps[key] !== null && partProps[key] !== undefined && (newProps[key] = partProps[key]);
    }

    newProps.key = newProps.id;

    return newProps;
  },

  _extendBodyProps(bodyProps) {
    let id = this.getId() + '-body';

    let newProps = {
      id: id
    };

    // default values is used if child is set as react element so null or undefined will not set!!!
    for (let key in bodyProps) {
      bodyProps[key] !== null && bodyProps[key] !== undefined && (newProps[key] = bodyProps[key]);
    }

    return UU5.Common.Tools.merge(newProps, { key: newProps.id });
  },
  //@@viewOff:componentSpecificHelpers

  // Render
  _buildChildren() {
    let header = this.state.header;
    let footer = this.state.footer;
    let bodyContent = this.state.content;

    let headerChild;
    let footerChild;

    if (!bodyContent && !header) {
      header = this.getDefault().header;
      bodyContent = this.getDefault().body;
    }

    header && (headerChild = this.buildHeaderChild({ header: header }));
    footer && (footerChild = this.buildFooterChild({ footer: footer }));

    let bodyProps = this._extendBodyProps({ content: bodyContent });

    let bodyChild;

    if (bodyProps.content) {
      bodyChild = this.buildChildren({
        children: React.createElement(Body, bodyProps)
      });
    }

    return [headerChild, bodyChild, footerChild];
  },

  //@@viewOn:render
  render() {
    // Todo uu5-size is maped to bootstrap-size
    // let size;
    // switch (this.state.size) {
    //   case 'l':
    //     size = 'lg';
    //     break;
    //   case 'm':
    //     size = 'md';
    //     break;
    //   default:
    //     size = 'sm';
    //     break;
    // }

    return this.getNestingLevel() ? <div {...this._getMainAttrs()}>
            <div className={this.getClassName().dialog + this.state.size}>
              {this._buildChildren()}
            </div>
            {this.getDisabledCover()}
          </div> : null;
  }
  //@@viewOff:render
});

export default Modal;
