// uu5g04/dist-esm/bricks/panel-body.js
import React from 'react';
import createReactClass from '../create-react-class.js';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import './panel-body.less';

export default createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Panel.Body"),
    classNames: {
      main: ns.css("panel-body"),
      body: ns.css("panel-body-body"),
      block: ns.css("panel-body-block"),
      expanding: ns.css("panel-body-expanding")
    },
    defaults: {
      parentTagName: 'UU5.Bricks.Panel',
      duration: 250
    },
    errors: {
      invalidParent: 'Parent of this component is not Panel.'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    _expanded: PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      _expanded: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function () {
    return {
      height: this._getHeight(this.props)
    };
  },

  componentWillMount: function () {
    //this.checkParentTagName(this.getDefault().parentTagName);

    if (!(this.getParent() && this.getParent().isPanel)) {
      this.showError("invalidParent");
    }
  },

  componentWillReceiveProps: function (nextProps) {
    var body = this;
    var height = this._getHeight(nextProps);

    if (height) {
      this.setState({ height: height }, function () {
        body.timer && clearTimeout(body.timer);
        body.timer = setTimeout(function () {
          body.setAsyncState({ height: null });
        }, body.getDefault().duration);
      });
    } else {
      this.setState({ height: this._getHeight(this.props) }, function () {
        body.timer && clearTimeout(body.timer);
        body.timer = setTimeout(function () {
          body.setAsyncState({ height: height });
        }, 1);
      });
    }
  },

  componentWillUnmount: function () {
    this.timer && clearTimeout(this.timer);
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getPanelBodyId: function () {
    return this.getId() + '-content';
  },

  _getHeight: function (props) {
    let result;
    if (props._expanded) {
      let element = document.getElementById(this._getPanelBodyId());
      element && (result = UU5.Common.Tools.getOuterHeight(element, true));
    } else {
      result = 0;
    }
    return result;
  },
  //@@viewOff:componentSpecificHelpers

  // Render
  _getMainAttrs: function () {
    var mainAttrs = this.getMainAttrs();

    if (this.state.height === null) {
      mainAttrs.className += ' ' + this.getClassName().block;
    } else {
      this.state.height > 0 && (mainAttrs.className += ' ' + this.getClassName().expanding);

      mainAttrs.style = mainAttrs.style || {};
      mainAttrs.style.height = this.state.height + 'px';

      var time = this.getDefault().duration / 1000;
      ['WebkitTransitionDuration', 'MozTransitionDuration', 'MsTransitionDuration', 'OTransitionDuration', 'transitionDuration'].forEach(function (style) {
        mainAttrs.style[style] = time + 's';
      });
    }

    return mainAttrs;
  },

  _buildChildren: function () {
    return this.buildChildren();
  },

  //@@viewOn:render
  render: function () {
    return <div {...this._getMainAttrs()}>
        <div className={this.getClassName().body} id={this._getPanelBodyId()} ref={panelBody => this._panelBody = panelBody}>
          {this._buildChildren()}
        </div>
        {this.getDisabledCover()}
      </div>;
  }
  //@@viewOff:render
});
