// uu5g04/dist-esm/bricks/progress-bar-cover.js
import React from 'react';
import createReactClass from '../create-react-class.js';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import './progress-bar-cover.less';

export default createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("ProgressBar.Cover"),
    classNames: {
      main: ns.css("progress-bar-cover")
    },
    errors: {
      invalidParent: 'Parent of this component is not ProgressBar.'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function () {
    if (!(this.getParent() && this.getParent().isProgressBar)) {
      this.showError("invalidParent");
    }
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return <div {...this.getMainAttrs()}>
        {this.props.children && React.Children.toArray(this.props.children)}
        {this.getDisabledCover()}
      </div>;
  }
  //@@viewOff:render
});
