// uu5g04/dist-esm/bricks/screen-size.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import ScreenSizeItem from './screen-size-item.js';

import './screen-size.less';

export const ScreenSize = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.ScreenSizeMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("ScreenSize"),
    nestingLevelList: UU5.Environment.getNestingLevelList('spa'),
    classNames: {
      main: ns.css("screen-size")
    },
    defaults: {
      childTagName: 'UU5.Bricks.ScreenSize.Item'
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    screenSize: PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      screenSize: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isScreenSize() {
    return true;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _expandChildProps: function (child) {
    let newChildProps = UU5.Common.Tools.mergeDeep({}, child.props);

    newChildProps.id = this.getId() + '-inner';
    newChildProps.controlled = true;
    newChildProps.hidden = this.isHidden();
    newChildProps.disabled = this.isDisabled();
    newChildProps.selected = this.isSelected();
    newChildProps.nestingLevel = this.getNestingLevel();

    return newChildProps;
  },

  _testScreenSize(child, screenSize) {
    let result = child && child.props && child.props.screenSize;
    if (result) {
      if (Array.isArray(result)) {
        result = result.indexOf(screenSize) > -1;
      } else {
        result = result === '*' ? true : result === screenSize;
      }
    }
    return result;
  },

  _getItem() {
    let item = null;
    let children = this.getChildren();
    let screenSize = this.getScreenSize();

    if (children) {
      children.some(child => {
        return item = this._testScreenSize(child, screenSize) ? child : null;
      });
      item = React.cloneElement(item, this._expandChildProps(item));
    }

    return item;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return this.getNestingLevel() ? this._getItem() : null;
  }
  //@@viewOff:render
});

ScreenSize.Item = ScreenSizeItem;

export default ScreenSize;
