// uu5g04/dist-esm/core/common/div.js
import React from 'react';
import createReactClass from 'create-react-class';
import ns from "./common-ns.js";
import BaseMixin from './base-mixin.js';
import ElementaryMixin from './elementary-mixin.js';
import ContentMixin from './content-mixin.js';
import NestingLevelMixin from './nesting-level-mixin.js';
import Environment from '../environment/environment.js';
import PureRenderMixin from "./pure-render-mixin";

export const Div = createReactClass({

  //@@viewOn:mixins
  mixins: [BaseMixin, ElementaryMixin, ContentMixin, NestingLevelMixin, PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Div"),
    classNames: {
      main: ns.css("div")
    },
    nestingLevelList: Environment.getNestingLevelList('bigBoxCollection', 'smallBox')
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return this.getNestingLevel() ? React.createElement('div', this.getMainAttrs(), this.getChildren(), this.getDisabledCover()) : null;
  }
  //@@viewOff:render
});

export default Div;
