// uu5g04/dist-esm/core/common/lsi-mixin.js
import React from 'react';
import PropTypes from 'prop-types';
import Tools from './tools.js';
import Environment from '../environment/environment.js';

export const LsiMixin = {

  //@@viewOn:statics
  statics: {
    "UU5.Common.LsiMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"],
      lsiEvent: Tools.events.lsi
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    language: PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      language: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function () {
    // initialize
    this.registerMixin("UU5.Common.LsiMixin");
    // state
    return {
      language: this.props.language || Tools.getLanguages()[0].location || Tools.getLanguages()[0].language
    };
  },

  componentDidMount: function () {
    window.UU5.Environment.EventListener.registerLsi(this.getId(), this._changeLanguage);
  },

  componentWillUnmount: function () {
    window.UU5.Environment.EventListener.unregisterLsi(this.getId(), this._changeLanguage);
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonLsiMixin: function () {
    return this.hasMixin("UU5.Common.LsiMixin");
  },

  getUU5CommonLsiMixinProps: function () {
    return {
      language: this.props.language
    };
  },

  getUU5CommonLsiMixinPropsToPass: function () {
    return this.getUU5CommonLsiMixinProps();
  },

  getLanguages: function () {
    return Tools.getLanguages();
  },

  getLanguage: function () {
    return this.state.language;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  // TODO: 'cs-CZ,en;q=0.8...' is set to state and never will be used, necessary is doing setState and render the component
  _setLanguage: function (language, setStateCallback) {
    this.setState({ language: language }, setStateCallback);
    return this;
  },

  _changeLanguage: function (language) {
    if (this.getLanguage() !== language) {
      if (typeof this.onChangeLanguage_ === 'function') {
        this.onChangeLanguage_(language);
      } else {
        this._setLanguage(language);
      }
    }
  }
  //@@viewOff:componentSpecificHelpers
};

export default LsiMixin;
