// uu5g04/dist-esm/core/common/pure-render-mixin.js
import React from 'react';
import PropTypes from 'prop-types';

export const PureRenderMixin = {

  //@@viewOn:statics
  statics: {
    "UU5.Common.PureRenderMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"]
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    pureRender: PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      pureRender: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  shouldComponentUpdate(nextProps, nextState) {
    let shouldUpdate;
    if (!this.props.pureRender) {
      shouldUpdate = true;
    } else if (typeof this.shouldComponentUpdate_ === 'function') {
      shouldUpdate = this.shouldComponentUpdate_(nextProps, nextState);
    } else {
      shouldUpdate = this.shouldRender(nextProps, nextState);
    }
    return shouldUpdate;
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonPureRenderMixin() {
    return this.hasMixin("UU5.Common.PureRenderMixin");
  },

  getUU5CommonPureRenderMixinProps() {
    return {
      pureRender: this.props.pureRender
    };
  },

  getUU5CommonPureRenderMixinPropsToPass() {
    return this.getUU5CommonPureRenderMixinProps();
  }
};

export default PureRenderMixin;
