// uu5g04/dist-esm/core/common/route-mixin.js
import React from 'react';
import PropTypes from 'prop-types';

export const RouteMixin = {

  //@@viewOn:mixins
  mixins: [],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    "UU5.Common.RouteMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"]
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    params: PropTypes.object
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      params: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState() {
    // initialize
    this.registerMixin("UU5.Common.RouteMixin");

    return {};
  },

  componentDidMount() {
    if (typeof this.onRouteChanged_ === 'function') {
      this.onRouteChanged_();
    }
  },

  componentDidUpdate(prevProps, prevState) {
    if (typeof this.onRouteChanged_ === 'function') {
      this.onRouteChanged_(prevProps, prevState);
    }
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonRouteMixin: function () {
    return this.hasMixin("UU5.Common.RouteMixin");
  },

  getUU5CommonRouteMixinProps: function () {
    return {};
  },

  getUU5CommonRouteMixinPropsToPass: function () {
    return this.getUU5CommonRouteMixinProps();
  }
  //@@viewOff:interface
};

export default RouteMixin;
