// uu_plus4u5g01/dist-node/productcatalogue/product-list.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from "prop-types";
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import ProductInfo from './product-info.js';

import './product-list.less';

export const ProductList = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.SectionMixin,
    UU5.Common.PureRenderMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.ProductCatalogue.ProductList',
    classNames: {
      main: 'plus4u5-product-catalogue-product-list',
      column: 'plus4u5-product-catalogue-product-list-column'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    type: PropTypes.oneOf(["simple", "detail"]),
    products: PropTypes.arrayOf(
      PropTypes.shape({
        logo: PropTypes.string,
        name: PropTypes.string,
        description: PropTypes.string,
        appCode: PropTypes.string,
        tidAwid: PropTypes.string
      })
    )
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      type: undefined, // default from product info
      products: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainProps() {
    let props = this.getMainPropsToPass();

    if (this.getLevel() === 0) {
      props.level = 1;
    }

    return props;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return (
      <UU5.Bricks.Row {...this._getMainProps()}>
        {this.props.products.map((props, key)=>{
          return (
            <UU5.Bricks.Column key={key} className={this.getClassName("column")} colWidth="xs12 s12 m6 l6 xl6">
              <ProductInfo {...props} type={this.props.type}/>
            </UU5.Bricks.Column>
          );
        })}
      </UU5.Bricks.Row>
    );
  }
  //@@viewOff:render
});

export default ProductList;
