// uu_plus4u5g01/dist-node/productcatalogue/uaf-products.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from "prop-types";
import * as UU5 from "uu5g04";
import * as Plus4U5 from 'uu_plus4u5g01';

import ProductList from './product-list.js';

import './uaf-products.less';

export const UAFProducts = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.SectionMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.ProductCatalogue.UAFProducts',
    classNames: {
      main: 'plus4u5-product-catalogue-uaf-products'
    },
    lsi: () => (Plus4U5.Environment.Lsi.ProductCatalogue.uafProducts),
    opt: {
      dummyLevel: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    type: PropTypes.oneOf(["simple", "detail"])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      type: undefined // default from product info
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainProps() {
    let props = this.getMainPropsToPass();

    if (props.header === null) {
      props.header = this.getLsiComponent("header");
    }

    return props;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return (
      <ProductList
        {...this._getMainProps()}
        products={Plus4U5.Environment.UAFProducts}
        type={this.props.type}
      />
    );
  }
  //@@viewOff:render
});

export default ProductList;
