// uu5g04/dist-esm/bricks/alert.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import Link from './link.js';
import Icon from './icon.js';

import './alert.less';

export const Alert = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.ColorSchemaMixin, UU5.Common.SectionMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Alert"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: ns.css("alert"),
      withHeader: ns.css("alert-with-header"),
      header: ns.css("alert-header"),
      headerWrapper: ns.css("alert-header-wrapper"),
      content: ns.css("alert-content"),
      position: ns.css("alert-"),
      block: ns.css("alert-block"),
      close: ns.css("alert-close")
    },
    defaults: {
      transitionDuration: 150,
      closeIcon: 'mdi-close'
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    position: PropTypes.oneOf(['left', 'center', 'right']),
    closeTimer: PropTypes.number,
    closeDisabled: PropTypes.bool,
    block: PropTypes.bool,
    onClose: PropTypes.func,
    onCloseAfter: PropTypes.func,
    onCloseBefore: PropTypes.func
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      position: 'center',
      closeTimer: null,
      closeDisabled: false,
      block: false,
      onClose: null,
      onCloseAfter: null,
      onCloseBefore: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillReceiveProps: function (nextProps) {
    nextProps.content !== this.props.content && this._clearTimeout();
  },

  componentWillUnmount: function () {
    this._clearTimeout();
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _clearTimeout: function () {
    if (this.timeout) {
      clearTimeout(this.timeout);
      this.timeout = null;
    }
    return this;
  },

  _hide: function () {
    var alert = this;
    this._clearTimeout();

    if (typeof this.props.onCloseBefore === 'function') {
      this.props.onCloseBefore(alert, this._hideFunc());
    } else {
      this._hideFunc();
    }

    return this;
  },

  _hideFunc: function () {
    var alert = this;
    this.hide(typeof this.props.onClose === 'function' || typeof this.props.onCloseAfter === 'function' ? function () {
      setTimeout(function () {
        typeof alert.props.onCloseAfter === 'function' ? alert.props.onCloseAfter(alert) : alert.props.onClose(alert);
      }, this.getDefault().transitionDuration);
    } : null);
  },

  //@@viewOff:componentSpecificHelpers

  // Render
  _getMainAttrs: function () {
    var mainAttrs = this.getMainAttrs();
    mainAttrs.className += ' ' + this.getClassName().position + this.props.position;
    this.props.block && (mainAttrs.className += ' ' + this.getClassName().block);

    mainAttrs.style = mainAttrs.style || {};

    var time = this.getDefault().transitionDuration / 1000;
    ['WebkitTransitionDuration', 'MozTransitionDuration', 'MsTransitionDuration', 'OTransitionDuration', 'transitionDuration'].forEach(function (style) {
      mainAttrs.style[style] = time + 's';
    });

    return mainAttrs;
  },

  _manageTimeout: function () {
    if (this.props.closeTimer) {
      if (this.isHidden()) {
        this._clearTimeout();
      } else if (!this.timeout) {
        this.timeout = setTimeout(this._hide, this.props.closeTimer);
      }
    }
  },

  _getChildren() {
    let result;

    if (this.props.header) {
      result = [<div className={this.getClassName('headerWrapper')} key="1">
          <div className={this.getClassName('header')}> {this.getHeader()} </div>
          {!this.props.closeDisabled && <Link className={this.getClassName().close} onClick={this._hide} parent={this} colorSchema="custom">
            <Icon icon={this.getDefault('closeIcon')} />
          </Link>}
        </div>, <div className={this.getClassName('content')} key="2"> {this.getChildren()} </div>, this.getDisabledCover()];
    } else {
      result = [<div className={this.getClassName('headerWrapper')} key="1">
          <div className={this.getClassName('content')}> {this.getChildren()} </div>
          {!this.props.closeDisabled && <Link className={this.getClassName().close} onClick={this._hide} parent={this} colorSchema="custom">
            <Icon icon={this.getDefault('closeIcon')} />
          </Link>}
        </div>, this.getDisabledCover()];
    }

    return result;
  },

  //@@viewOn:render
  render: function () {
    let mainAttrs = this._getMainAttrs();
    if (this.props.header) {
      mainAttrs.className += ' ' + this.getClassName('withHeader');
    }

    this._manageTimeout();

    return this.getNestingLevel() ? <div {...mainAttrs}>
            {this._getChildren()}
          </div> : null;
  }
  //@@viewOff:render
});

export default Alert;
