// uu5g04/dist-esm/bricks/backdrop.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import './backdrop.less';

export const Backdrop = createReactClass({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Backdrop"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'smallBox'),
    classNames: {
      main: ns.css("backdrop"),
      background: ns.css("backdrop-background")
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    onClick: PropTypes.func,
    onMouseOver: PropTypes.func,
    background: PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      onClick: null,
      onMouseOver: null,
      background: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    var mainAttrs = this.getMainAttrs();
    mainAttrs.id = this.getId();
    if (typeof this.props.onClick === 'function') {
      mainAttrs.onClick = e => this.props.onClick(this, e);
    }
    if (typeof this.props.onMouseOver === 'function') {
      mainAttrs.onMouseOver = e => this.props.onMouseOver(this, e);
    }
    if (this.props.background) {
      mainAttrs.className += ' ' + this.getClassName('background');
    }

    return this.getNestingLevel() ? <div {...mainAttrs}>
            {this.getDisabledCover()}
          </div> : null;
  }
  //@@viewOff:render
});

export default Backdrop;
