// uu5g04/dist-esm/bricks/carousel-item.js
import React from 'react';
import createReactClass from '../create-react-class.js';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";
import { Div } from './factory.js';

export default createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Carousel.Item"),
    classNames: {
      main: ns.css("carousel-item")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _buildMainAttrs() {
    const mainAttrs = this.buildMainAttrs();
    mainAttrs.style = this.props.style;
    return mainAttrs;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return <Div {...this._buildMainAttrs()}>
      {this.getChildren()}
      </Div>;
  }
  //@@viewOff:render
});
