// uu5g04/dist-esm/bricks/code.js
import React from 'react';
import createReactClass from 'create-react-class';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import Factory from './factory.js';

import './code.less';

const Code_ = Factory.createTag('code', '_code');

export const Code = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Code"),
    nestingLevel: 'inline',
    classNames: {
      main: ns.css("code")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainProps: function () {
    var mainProps = this.getMainPropsToPass();

    mainProps.id = this.getId() + '-code';
    mainProps.ignoreInnerHTML = true;
    mainProps.language = 'code';

    return mainProps;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return this.getNestingLevel() ? <Code_ {...this._getMainProps()}>
            {this.props.children && React.Children.toArray(this.props.children)}
          </Code_> : null;
  }
  //@@viewOff:render
});

export default Code;
