// uu5g04/dist-esm/bricks/column.js
import React from 'react';
import createReactClass from '../create-react-class.js';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import './col.less';
import './column.less';

export const Column = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.SectionMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Column"),
    nestingLevelList: UU5.Environment.getNestingLevelList('spa', 'smallBoxCollection'),
    classNames: {
      main: ns.css("column"),
      spacing: ns.css("column-spacing"),
      noSpacing: ns.css("column-nospacing")
    },
    defaults: {
      colWidth: { xs: 12, s: 12, m: 12, l: 12, xl: 12 },
      columnRegexp: /^([a-z]+)(\d+)$/
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    noSpacing: PropTypes.bool,
    width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    // BS Column magic
    colWidth: PropTypes.oneOfType([PropTypes.shape({
      xs: PropTypes.number,
      s: PropTypes.number,
      m: PropTypes.number,
      l: PropTypes.number,
      xl: PropTypes.number
    }), PropTypes.string])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      noSpacing: false,
      width: null,
      colWidth: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _buildColumnClassName: function (className) {
    return className + ' ' + this._buildColWidthClassName(this.props.colWidth);
  },

  _buildColWidthClassName: function (colWidth) {
    var newBsColWidth = colWidth || this.getDefault('colWidth');

    if (typeof newBsColWidth === 'string') {
      var colWidthArray = newBsColWidth.split(' ');
      newBsColWidth = {};
      colWidthArray.forEach(colWidth => {
        var match = colWidth.match(this.getDefault().columnRegexp);
        newBsColWidth[match[1]] = parseInt(match[2]);
      });
    }

    var sizeClassNames = [];
    var lowerWidth = 12;

    !newBsColWidth.xs && lowerWidth && (newBsColWidth.xs = lowerWidth);
    (lowerWidth = newBsColWidth.xs) && sizeClassNames.push('uu5-col-xs' + newBsColWidth.xs);
    !newBsColWidth.s && lowerWidth && (newBsColWidth.s = lowerWidth);
    (lowerWidth = newBsColWidth.s) && sizeClassNames.push('uu5-col-s' + newBsColWidth.s);
    !newBsColWidth.m && lowerWidth && (newBsColWidth.m = lowerWidth);
    (lowerWidth = newBsColWidth.m) && sizeClassNames.push('uu5-col-m' + newBsColWidth.m);
    !newBsColWidth.l && lowerWidth && (newBsColWidth.l = lowerWidth);
    (lowerWidth = newBsColWidth.l) && sizeClassNames.push('uu5-col-l' + newBsColWidth.l);
    !newBsColWidth.xl && lowerWidth && (newBsColWidth.xl = lowerWidth);
    newBsColWidth.xl && sizeClassNames.push('uu5-col-xl' + newBsColWidth.xl);

    newBsColWidth['xs-offset'] && sizeClassNames.push('uu5-col-xs-offset-' + newBsColWidth['xs-offset']);
    newBsColWidth['s-offset'] && sizeClassNames.push('uu5-col-s-offset-' + newBsColWidth['s-offset']);
    newBsColWidth['m-offset'] && sizeClassNames.push('uu5-col-m-offset-' + newBsColWidth['m-offset']);
    newBsColWidth['l-offset'] && sizeClassNames.push('uu5-col-l-offset-' + newBsColWidth['l-offset']);
    newBsColWidth['xl-offset'] && sizeClassNames.push('uu5-col-xl-offset-' + newBsColWidth['xl-offset']);

    return sizeClassNames.join(' ');
  },

  _getMainAttrs() {
    let mainAttrs = this.getMainAttrs();
    mainAttrs.className += " " + this.getClassName(this.props.noSpacing ? "noSpacing" : "spacing");

    if (this.props.width) {
      mainAttrs.style = UU5.Common.Tools.merge(mainAttrs.style, { width: this.props.width });
    } else {
      mainAttrs.className = mainAttrs.className + ' ' + this._buildColWidthClassName(this.props.colWidth);
    }

    return mainAttrs;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getNestingLevel() ? <div {...this._getMainAttrs()}>
            {this.getHeaderChild()}
            {this.getChildren()}
            {this.getFooterChild()}
            {this.getDisabledCover()}
          </div> : null;
  }
  //@@viewOff:render
});

export default Column;
