// uu5g04/dist-esm/bricks/row.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import './row.less';

export const Row = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.SectionMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Row"),
    nestingLevelList: ['spa', 'page', 'bigBoxCollection', 'bigBoxCollection', 'boxCollection', 'box', 'smallBoxCollection'],
    classNames: {
      main: ns.css("row"),
      spacing: ns.css("row-spacing"),
      noSpacing: ns.css("row-nospacing"),
      standard: ns.css("row-standard"),
      flex: ns.css("row-flex")
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    noSpacing: PropTypes.bool,
    display: PropTypes.oneOf(['standard', 'flex'])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      noSpacing: false,
      display: 'standard'
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainAttrs() {
    let mainAttrs = this.getMainAttrs();
    mainAttrs.className += ' ' + this.getClassName(this.props.display);
    mainAttrs.className += ' ' + this.getClassName(this.props.noSpacing ? 'noSpacing' : 'spacing');
    return mainAttrs;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getNestingLevel() ? <div {...this._getMainAttrs()}>
          {this.getHeaderChild()}
          {this.getChildren()}
          {this.getFooterChild()}
          {this.getDisabledCover()}
        </div> : null;
  }
  //@@viewOff:render
});

export default Row;
