// uu5g04/dist-esm/bricks/swiper.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import SwiperBody from './swiper-body.js';
import SwiperMenu from './swiper-menu.js';

import './swiper.less';

export const Swiper = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.SwipeMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Swiper"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: ns.css("swiper")
    },
    defaults: {
      bodyTagName: 'UU5.Bricks.Swiper.Body',
      menuTagName: 'UU5.Bricks.Swiper.Menu'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    leftMenuOpen: PropTypes.bool,
    rightMenuOpen: PropTypes.bool,
    onSwipeOpenLeftMenu: PropTypes.func,
    onSwipeCloseLeftMenu: PropTypes.func,
    onSwipeOpenRightMenu: PropTypes.func,
    onSwipeCloseRightMenu: PropTypes.func,
    allowBodyTags: PropTypes.array,
    allowMenuTags: PropTypes.array
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      leftMenuOpen: false,
      rightMenuOpen: false,
      onSwipeOpenLeftMenu: null,
      onSwipeCloseLeftMenu: null,
      onSwipeOpenRightMenu: null,
      onSwipeCloseRightMenu: null,
      allowBodyTags: [],
      allowMenuTags: []
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function () {
    return {
      leftMenuOpen: this.props.leftMenuOpen,
      rightMenuOpen: this.props.rightMenuOpen
    };
  },

  componentWillReceiveProps(nextProps) {
    if (nextProps.controlled) {
      this.setState({
        leftMenuOpen: nextProps.leftMenuOpen,
        rightMenuOpen: nextProps.rightMenuOpen
      });
    }
    return this;
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isSwiper() {
    return true;
  },

  openLeftMenu: function (setStateCallback) {
    this.setState({
      leftMenuOpen: true,
      rightMenuOpen: false
    }, setStateCallback);
    return this;
  },

  closeLeftMenu: function (setStateCallback) {
    this.setState({
      leftMenuOpen: false
    }, setStateCallback);
    return this;
  },

  toggleLeftMenu: function (setStateCallback) {
    this.setState(function (state) {
      var newState = { leftMenuOpen: !state.leftMenuOpen };
      !state.leftMenuOpen && (newState.rightMenuOpen = false);
      return newState;
    }, setStateCallback);
    return this;
  },

  openRightMenu: function (setStateCallback) {
    this.setState({
      leftMenuOpen: false,
      rightMenuOpen: true
    }, setStateCallback);
    return this;
  },

  closeRightMenu: function (setStateCallback) {
    this.setState({
      rightMenuOpen: false
    }, setStateCallback);
    return this;
  },

  toggleRightMenu: function (setStateCallback) {
    this.setState(function (state) {
      var newState = { rightMenuOpen: !state.rightMenuOpen };
      !state.rightMenuOpen && (newState.leftMenuOpen = false);
      return newState;
    }, setStateCallback);
    return this;
  },

  isLeftMenuOpen: function () {
    return this.state.leftMenuOpen;
  },

  isRightMenuOpen: function () {
    return this.state.rightMenuOpen;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  shouldChildRender_: function (child) {
    var childTagName = UU5.Common.Tools.getChildTagName(child);
    return this._getAllowMenuTags().indexOf(childTagName) > -1 || this._getAllowBodyTags().indexOf(childTagName) > -1;
  },

  expandChildProps_: function (child) {
    var newChildProps = UU5.Common.Tools.mergeDeep({}, child.props);
    if (this._getAllowMenuTags().indexOf(UU5.Common.Tools.getChildTagName(child)) > -1) {
      if (child.props.pullRight) {
        newChildProps._open = this.isRightMenuOpen();
      } else {
        newChildProps._open = this.isLeftMenuOpen();
      }
    }
    return newChildProps || child.props;
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getAllowBodyTags() {
    return this.props.allowBodyTags.concat(this.getDefault().bodyTagName);
  },

  _getAllowMenuTags() {
    return this.props.allowMenuTags.concat(this.getDefault().menuTagName);
  },

  _onSwipeEnd: function () {
    if (this.isSwipedRight()) {
      if (this.isRightMenuOpen()) {
        if (typeof this.props.onSwipeCloseRightMenu === 'function') {
          this.props.onSwipeCloseRightMenu(this);
        } else {
          this.closeRightMenu();
        }
      } else {
        if (typeof this.props.onSwipeOpenLeftMenu === 'function') {
          this.props.onSwipeOpenLeftMenu(this);
        } else {
          this.openLeftMenu();
        }
      }
    } else if (this.isSwipedLeft()) {
      if (this.isLeftMenuOpen()) {
        if (typeof this.props.onSwipeCloseLeftMenu === 'function') {
          this.props.onSwipeCloseLeftMenu(this);
        } else {
          this.closeLeftMenu();
        }
      } else {
        if (typeof this.props.onSwipeOpenRightMenu === 'function') {
          this.props.onSwipeOpenRightMenu(this);
        } else {
          this.openRightMenu();
        }
      }
    }
    return this;
  },
  //@@viewOff:componentSpecificHelpers

  // Render
  _buildChildren: function () {
    var menuLeft;
    var menuRight;
    var body;

    var children = this.getChildren();
    if (children) {
      children.forEach(function (child) {
        if (this._getAllowBodyTags().indexOf(UU5.Common.Tools.getChildTagName(child)) > -1) {
          body = child;
        } else if (this._getAllowMenuTags().indexOf(UU5.Common.Tools.getChildTagName(child)) > -1) {
          if (child.props.pullRight) {
            menuRight = menuRight || child;
          } else {
            menuLeft = menuLeft || child;
          }
        }
      }.bind(this));
    }

    var newChildren = [];
    menuLeft && newChildren.push(menuLeft);
    menuRight && newChildren.push(menuRight);
    body && newChildren.push(body);
    return newChildren;
  },

  //@@viewOn:render
  render: function () {
    return this.getNestingLevel() ? <div {...this.getMainAttrs()} onTouchStart={this.swipeOnTouchStart} onTouchMove={this.swipeOnTouchMove} onTouchEnd={this.swipeOnTouchEnd.bind(this, this._onSwipeEnd)}>

          {this._buildChildren()}
          {this.getDisabledCover()}
        </div> : null;
  }
  //@@viewOff:render
});

Swiper.Menu = SwiperMenu;
Swiper.Body = SwiperBody;

export default Swiper;
