// uu5g04/dist-esm/bricks/table.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import TableCol from './table-col.js';
import TableColGroup from './table-col-group.js';
import TableTBody from './table-tbody.js';
import TableTd from './table-td.js';
import TableTFoot from './table-tfoot.js';
import TableTh from './table-th.js';
import TableTHead from './table-thead.js';
import TableTr from './table-tr.js';

import './table.less';

export const Table = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin, UU5.Common.ColorSchemaMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Table"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'smallBox'),
    classNames: {
      main: ns.css("table"),
      table: ns.css("table-table"),
      striped: ns.css("table-striped"),
      bordered: ns.css("table-bordered"),
      hover: ns.css("table-hover"),
      condensed: ns.css("table-condensed"),
      responsive: ns.css("table-responsive")
    },
    defaults: {
      childTagNames: ['UU5.Bricks.Table.Tr', 'UU5.Bricks.Table.THead', 'UU5.Bricks.Table.TBody', 'UU5.Bricks.Table.TFoot', 'UU5.Bricks.Table.ColGroup']
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    striped: PropTypes.bool,
    bordered: PropTypes.bool,
    hover: PropTypes.bool,
    condensed: PropTypes.bool,
    responsive: PropTypes.bool,
    allowTags: PropTypes.arrayOf(PropTypes.string)
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      striped: false,
      bordered: false,
      hover: false,
      condensed: false,
      responsive: false,
      allowTags: []
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isTable() {
    return true;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  shouldChildRender_: function (child) {
    var childTagName = UU5.Common.Tools.getChildTagName(child);
    var childTagNames = this.props.allowTags.concat(this.getDefault().childTagNames);
    return childTagNames.indexOf(childTagName) > -1;
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainProps: function () {
    var props = this.getMainAttrs();
    this.props.responsive && (props.className += ' ' + this.getClassName().responsive);
    return props;
  },

  _buildTableAttrs: function () {
    var tableAttrs = { className: this.getClassName().table };
    this.props.striped && (tableAttrs.className += ' ' + this.getClassName().striped);
    this.props.bordered && (tableAttrs.className += ' ' + this.getClassName().bordered);
    this.props.hover && (tableAttrs.className += ' ' + this.getClassName().hover);
    this.props.condensed && (tableAttrs.className += ' ' + this.getClassName().condensed);
    return tableAttrs;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return this.getNestingLevel() ? <div {...this._getMainProps()}>
          {this.getHeaderChild()}
          <table {...this._buildTableAttrs()}>
            {this.getChildren()}
          </table>
          {this.getFooterChild()}
          {this.getDisabledCover()}
        </div> : null;
  }
  //@@viewOff:render
});

Table.ColGroup = TableColGroup;
Table.Col = TableCol;
Table.TBody = TableTBody;
Table.Td = TableTd;
Table.TFoot = TableTFoot;
Table.Th = TableTh;
Table.THead = TableTHead;
Table.Tr = TableTr;

export default Table;
