// uu5g04/dist-esm/bricks/table-th.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./bricks-ns.js";

import './table-th.less';

export default createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.ContentMixin, UU5.Common.ColorSchemaMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Table.Th"),
    nestingLevelList: UU5.Environment.getNestingLevelList('boxCollection', 'inline'),
    classNames: {
      main: ns.css("table-th uu5-common-text")
      //bg: 'uu5-common-bg'
    },
    defaults: {
      parentTagName: 'UU5.Bricks.Table.Tr'
    },
    opt: {
      nestingLevelWrapper: true
    },
    errors: {
      invalidParent: 'Parent of this component is not Table.'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    colSpan: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    rowSpan: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
  },

  getDefaultProps: function () {
    return {
      colSpan: null,
      rowSpan: null
    };
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function () {
    //this.checkParentTagName(this.getDefault().parentTagName);

    if (!(this.getParent() && (this.getParent().isTable || this.getParent().isTr))) {
      this.showError("invalidParent");
    }
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainProps: function () {
    const props = this.getMainAttrs();
    //this.props.background && (props.className += ' ' + this.getClassName().bg);
    this.props.colSpan && (props.colSpan = this.props.colSpan);
    this.props.rowSpan && (props.rowSpan = this.props.rowSpan);
    return props;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return this.getNestingLevel() ? <th {...this._getMainProps()}>
          {this.getChildren()}
          {this.getDisabledCover()}
        </th> : null;
  }
  //@@viewOff:render
});
