// uu5g04/dist-node/forms/datepicker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Datepicker = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('../create-react-class.js');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _formsNs = require('./forms-ns.js');

var _formsNs2 = _interopRequireDefault(_formsNs);

var _textInput = require('./internal/text-input.js');

var _textInput2 = _interopRequireDefault(_textInput);

var _calendar = require('./calendar.js');

var _calendar2 = _interopRequireDefault(_calendar);

var _textInputMixin = require('./mixins/text-input-mixin.js');

var _textInputMixin2 = _interopRequireDefault(_textInputMixin);

require('./datepicker.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Datepicker = exports.Datepicker = (0, _createReactClass2.default)({
  displayName: 'Datepicker',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, _textInputMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _formsNs2.default.name("Datepicker"),
    classNames: {
      main: _formsNs2.default.css("datepicker"),
      open: _formsNs2.default.css("datepicker-open"),
      menu: _formsNs2.default.css("input-menu")
    },
    defaults: {
      format: 'dd.mm.Y',
      regexpSpace: / /g
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    value: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.string]),
    dateFrom: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.string]),
    dateTo: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.string]),
    buttonHidden: _propTypes2.default.bool,
    iconOpen: _propTypes2.default.string,
    iconClosed: _propTypes2.default.string,
    format: _propTypes2.default.string,
    country: _propTypes2.default.string,
    nanMessage: _propTypes2.default.any,
    beforeRangeMessage: _propTypes2.default.any,
    afterRangeMessage: _propTypes2.default.any,
    parseDate: _propTypes2.default.func,
    disableBackdrop: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      value: null,
      dateFrom: null,
      dateTo: null,
      buttonHidden: false,
      iconOpen: 'mdi-calendar',
      iconClosed: 'mdi-calendar',
      format: null,
      country: null,
      nanMessage: 'Please insert a valid date.',
      beforeRangeMessage: 'Date is out of range.',
      afterRangeMessage: 'Date is out of range.',
      parseDate: null,
      icon: 'mdi-calendar',
      disableBackdrop: false
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      open: false,
      format: this.props.format,
      country: this.props.country
    };
  },
  componentWillMount: function componentWillMount() {
    var value = this._prepareStringDate(this.state.value);

    if (this.props.onValidate && typeof this.props.onValidate === 'function') {
      var result = this._isValidDateResult({ value: value });
      if (result) {
        if ((typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object') {
          if (result.feedback) {
            this.setFeedback(result.feedback, result.message, result.value);
          } else {
            this._validateOnChange({ value: value, event: null, component: this });
          }
        }
      }
    } else {
      this.setFeedback(this.props.feedback, this.props.message, value);
    }
    return this;
  },
  componentDidMount: function componentDidMount() {
    UU5.Environment.EventListener.registerDateTime(this.getId(), this._change);
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (this.props.controlled) {
      var value = this._prepareStringDate(nextProps.value);
      var result = this._isValidDateResult({ value: value });
      if (result) {
        if ((typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object') {
          if (result.feedback) {
            this.setFeedback(result.feedback, result.message, result.value);
          } else {
            this.setFeedback(nextProps.feedback, nextProps.message, value);
          }
        }
      }
    }
    return this;
  },
  componentWillUnmount: function componentWillUnmount() {
    this._removeEvent();
    UU5.Environment.EventListener.unregisterDateTime(this.getId(), this._change);
  },


  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  close: function close(setStateCallback, e) {
    var _this = this;

    this._removeEvent();
    var opt = { value: this._formatDate(this.getValue()), event: e, component: this };
    setStateCallback = typeof setStateCallback === 'function' ? setStateCallback : function () {
      return _this._onBlurCalendar(opt);
    };
    this.setState({ open: false }, setStateCallback);
    return this;
  },
  toggle: function toggle(setStateCallback, e) {
    var _this2 = this;

    var opt = { value: this._formatDate(this.getValue()), event: e, component: this };
    setStateCallback = setStateCallback || this.isOpen() ? function () {
      return _this2._onFocusCalendar(opt);
    } : function () {
      return _this2._onBlurCalendar(opt);
    };
    //this.setState((state) => ({open: !state.open}), setStateCallback);
    this.setState(function (state) {
      state.open ? this._removeEvent() : this._addEvent();
      return { open: !state.open };
    }, setStateCallback);
    return this;
  },
  parseDate: function parseDate(stringDate) {
    var config = {
      "^(\\d{1,2})\\.(\\d{1,2})\\.(\\d{4})$": '$3-$2-$1',
      "^(\\d{1,2})\\/(\\d{1,2})\\/(\\d{4})$": '$3-$1-$2',
      "^(\\d{4})-(\\d{1,2})-(\\d{1,2})$": '$1-$2-$3',
      "^(\\d{1,2})-(\\d{1,2})-(\\d{4})$": '$3-$2-$1'
    };

    var date = null;
    if (stringDate) {
      stringDate = stringDate.replace(this.getDefault().regexpSpace, '');

      for (var pattern in config) {
        var regExp = new RegExp(pattern);
        if (regExp.test(stringDate)) {
          var replacedDate = stringDate.replace(regExp, config[pattern]);
          var splitter = replacedDate.split('-');
          //splitter[1] = UU5.Common.Tools.rjust(splitter[1], 2, '0');
          //splitter[2] = UU5.Common.Tools.rjust(splitter[2], 2, '0');
          //replacedDate = splitter.join('-');
          splitter[0] = Number(splitter[0]);
          splitter[1] = Number(splitter[1]) - 1;
          splitter[2] = Number(splitter[2]);

          date = new Date(splitter[0], splitter[1], splitter[2]);
          break;
        }
      }
    }
    return date;
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  // TODO: tohle je ještě otázka - je potřeba nastavit hodnotu z jiné komponenty (musí být validace) a z onChange (neměla by být validace)
  setValue_: function setValue_(value, setStateCallback) {
    value = this._prepareStringDate(value);
    if (this._checkRequired({ value: value })) {
      if (typeof this.props.onValidate === 'function') {
        this._validateOnChange({ value: value, event: null, component: this });
      } else {
        this.setInitial(null, value, setStateCallback);
      }
    }

    return this;
  },
  getValue_: function getValue_() {
    var date = void 0;

    if (this.state.value instanceof Date) {
      date = this.state.value;
    } else {
      date = this._parseDate(this.state.value);
    }

    return date;
  },

  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _findTarget: function _findTarget(item) {
    var result = false;
    var id = this.getId();

    if (item.id === id) {
      result = true;
    } else if (item.parentElement) {
      result = this._findTarget(item.parentElement);
    }
    return result;
  },
  _addEvent: function _addEvent() {
    var _this3 = this;

    !this.props.disableBackdrop && UU5.Environment.EventListener.addWindowEvent('click', this.getId(), function (e) {
      var isDatepicker = _this3._findTarget(e.target);

      if (!_this3._stopPropagation && !isDatepicker && _this3.isOpen()) {
        _this3.close();
      } else {
        _this3._stopPropagation = false;
      }
    });
    return this;
  },
  _removeEvent: function _removeEvent() {
    !this.props.disableBackdrop && UU5.Environment.EventListener.removeWindowEvent('click', this.getId());
    return this;
  },
  _change: function _change(opt) {
    if (typeof this.props.onChange === 'function') {
      this.props.onChange(opt);
    } else {
      this._setOptions(opt);
    }
    return this;
  },
  _setOptions: function _setOptions(opt, setStateCallback) {
    this.setState({
      format: opt.format === undefined ? this.state.format : opt.format,
      country: opt.country === undefined ? this.state.country : opt.country ? opt.country.toLowerCase() : opt.country
    }, setStateCallback);
    return this;
  },
  _onChange: function _onChange(e) {
    var opt = { value: e.target.value, event: e, component: this };
    var date = this._parseDate(opt.value);
    var formatedDate = date ? this._formatDate(date) : null;

    if (!opt.value || formatedDate) {
      opt.value = formatedDate;
      if (!this.isDisabled() && !this.isReadOnly()) {
        if (typeof this.props.onChange === 'function') {
          this.props.onChange(opt);
        } else {
          if (this.props.validateOnChange) {
            this._validateOnChange(opt);
          } else {
            if (opt.value) {
              if (this._checkRequired({ value: opt.value })) {
                opt.required = this.props.required;
                var result = this.getChangeFeedback(opt);
                this.setFeedback(result.feedback, result.message, result.value);
              }
            } else {
              this.setState({ value: opt.value });
            }
          }
        }
      }
    } else {
      this.setState({ value: opt.value });
    }
    return this;
  },
  _onFocus: function _onFocus(e) {
    var opt = { value: e.target.value, event: e, component: this };

    this._onFocusCalendar(opt);

    return this;
  },
  _onFocusCalendar: function _onFocusCalendar(opt) {
    if (typeof this.props.onFocus === 'function') {
      this.props.onFocus(opt);
    } else {
      var result = this.getFocusFeedback(opt);
      result && this.setFeedback(result.feedback, result.message, result.value);
    }

    return this;
  },
  _onBlur: function _onBlur(e) {
    var opt = { value: e.target.value, event: e, component: this };
    this._onBlurCalendar(opt);
    return this;
  },
  _onBlurCalendar: function _onBlurCalendar(opt) {
    if (!this.isOpen()) {
      if (typeof this.props.onBlur === 'function') {
        this.props.onBlur(opt);
      } else {
        if (this._checkRequired({ value: opt.value }) && !this.props.validateOnChange) {
          opt.required = this.props.required;
          var blurResult = this.getBlurFeedback(opt);
          var result = this._isValidDateResult(blurResult);
          this.setFeedback(result.feedback, result.message, result.value);
        }
      }
    }
    return this;
  },
  _isValidDateResult: function _isValidDateResult(opt) {
    return this._validateDateRangeResult(this._validateDateResult(opt));
  },
  _validateDateResult: function _validateDateResult(opt) {
    var result = opt;
    var date = this._parseDate(opt.value);
    if (!date && opt.value) {
      result.feedback = 'error';
      result.message = this.props.nanMessage;
    }
    return result;
  },
  _getDate: function _getDate(date) {
    if (typeof date === 'string') {
      date = this._parseDate(date);
    }
    return date;
  },
  _validateDateRangeResult: function _validateDateRangeResult(opt) {
    var result = opt;

    var date = this._parseDate(opt.value);
    if (date) {
      var dateFrom = this._getDateFrom();
      var dateTo = this._getDateTo();
      if (dateFrom && date < dateFrom) {
        result.feedback = 'error';
        result.message = this.props.beforeRangeMessage;
      } else if (dateTo && date > dateTo) {
        result.feedback = 'error';
        result.message = this.props.afterRangeMessage;
      }
    }

    return result;
  },
  _validateOnChange: function _validateOnChange(opt) {
    var result = typeof this.props.onValidate === 'function' ? this.props.onValidate(opt) : null;
    if (result) {
      if ((typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object') {
        if (result.feedback) {
          this.setFeedback(result.feedback, result.message, result.value);
        } else {
          this.setState({ value: opt.value });
        }
      } else {
        this.showError('validateError', null, { context: { event: e, func: this.props.onValidate, result: result } });
      }
    } else if (this.props.required && this.state.value) {
      this.setSuccess(null, this.state.value);
    }
    return this;
  },
  _getFeedbackIcon: function _getFeedbackIcon() {
    return this.isOpen() ? this.props.iconOpen : this.props.iconClosed;
  },
  _onCalendarChange: function _onCalendarChange(opt) {
    var _this4 = this;

    var date = this._formatDate(opt.value);
    opt.value = date;
    opt.component = this;
    if (typeof this.props.onChange === 'function') {
      this.setState({ open: false }, this.props.onChange(opt));
    } else {
      this.setValue(date, function () {
        return _this4.close();
      });
    }
    return this;
  },
  _getDateFrom: function _getDateFrom() {
    var dateFrom = void 0;
    if (this.props.dateFrom) {
      if (typeof this.props.dateFrom === 'string') {
        dateFrom = this._parseDate(this.props.dateFrom);
      } else if (this.props.dateFrom instanceof Date) {
        dateFrom = this.props.dateFrom;
      }
    }
    return dateFrom;
  },
  _getDateTo: function _getDateTo() {
    var dateTo = void 0;
    if (this.props.dateTo) {
      if (typeof this.props.dateTo === 'string') {
        dateTo = this._parseDate(this.props.dateTo);
      } else if (this.props.dateTo instanceof Date) {
        dateTo = this.props.dateTo;
      }
    }
    return dateTo;
  },
  _getCalendarProps: function _getCalendarProps() {
    var date = typeof this.getValue() === 'string' ? this._parseDate(this.getValue()) : this.getValue();

    return {
      className: this.getClassName().menu,
      date: date,
      dateFrom: this._getDateFrom(),
      dateTo: this._getDateTo(),
      ref_: this._refCalendar,
      hidden: !this.isOpen(),
      onChange: this._onCalendarChange
    };
  },
  _prepareStringDate: function _prepareStringDate(date) {
    var result = date;
    if (date instanceof Date) {
      result = this._formatDate(date);
    }
    return result;
  },
  _formatDate: function _formatDate(date) {
    var result = null;

    if (date) {
      if (this.state.format) {
        result = UU5.Common.Tools.formatDate(date, this.state.format);
      } else if (this.state.country) {
        result = this._formatDateByCountry(date, this.state.country);
      } else {
        result = UU5.Common.Tools.toLocaleDateString(date, UU5.Common.Tools.getLanguage());
      }
    }
    return result;
  },
  _formatDateByCountry: function _formatDateByCountry(date, country) {
    var result = void 0;
    if (UU5.Environment.dateTimeFormat[country]) {
      result = UU5.Common.Tools.formatDate(date, UU5.Environment.dateTimeFormat[country]);
    } else {
      result = UU5.Common.Tools.toLocaleDateString(date, country);
    }
    return result;
  },
  _isDateReversed: function _isDateReversed() {
    var dateFormat = this.state.country || UU5.Common.Tools.getLanguage();
    return dateFormat.toLowerCase() === 'en' || dateFormat.toLowerCase() === 'en-us';
  },
  _parseDate: function _parseDate(stringDate) {
    var date = null;
    if (this.props.parseDate && typeof this.props.parseDate === 'function') {
      date = this.props.parseDate(stringDate);
    } else {
      var config = {
        "^(\\d{1,2})\\.(\\d{1,2})\\.(\\d{4})$": '$3-$2-$1',
        "^(\\d{1,2})\\/(\\d{1,2})\\/(\\d{4})$": this._isDateReversed() ? '$3-$1-$2' : '$3-$2-$1',
        "^(\\d{4})-(\\d{1,2})-(\\d{1,2})$": '$1-$2-$3',
        "^(\\d{1,2})-(\\d{1,2})-(\\d{4})$": '$3-$2-$1'
      };

      stringDate = stringDate && stringDate.replace(this.getDefault().regexpSpace, '');

      for (var pattern in config) {
        var regExp = new RegExp(pattern);
        if (regExp.test(stringDate)) {
          var replacedDate = stringDate.replace(regExp, config[pattern]);

          // because of safari and IE must be date in format YYYY-MM-DD (not YYYY-M-D)
          var parser = replacedDate.split('-');
          replacedDate = [parser[0], UU5.Common.Tools.rjust(parser[1], 2, '0'), UU5.Common.Tools.rjust(parser[2], 2, '0')].join('-');

          date = Date.parse(replacedDate) ? new Date(replacedDate) : null;
          date && date.setHours(0, 0, 0, 0);
          break;
        }
      }
    }
    return date;
  },
  _getTextInputAttrs: function _getTextInputAttrs() {
    var _this5 = this;

    var props = {};

    if (!this.isReadOnly() && !this.isDisabled()) {
      props.onClick = function () {
        _this5.open();
      };
    }

    return props;
  },
  _getMainAttrs: function _getMainAttrs() {
    var attrs = this._getInputAttrs();
    attrs.id = this.getId();

    if (this.isOpen()) {
      attrs.className += ' ' + this.getClassName().open;
    }

    return attrs;
  },
  _getPlacehoder: function _getPlacehoder() {
    var format = this.state.format || UU5.Environment.dateTimeFormat[this.state.country];
    var placeholder = format;
    if (this.props.placeholder && format) {
      placeholder = this.props.placeholder + ' - ' + format;
    } else if (this.props.placeholder) {
      placeholder = this.props.placeholder;
    }
    return placeholder;
  },


  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var _this6 = this;

    var inputId = this.getId() + '-input';
    /*let buttons = !this.isReadOnly() && !this.props.buttonHidden ? [{
        icon: this.isOpen() ? this.props.iconOpen : this.props.iconClosed,
        disabled: this.isDisabled(),
        onClick: this.toggle,
        pressed: this.isOpen(),
        colorSchema: 'default'
      }] : null;*/

    var date = this._parseDate(this.state.value);

    var formatedDate = date ? this._formatDate(date) : this.state.value;
    return _react2.default.createElement(
      'div',
      this._getMainAttrs(),
      this.getLabel(inputId),
      this.getInputWrapper([_react2.default.createElement(_textInput2.default, { id: inputId, name: this.props.name || inputId, value: formatedDate || '', placeholder: this._getPlacehoder(), type: 'text', onChange: this._onChange, onBlur: this._onBlur, onFocus: this._onFocus, onKeyDown: this.onKeyDown, mainAttrs: this._getTextInputAttrs() || this.props.inputAttrs, disabled: this.isDisabled() || this.isLoading(), readonly: this.isReadOnly(), icon: this._getFeedbackIcon(), iconClickable: false, loading: this.isLoading(), ref_: function ref_(item) {
          return _this6._textInput = item;
        }, feedback: this.getFeedback() }), _react2.default.createElement(_calendar2.default, this._getCalendarProps())])
    );
  }
  //@@viewOn:render

});

exports.default = Datepicker;
