// uu5g04/dist-node/forms/forms-lsi.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var FormsLsi = exports.FormsLsi = {
  message: {
    patternMessage: { en: 'Enter a value that matches the specified format.', cs: 'Zadejte hodnotu, která odpovídá požadovanému formátu.' },
    requiredMessage: { en: 'This field is required.', cs: 'Toto pole je povinné.' },
    requiredMessageChoice: { en: 'Select some item.', cs: 'Vyberte položku.' },
    requiredMessageGroup: { en: 'Choose at least one option.', cs: 'Vyberte alespoň jednu položku.' }
  },
  form: {
    validContent: { en: 'Form is valid.', cs: 'Formulář je validní.' },
    invalidContent: { en: 'Form is not valid.', cs: 'Formulář není validní.' },
    ok: { en: 'OK', cs: 'OK' },
    validate: { en: 'Validate', cs: 'Zvalidovat' },
    reset: { en: 'Reset', cs: 'Reset' },
    cancel: { en: 'Cancel', cs: 'Zrušit' }
  },
  calendar: {
    dayNames: {
      cs: ['Pondělí', 'Úterý', 'Středa', 'Čtvrtek', 'Pátek', 'Sobota', 'Neděle'],
      nl: ['Maandag', 'Dinsdag', 'Woensdag', 'Donderdag', 'Vrijdag', 'Zaterdag', 'Zondag'],
      en: ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
      'en-gb': ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
      'en-us': ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'],
      fr: ['Lundi', 'Mardi', 'Mercredi', 'Jeudi', 'Vendredi', 'Samedi', 'Dimanche'],
      de: ['Montag', 'Dienstag', 'Mittwoch', 'Donnerstag', 'Freitag', 'Samstag', 'Sonntag'],
      pl: ['Poniedziałek', 'Wtorek', 'Środa', 'Czwartek', 'Piątek', 'Sobota', 'Niedziela'],
      ru: ['Понедельник', 'Вторник', 'Cреда', 'Четверг', 'Пятница', 'Cуббота', 'Воскресенье'],
      sk: ['Pondelok', 'Utorok', 'Streda', 'Štvrtok', 'Piatok', 'Sobota', 'Nedeľa'],
      es: ['Lunes', 'Martes', 'Miércoles', 'Jueves', 'Viernes', 'Sábado', 'Domingo'],
      uk: ['Понеділок', 'Вівторок', 'Cереда', 'Четвер', "П'ятниця", 'Cубота', 'Неділя']
    },
    monthNames: {
      cs: ['Leden', 'Únor', 'Březen', 'Duben', 'Květen', 'Červen', 'Červenec', 'Srpen', 'Září', 'Říjen', 'Listopad', 'Prosinec'],
      nl: ['Januari', 'Februari', 'Maart', 'April', 'Mei', 'Juni', 'Juli', 'Augustus', 'September', 'Oktober', 'November', 'December'],
      en: ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
      'en-gb': ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
      'en-us': ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
      fr: ['Janvier', 'Février', 'Mars', 'Avril', 'Mai', 'Juin', 'Juillet', 'Août', 'Septembre', 'Octobre', 'Novembre', 'Décembre'],
      de: ['Januar', 'Februar', 'März', 'April', 'Mai', 'Juni', 'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember'],
      pl: ['Styczeń', 'Luty', 'Marzec', 'Kwiecień', 'Maj', 'Czerwiec', 'Lipiec', 'Sierpień', 'Wrzesień', 'Październik', 'Listopad', 'Grudzień'],
      ru: ['Январь', 'Февраль', 'Март', 'Апрель', 'Май', 'Июнь', 'Июль', 'Август', 'Сентябрь', 'Октябрь', 'Ноябрь', 'Декабрь'],
      sk: ['Január', 'Február', 'Marec', 'Apríl', 'Máj', 'Jún', 'Júl', 'August', 'September', 'Október', 'November', 'December'],
      es: ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre'],
      uk: ['Січень', 'Лютий', 'Березень', 'Квітень', 'Травень', 'Червень', 'Липень', 'Серпень', 'Вересень', 'Жовтень', 'Листопад', 'Грудень']
    }
  },
  select: {
    selectAll: {
      cs: 'Vybrat vše',
      nl: 'Selecteer alles',
      en: 'Select all',
      'en-gb': 'Select all',
      'en-us': 'Select all',
      fr: 'Tout sélectionner',
      de: 'Alles auswählen',
      pl: 'Zaznacz wszystko',
      ru: 'Выбрать все',
      sk: 'Vybrať všetko',
      es: 'Seleccionar todo',
      uk: 'Вибрати все'
    },
    unselectAll: {
      cs: 'Odebrat vše',
      nl: 'Deselecteer alles',
      en: 'Deselect all',
      'en-gb': 'Deselect all',
      'en-us': 'Deselect all',
      fr: 'Tout déselectionner',
      de: 'Alle abwählen',
      pl: 'Odznacz wszystkie',
      ru: 'Удалить все',
      sk: 'Odznačiť všetko',
      es: 'Anular la selección de todo',
      uk: 'Cкасувати всі'
    }
  }
};

exports.default = FormsLsi;
