// uu5g04/dist-esm/forms/internal/item-list.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "../forms-ns.js";

import './item-list.less';

export default createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("ItemList"),
    classNames: {
      main: ns.css("item-list"),
      header: ns.css("item-list-header"),
      body: ns.css("item-list-body"),
      footer: ns.css("item-list-footer"),
      selectedItem: ns.css("item-list-selected uu5-common-bg")
    },
    defaults: {
      childTagName: 'UU5.Forms.Select.option'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    allowTags: PropTypes.arrayOf(PropTypes.string)
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      allowTags: []
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  changeValue(index, e, setStateCallback) {
    if (typeof this.props.onChange === 'function') {
      this.props.onChange({ value: index, event: e, component: this, setStateCallback: setStateCallback });
    } else if (typeof setStateCallback === 'function') {
      setStateCallback();
    }
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  shouldChildRender_: function (child) {
    var childTagName = UU5.Common.Tools.getChildTagName(child);
    var childTagNames = this.props.allowTags.concat(this.getDefault().childTagName);
    return childTagNames.indexOf(childTagName) > -1;
  },

  expandChildProps_: function (child, i) {
    var props = UU5.Common.Tools.mergeDeep({}, child.props);

    if (this.props.value && this.props.value.indexOf(i) > -1) {
      props.selected = true;
      props.className = props.className ? props.className + ' ' + this.getClassName().selectedItem : this.getClassName().selectedItem;
      props.colorSchema = 'info';
    }

    props.mainAttrs = props.mainAttrs || {};
    props.mainAttrs.id = this.getId() + '-item-' + i;

    var childOnClick = props.onClick;
    props.onClick = opt => {
      this.changeValue(opt.value, opt.event, childOnClick);
    };

    return props;
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getHeader() {
    let result;
    if (this.props.header) {
      result = <div className={this.getClassName().header}>
        {this.getHeader()}
      </div>;
    }
    return result;
  },

  _getFooter() {
    let result;
    if (this.props.footer) {
      result = <div className={this.getClassName().footer}>
        {this.getFooter()}
      </div>;
    }
    return result;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return <div {...this.getMainAttrs()}>
        {this._getHeader()}
        <div className={this.getClassName().body}>
          {this.getChildren()}
        </div>
        {this._getFooter()}
      </div>;
  }
  //@@viewOn:render
});
