// uu5g04/dist-esm/forms/internal/label.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "../forms-ns.js";

import './label.less';

export default createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Label"),
    classNames: {
      main: ns.css("label"),
      required: ns.css("label-required")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    for: PropTypes.string,
    required: PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      for: null,
      required: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  shouldComponentUpdate_(newProps, newState) {
    let result = false;
    if (newProps.for != this.props.for || newProps.content != this.props.content || newProps.children != this.props.children) {
      result = true;
    }
    return result;
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainAttrs() {
    let mainAttrs = this.getMainAttrs();
    this.props.for && (mainAttrs.htmlFor = this.props.for);
    mainAttrs.className += ' ' + this.props.colWidth;
    if (this.props.required) {
      mainAttrs.className += ' ' + this.getClassName("required");
    }
    return mainAttrs;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return <label {...this._getMainAttrs()}>
        {this.getChildren()}
      </label>;
  }
  //@@viewOn:render
});
