// uu5g04/dist-node/bricks/backdrop.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Backdrop = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

require('./backdrop.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Backdrop = exports.Backdrop = (0, _createReactClass2.default)({
  displayName: 'Backdrop',

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Backdrop"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'smallBox'),
    classNames: {
      main: _bricksNs2.default.css("backdrop"),
      background: _bricksNs2.default.css("backdrop-background")
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    onClick: _propTypes2.default.func,
    onMouseOver: _propTypes2.default.func,
    background: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      onClick: null,
      onMouseOver: null,
      background: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var _this = this;

    var mainAttrs = this.getMainAttrs();
    mainAttrs.id = this.getId();
    if (typeof this.props.onClick === 'function') {
      mainAttrs.onClick = function (e) {
        return _this.props.onClick(_this, e);
      };
    }
    if (typeof this.props.onMouseOver === 'function') {
      mainAttrs.onMouseOver = function (e) {
        return _this.props.onMouseOver(_this, e);
      };
    }
    if (this.props.background) {
      mainAttrs.className += ' ' + this.getClassName('background');
    }

    return this.getNestingLevel() ? _react2.default.createElement(
      'div',
      mainAttrs,
      this.getDisabledCover()
    ) : null;
  }
  //@@viewOff:render
});

exports.default = Backdrop;
