// uu5g04/dist-node/bricks/blockquote.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Blockquote = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _blockquoteFooter = require('./blockquote-footer.js');

var _blockquoteFooter2 = _interopRequireDefault(_blockquoteFooter);

require('./blockquote.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Blockquote = exports.Blockquote = (0, _createReactClass2.default)({
  displayName: 'Blockquote',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.ColorSchemaMixin, UU5.Common.ContentMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Blockquote"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("blockquote"),
      bg: _bricksNs2.default.css("blockquote-bg"),
      right: 'blockquote-reverse'
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    background: _propTypes2.default.bool,
    alignment: _propTypes2.default.oneOf(['left', 'right']),
    footer: _propTypes2.default.any,
    footerAlignment: _propTypes2.default.oneOf(['left', 'right'])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      background: false,
      alignment: 'left',
      footer: null,
      footerAlignment: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _buildMainAttrs: function _buildMainAttrs() {
    var mainAttrs = this.getMainAttrs();
    this.props.background && (mainAttrs.className += ' ' + this.getClassName().bg);
    this.props.alignment === 'right' && (mainAttrs.className += ' ' + this.getClassName().right);
    return mainAttrs;
  },

  _getFooterAlignment: function _getFooterAlignment() {
    return this.props.footerAlignment || this.props.alignment;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? _react2.default.createElement(
      'blockquote',
      this._buildMainAttrs(),
      this.getChildren(),
      this.props.footer && _react2.default.createElement(_blockquoteFooter2.default, { content: this.props.footer, alignment: this._getFooterAlignment() }),
      this.getDisabledCover()
    ) : null;
  }
  //@@viewOff:render
});

exports.default = Blockquote;
