// uu5g04/dist-node/bricks/button-switch.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonSwitch = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _button = require('./button.js');

var _button2 = _interopRequireDefault(_button);

require('./button-switch.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ButtonSwitch = exports.ButtonSwitch = (0, _createReactClass2.default)({
  displayName: 'ButtonSwitch',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.ContentMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("ButtonSwitch"),
    nestingLevel: 'smallBox',
    classNames: {
      main: _bricksNs2.default.css("button-switch")
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    onProps: _propTypes2.default.object,
    offProps: _propTypes2.default.object,
    switchedOn: _propTypes2.default.bool,
    props: _propTypes2.default.object,
    size: _propTypes2.default.oneOf(['s', 'm', 'l', 'xl']),
    bgStyle: _propTypes2.default.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      onProps: null,
      offProps: null,
      switchedOn: false,
      props: null,
      size: 'm',
      bgStyle: undefined
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      switchedOn: this.props.switchedOn
    };
  },

  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (nextProps.controlled) {
      this.setState({ switchedOn: nextProps.switchedOn });
    }
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  switchOn: function switchOn(setStateCallback) {
    return this.setState({ switchedOn: true }, setStateCallback);
  },

  switchOff: function switchOff(setStateCallback) {
    return this.setState({ switchedOn: false }, setStateCallback);
  },

  toggle: function toggle(setStateCallback) {
    this.setState(function (state) {
      return { switchedOn: !state.switchedOn };
    }, setStateCallback);
    return this;
  },

  isSwitchOn: function isSwitchOn() {
    return this.state.switchedOn;
  },

  isSwitchOff: function isSwitchOff() {
    return !this.isSwitchOn();
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return _react2.default.createElement(
      _button2.default,
      _extends({}, this.getMainPropsToPass(), { size: this.props.size, bgStyle: this.props.bgStyle }, this.isSwitchOn() ? this.props.onProps : this.props.offProps, this.props.props),
      this.getChildren()
    );
  }
  //@@viewOff:render
});

exports.default = ButtonSwitch;
