// uu5g04/dist-node/bricks/column.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Column = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

require('./column.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Column = exports.Column = (0, _createReactClass2.default)({
  displayName: 'Column',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.SectionMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Column"),
    nestingLevelList: UU5.Environment.getNestingLevelList('spa', 'smallBoxCollection'),
    classNames: {
      main: _bricksNs2.default.css("column"),
      spacing: _bricksNs2.default.css("column-spacing"),
      noSpacing: _bricksNs2.default.css("column-nospacing")
    },
    defaults: {
      colWidth: { xs: 12, s: 12, m: 12, l: 12, xl: 12 },
      columnRegexp: /^((?:offset-)?[a-z]+)(?:-)?(\d+)$/
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    noSpacing: _propTypes2.default.bool,
    width: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
    // BS Column magic
    colWidth: _propTypes2.default.oneOfType([_propTypes2.default.shape({
      xs: _propTypes2.default.number,
      s: _propTypes2.default.number,
      m: _propTypes2.default.number,
      l: _propTypes2.default.number,
      xl: _propTypes2.default.number
    }), _propTypes2.default.string])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      noSpacing: false,
      width: null,
      colWidth: null
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainAttrs: function _getMainAttrs() {
    var mainAttrs = this.getMainAttrs();
    mainAttrs.className += " " + this.getClassName(this.props.noSpacing ? "noSpacing" : "spacing");

    if (this.props.width) {
      mainAttrs.style = UU5.Common.Tools.merge(mainAttrs.style, { width: this.props.width });
    } else {
      mainAttrs.className = mainAttrs.className + ' ' + UU5.Common.Tools.buildColWidthClassName(this.props.colWidth || this.getDefault().colWidth);
    }

    return mainAttrs;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? _react2.default.createElement(
      'div',
      this._getMainAttrs(),
      this.getHeaderChild(),
      this.getChildren(),
      this.getFooterChild(),
      this.getDisabledCover()
    ) : null;
  }
  //@@viewOff:render

});

exports.default = Column;
