// uu5g04/dist-node/bricks/console.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Console = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _factory = require('./factory');

var _text = require('./text');

var _text2 = _interopRequireDefault(_text);

require('./console.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

//@@viewOff:import

var Console = exports.Console = (0, _createReactClass2.default)({
  displayName: 'Console',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Console"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'smallBox'),
    classNames: {
      main: _bricksNs2.default.css("console"),
      arrow: _bricksNs2.default.css("console-with-arrow")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      content: []
    };
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  info: function info(input, setStateCallback) {
    this._log(input, "info", setStateCallback);
    return this;
  },
  warning: function warning(input, setStateCallback) {
    this._log(input, "warning", setStateCallback);
    return this;
  },
  error: function error(input, setStateCallback) {
    this._log(input, "danger", setStateCallback);
    return this;
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _log: function _log(input, color, setStateCallback) {
    var _this = this;

    this.setState(function (state) {
      var content = state.content.slice();
      content.push({
        content: _this._prepareInput(input),
        color: color
      });
      return { content: content };
    }, setStateCallback);
  },
  _prepareInput: function _prepareInput(input) {
    if (typeof input === "function") {
      return input.toString().split("{")[0];
    }
    return JSON.stringify(input, null, 2);
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var divs = [];
    for (var i = 0; i < this.state.content.length; i++) {
      divs.push(_react2.default.createElement(
        _factory.Div,
        { key: i },
        _react2.default.createElement(_text2.default, { className: this.getClassName().arrow, colorSchema: this.state.content[i].color, content: this.state.content[i].content })
      ));
    }
    return _react2.default.createElement(
      _factory.Pre,
      this.getMainPropsToPass(),
      divs
    );
  }
  //@@viewOff:render

}); //@@viewOn:import
exports.default = Console;
