// uu5g04/dist-node/bricks/context-menu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ContextMenu = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _popover = require('./popover');

var _popover2 = _interopRequireDefault(_popover);

var _contextMenuItem = require('./context-menu-item.js');

var _contextMenuItem2 = _interopRequireDefault(_contextMenuItem);

require('./context-menu.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ContextMenu = exports.ContextMenu = (0, _createReactClass2.default)({
  displayName: 'ContextMenu',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin, UU5.Common.CcrWriterMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("ContextMenu"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("context-menu"),
      ul: _bricksNs2.default.css("context-menu-ul")
    },
    defaults: {
      transitionDuration: 100,
      childTagName: 'UU5.Bricks.ContextMenu.Item'
    },
    opt: {
      nestingLevelRoot: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    shown: _propTypes2.default.bool,
    parentElement: _propTypes2.default.object,
    allowTags: _propTypes2.default.arrayOf(_propTypes2.default.string)
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      shown: false,
      parentElement: null,
      allowTags: []
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isContextMenu: function isContextMenu() {
    return true;
  },
  open: function open(opt, setStateCallback) {
    this._menu && typeof this._menu.open === "function" && this._menu.open(opt, setStateCallback);
    return this;
  },
  close: function close(setStateCallback) {
    this._menu && typeof this._menu.close === "function" && this._menu.close(setStateCallback);
    return this;
  },
  isOpen: function isOpen() {
    return this._menu && typeof this._menu.isOpen === "function" && this._menu.isOpen();
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  shouldChildRender_: function shouldChildRender_(child) {
    var childTagName = UU5.Common.Tools.getChildTagName(child);
    var childTagNames = this.props.allowTags.concat(this.getDefault().childTagName);
    return childTagNames.indexOf(childTagName) > -1;
  },

  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainAttrs: function _getMainAttrs() {
    var props = this.getMainAttrs();

    props.shown = this.props.shown;
    props.parentElement = this.props.parentElement;
    props.header = this.getHeader();
    props.footer = this.getFooter();
    props.nestingLevel = this.getNestingLevel();

    return props;
  },


  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var _this = this;

    var children = this.buildChildren({ content: this.state.content, children: this.props.children });
    return _react2.default.createElement(
      _popover2.default,
      _extends({}, this._getMainAttrs(), { ref_: function ref_(menu) {
          return _this._menu = menu;
        }, parent: this }),
      _react2.default.createElement(
        'ul',
        { className: this.getClassName().ul },
        children
      )
    );
  }
  //@@viewOff:render

});

ContextMenu.Item = _contextMenuItem2.default;
exports.default = ContextMenu;
