// uu5g04/dist-node/bricks/draggable-mixin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DraggableMixin = undefined;

var _uu5g = require("uu5g04");

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require("./bricks-ns.js");

var _bricksNs2 = _interopRequireDefault(_bricksNs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var DraggableMixin = exports.DraggableMixin = {

  //@@viewOn:statics
  statics: {
    UU5_Bricks_DraggableMixin: {
      requiredMixins: ["UU5.Common.BaseMixin"],
      classNames: {
        main: _bricksNs2.default.css("draggable")
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    // initialize
    this.registerMixin('UU5_Bricks_DraggableMixin');
    return {};
  },

  componentWillUnmount: function componentWillUnmount() {
    UU5.Environment.EventListener.removeWindowEvent('mousemove', this.getId());
    UU5.Environment.EventListener.removeWindowEvent('touchmove', this.getId());
    UU5.Environment.EventListener.removeWindowEvent('mouseup', this.getId());
    UU5.Environment.EventListener.removeWindowEvent('touchend', this.getId());
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5_Bricks_DraggableMixin: function hasUU5_Bricks_DraggableMixin() {
    return this.hasMixin('UU5_Bricks_DraggableMixin');
  },

  startDragging: function startDragging(draggedItem, x, y) {
    UU5.Environment.EventListener.addWindowEvent("mousemove", this.getId(), this._onMouseMove);
    UU5.Environment.EventListener.addWindowEvent('touchmove', this.getId(), this._onMouseMove);
    UU5.Environment.EventListener.addWindowEvent('mouseup', this.getId(), this.stopDragging);
    UU5.Environment.EventListener.addWindowEvent('touchend', this.getId(), this.stopDragging);

    this.draggedItem = draggedItem;
    this.x = x;
    this.startX = this.startX || x;
    this.y = y;
    this.startY = this.startY || y;
    this.width = UU5.Common.Tools.getWidth(this);
    this.height = UU5.Common.Tools.getHeight(this);
    this.itemMarginTop = this.getStylePropertyValue(this.draggedItem, 'margin-top');
    this.itemMarginRight = this.getStylePropertyValue(this.draggedItem, 'margin-right');
    this.itemMarginBottom = this.getStylePropertyValue(this.draggedItem, 'margin-bottom');
    this.itemMarginLeft = this.getStylePropertyValue(this.draggedItem, 'margin-left');
    return this;
  },


  dragStart: function dragStart(draggedItem, x, y) {
    // TODO: deprecated
    return this.startDragging(draggedItem, x, y);
  },

  stopDragging: function stopDragging() {
    if (this.draggedItem) {
      UU5.Environment.EventListener.removeWindowEvent('mousemove', this.getId());
      UU5.Environment.EventListener.removeWindowEvent('touchmove', this.getId());
      UU5.Environment.EventListener.removeWindowEvent('mouseup', this.getId());
      UU5.Environment.EventListener.removeWindowEvent('touchend', this.getId());

      typeof this.draggedItem.moveEnd === "function" && this.draggedItem.moveEnd();
      this.draggedItem = null;
    }
    return this;
  },


  getXOffset: function getXOffset() {
    return this.findDOMNode().offsetLeft;
  },

  getYOffset: function getYOffset() {
    return this.findDOMNode().offsetTop;
  },

  getClientLeft: function getClientLeft() {
    return this.findDOMNode().clientLeft;
  },

  getClientTop: function getClientTop() {
    return this.findDOMNode().clientTop;
  },

  getStylePropertyValue: function getStylePropertyValue(object, property) {
    return this._getNumber(window.getComputedStyle(object.findDOMNode(), null).getPropertyValue(property));
  },

  getPaddingLeft: function getPaddingLeft() {
    return this.getStylePropertyValue(this, 'padding-left');
  },

  getPaddingTop: function getPaddingTop() {
    return this.getStylePropertyValue(this, 'padding-top');
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getNumber: function _getNumber(value) {
    return Number(value.match(/\d+/)[0]);
  },

  _onMouseMove: function _onMouseMove(e) {
    e.preventDefault();
    if (this.draggedItem) {

      var posX = e.pageX || e.touches && e.touches[0].pageX;
      var posY = e.pageY || e.touches && e.touches[0].pageY;

      var x = posX - this.x;
      var y = posY - this.y;

      var itemDOMNode = this.draggedItem.findDOMNode();
      var itemOffsetLeft = itemDOMNode.offsetLeft;
      var itemOffsetTop = itemDOMNode.offsetTop;
      var itemOffsetWidth = itemDOMNode.offsetWidth;
      var itemOffsetHeight = itemDOMNode.offsetHeight;

      var parentX1 = this.getXOffset() + this.getClientLeft() + this.getPaddingLeft();
      var parentY1 = this.getYOffset() + this.getClientTop() + this.getPaddingTop();
      var itemX1 = itemOffsetLeft - this.itemMarginLeft;
      var itemY1 = itemOffsetTop - this.itemMarginTop;

      var parentX2 = this.width;
      var parentY2 = this.height;
      var itemX2 = itemOffsetLeft + itemOffsetWidth + this.itemMarginLeft + this.itemMarginRight;
      var itemY2 = itemOffsetTop + itemOffsetHeight + this.itemMarginTop + this.itemMarginBottom;

      if (itemX1 + x <= parentX1) {
        x = null;
      } else if (itemX2 + x - parentX1 - this.itemMarginLeft >= parentX2) {
        x = null;
      }

      if (itemY1 + y <= parentY1) {
        y = null;
      } else if (itemY2 + y - parentY1 - this.itemMarginTop >= parentY2) {
        y = null;
      }

      this.x = posX;
      this.y = posY;

      this.draggedItem.moveToPosition(x, y);
    }

    return this;
  }
  //@@viewOff:componentSpecificHelpers
};

exports.default = DraggableMixin;
