// uu5g04/dist-node/bricks/dropdown-item.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _link = require('./link.js');

var _link2 = _interopRequireDefault(_link);

var _icon = require('./icon.js');

var _icon2 = _interopRequireDefault(_icon);

var _span = require('./span.js');

var _span2 = _interopRequireDefault(_span);

require('./dropdown-item.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = (0, _createReactClass2.default)({
  displayName: 'dropdown-item',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Dropdown.Item"),
    nestingLevelList: UU5.Environment.getNestingLevelList('smallBoxCollection'),
    classNames: {
      main: _bricksNs2.default.css("dropdown-item"),
      link: _bricksNs2.default.css("dropdown-item-link"),
      divider: _bricksNs2.default.css("dropdown-item-divider"),
      header: _bricksNs2.default.css("dropdown-item-header"),
      subMenu: _bricksNs2.default.css("dropdown-submenu", "dropdown-menu"),
      subMenuItem: _bricksNs2.default.css("dropdown-item-submenu"),
      dropup: _bricksNs2.default.css("dropdown-item-submenu-dropup"),
      disabledItem: 'disabled'
    },
    defaults: {
      childTagName: 'UU5.Bricks.Dropdown.Item'
    },
    errors: {
      maxLevel: 'Max submenu level is 3!',
      invalidParent: 'Parent of this component is not Dropdown.'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    header: _propTypes2.default.bool,
    divider: _propTypes2.default.bool,
    label: _propTypes2.default.any, // content
    href: _propTypes2.default.string,
    onClick: _propTypes2.default.func,
    smoothScroll: _propTypes2.default.number,
    offset: _propTypes2.default.number,
    target: _propTypes2.default.string,
    dropup: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      header: false,
      divider: false,
      label: null,
      href: null,
      onClick: null,
      smoothScroll: null,
      offset: null,
      target: '_self',
      dropup: false
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function componentWillMount() {
    if (!this.getParentByType('isDropdown')) {
      this.showError("invalidParent");
    }
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  expandChildProps_: function expandChildProps_(child) {
    var newChildProps = UU5.Common.Tools.mergeDeep({}, child.props);

    newChildProps.dropup = newChildProps.dropup || this.props.dropup;

    return newChildProps;
  },

  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onClickHandler: function _onClickHandler() {
    var parent = this.getParentByType('isDropdown');
    parent && parent.close();
    return this;
  },
  _getStandardItem: function _getStandardItem(props) {
    var linkProps = {
      className: this.getClassName().link,
      content: this.props.label,
      parent: this,
      disabled: this.isDisabled(),
      colorSchema: 'custom'
    };

    if (!this.isDisabled()) {
      linkProps.onClick = this.props.onClick;
      linkProps.href = this.props.href;
      linkProps.smoothScroll = this.props.smoothScroll;
      linkProps.offset = this.props.offset;
      linkProps.target = this.props.target;

      props.onClick = this._onClickHandler;
    }

    return _react2.default.createElement(
      'li',
      props,
      _react2.default.createElement(_link2.default, linkProps)
    );
  },
  _getDividerItem: function _getDividerItem(props) {
    props.className += ' ' + this.getClassName().divider;
    return _react2.default.createElement('li', props);
  },
  _getHeaderItem: function _getHeaderItem(props) {
    props.className += ' ' + this.getClassName().header;
    return _react2.default.createElement(
      'li',
      props,
      _react2.default.createElement(_span2.default, { content: this.props.label })
    );
  },
  _getContentItem: function _getContentItem(props) {
    return _react2.default.createElement(
      'li',
      props,
      this.getChildren()
    );
  },
  _getSubMenuItem: function _getSubMenuItem(props) {
    var linkProps = {
      className: this.getClassName().link,
      parent: this,
      disabled: this.isDisabled(),
      colorSchema: 'custom'
    };

    if (!this.isDisabled()) {
      linkProps.onClick = this.props.onClick;
      linkProps.href = this.props.href;
      linkProps.smoothScroll = this.props.smoothScroll;
      linkProps.offset = this.props.offset;
      linkProps.target = this.props.target;

      this.props.onClick && (props.onClick = this._onClickHandler);
      props.className += ' ' + this.getClassName().subMenuItem;
    }

    if (this.getParent() !== null && this.getParent().getParent() !== null && this.getParent().getParent().getParent() !== null) {
      if (this.getParent().getParent().getParent().getTagName() === 'UU5.Bricks.Dropdown') {
        this.showError('maxLevel');
        return _react2.default.createElement(
          'li',
          props,
          _react2.default.createElement(_link2.default, _extends({}, linkProps, { content: this.props.label }))
        );
      }
    }

    return _react2.default.createElement(
      'li',
      props,
      _react2.default.createElement(
        _link2.default,
        linkProps,
        this.props.label,
        _react2.default.createElement(_icon2.default, { icon: 'mdi-menu-right' })
      ),
      _react2.default.createElement(
        'ul',
        { className: this.getClassName().subMenu + (this.props.dropup ? ' ' + this.getClassName().dropup : '') },
        this._getChildren()
      )
    );
  },
  _getChildren: function _getChildren() {
    var contentProps = {};

    if (this.props.items) {
      contentProps = {
        content: {
          tag: this.getDefault().childTagName,
          propsArray: this.props.items
        }
      };
    } else if (this.props.children) {
      contentProps = { children: this.props.children };
    }
    contentProps.nestingLevel = this.getNestingLevel();

    return this.buildChildren(contentProps);
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var mainAttrs = this.getMainAttrs();

    if (this.isDisabled()) {
      mainAttrs.className += ' ' + this.getClassName().disabledItem;
    }

    var result = null;

    if (this.getNestingLevel()) {
      if (this.getChildren() !== null) {
        result = this._getSubMenuItem(mainAttrs);
      } else {
        if (this.getContent() || this.props.children) {
          result = this._getContentItem(mainAttrs);
        } else if (this.props.divider) {
          result = this._getDividerItem(mainAttrs);
        } else if (this.props.header) {
          result = this._getHeaderItem(mainAttrs);
        } else {
          result = this._getStandardItem(mainAttrs);
        }
      }
    }

    return result;
  }
  //@@viewOff:render

});
