// uu5g04/dist-node/bricks/image.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Image = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

require('./image.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Image = exports.Image = (0, _createReactClass2.default)({
  displayName: 'Image',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Image"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("image"),
      type: _bricksNs2.default.css("image-"),
      disabledWrapper: 'uu5-common-disabled-cover-wrapper'
    }
  },
  //@@viewOff:statics

  // TODO: strictCircle -> no ellipse but cut a circle from different image size - e.g. http://sixrevisions.com/css/circular-images-css/
  //@@viewOn:propTypes
  propTypes: {
    type: _propTypes2.default.oneOf(['rounded', 'circle', 'thumbnail']),
    src: _propTypes2.default.string,
    responsive: _propTypes2.default.bool,
    alt: _propTypes2.default.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      type: null,
      src: '',
      responsive: true,
      alt: undefined
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getAlt: function _getAlt() {
    var alt = void 0;

    if (typeof this.props.alt !== 'undefined') {
      alt = this.props.alt;
    } else {
      alt = this.getName() || UU5.Common.Tools.getFileName(this.props.src) || '';
    }

    return alt;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var mainAttrs = this.getMainAttrs();

    this.props.type && (mainAttrs.className += ' ' + this.getClassName('type') + this.props.type);
    this.props.responsive && (mainAttrs.className += ' ' + this.getClassName('type') + 'responsive');
    mainAttrs.src = this.props.src;
    mainAttrs.alt = this._getAlt();

    var image = _react2.default.createElement('img', mainAttrs);

    if (this.isDisabled()) {
      image = _react2.default.createElement(
        'div',
        { className: this.getClassName().disabledWrapper },
        image,
        this.getDisabledCover()
      );
    }

    return this.getNestingLevel() ? image : null;
  }
  //@@viewOff:render
});

exports.default = Image;
