// uu5g04/dist-node/bricks/modal-header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _link = require('./link.js');

var _link2 = _interopRequireDefault(_link);

var _header = require('./header.js');

var _header2 = _interopRequireDefault(_header);

var _icon = require('./icon.js');

var _icon2 = _interopRequireDefault(_icon);

require('./modal-header.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = (0, _createReactClass2.default)({
  displayName: 'modal-header',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Modal.Header"),
    classNames: {
      main: _bricksNs2.default.css("modal-header"),
      close: _bricksNs2.default.css("modal-header-close"),
      title: _bricksNs2.default.css("modal-header-title")
    },
    defaults: {
      parentTagName: 'UU5.Bricks.Modal',
      closeIcon: 'mdi-close'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    _sticky: _propTypes2.default.bool,
    _onClose: _propTypes2.default.func
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      _sticky: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function componentWillMount() {
    this.checkParentTagName(this.getDefault().parentTagName);
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onCloseHandler: function _onCloseHandler(e) {
    typeof this.props._onClose === 'function' && this.props._onClose(e);
  },

  _getCloseButton: function _getCloseButton() {
    var button;
    if (!this.props._sticky) {
      button = _react2.default.createElement(
        _link2.default,
        { className: this.getClassName().close, onClick: this._onCloseHandler, colorSchema: 'custom' },
        _react2.default.createElement(_icon2.default, { icon: this.getDefault('closeIcon') })
      );
    }
    return button;
  },

  _buildChildren: function _buildChildren() {
    var content = this.getContent();
    if (typeof content === 'string') {
      return _react2.default.createElement(_header2.default, { level: 4, content: content });
    } else {
      return _react2.default.createElement(
        UU5.Common.Div,
        { content: this.props.content },
        this.props.children
      );
    }
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var mainAttrs = this.getMainAttrs();

    return _react2.default.createElement(
      'div',
      mainAttrs,
      _react2.default.createElement(
        'div',
        { className: this.getClassName('title') },
        this.getChildren()
      ),
      this._getCloseButton(),
      this.getDisabledCover()
    );
  }
  //@@viewOff:render
});
