// uu5g04/dist-node/bricks/nav-bar-header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('../create-react-class.js');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _button = require('./button.js');

var _button2 = _interopRequireDefault(_button);

var _icon = require('./icon.js');

var _icon2 = _interopRequireDefault(_icon);

require('./nav-bar-header.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = (0, _createReactClass2.default)({
  displayName: 'nav-bar-header',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("NavBar.Header"),
    classNames: {
      main: _bricksNs2.default.css("nav-bar-header navbar-header"),
      hamburger: _bricksNs2.default.css("nav-bar-header-hamburger"),
      brand: _bricksNs2.default.css("nav-bar-header-brand")
    },
    defaults: {
      parentTagName: 'UU5.Bricks.NavBar'
    },
    errors: {
      invalidParent: 'Parent of this component is not NavBar.'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    _icon: _propTypes2.default.string,
    _hamburger: _propTypes2.default.bool,
    _onOpen: _propTypes2.default.func,
    _onClose: _propTypes2.default.func
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      _icon: 'mdi-menu',
      _hamburger: true,
      _onOpen: null,
      _onClose: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function componentWillMount() {
    this.checkParentTagName(this.getDefault().parentTagName);

    if (!(this.getParent() && this.getParent().isNavBar)) {
      this.showError("invalidParent");
    }
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onClickHamburger: function _onClickHamburger() {
    var header = this;
    this.getParent() && this.getParent().toggle(function () {
      if (this.isOpen()) {
        header.props._onOpen && header.props._onOpen(this);
      } else {
        header.props._onClose && header.props._onClose(this);
      }
    });
  },

  _getHamburgerProps: function _getHamburgerProps() {
    return {
      className: this.getClassName().hamburger,
      onClick: this._onClickHamburger
    };
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var hamburger = null;
    if (this.props._hamburger) {
      hamburger = _react2.default.createElement(
        _button2.default,
        this._getHamburgerProps(),
        _react2.default.createElement(_icon2.default, { icon: this.props._icon })
      );
    }

    var children = this.getChildren();
    if (children) {
      children = _react2.default.createElement(
        'span',
        { className: this.getClassName().brand },
        children
      );
    }

    return this.getNestingLevel() ? _react2.default.createElement(
      'div',
      this.getMainAttrs(),
      hamburger,
      children,
      this.getDisabledCover()
    ) : null;
  }
  //@@viewOff:render
});
