// uu5g04/dist-node/bricks/newspaper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Newspaper = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _section = require('./section.js');

var _section2 = _interopRequireDefault(_section);

require('./newspaper.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Newspaper = exports.Newspaper = (0, _createReactClass2.default)({
  displayName: 'Newspaper',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin, UU5.Common.ColorSchemaMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Newspaper"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("newspaper uu5-common-text"),
      //bg: 'uu5-common-bg',
      columns: 'uu5-common-newspaper-layout-'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    columnsCount: _propTypes2.default.oneOf([1, 2, 3, 4, 5, 6])
    //background: PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      columnsCount: 2
      //background: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainProps: function _getMainProps() {
    var mainProps = this.getMainPropsToPass();
    mainProps.className += ' ' + this.getClassName().columns + this.props.columnsCount;
    //this.props.background && (mainProps.className += ' ' + this.getClassName().bg);
    return mainProps;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? _react2.default.createElement(
      _section2.default,
      this._getMainProps(),
      this.props.children && _react2.default.Children.toArray(this.props.children)
    ) : null;
  }
  //@@viewOff:render
});

exports.default = Newspaper;
