// uu5g04/dist-node/bricks/progress-bar-item.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

require('./progress-bar-item.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = (0, _createReactClass2.default)({
  displayName: 'progress-bar-item',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.ColorSchemaMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("ProgressBar.Item"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'smallBox'),
    classNames: {
      main: _bricksNs2.default.css("progress-bar-item"),
      striped: _bricksNs2.default.css("progress-bar-item-striped"),
      active: _bricksNs2.default.css("progress-bar-item-active")
    },
    errors: {
      invalidParent: 'Parent of this component is not ProgressBar.'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    progress: _propTypes2.default.number,
    striped: _propTypes2.default.bool,
    animated: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      progress: 0,
      striped: false,
      animated: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      progress: this.props.progress,
      content: this.getContent(),
      striped: this.props.striped,
      animated: this.props.animated
    };
  },

  componentWillMount: function componentWillMount() {
    if (!(this.getParent() && this.getParent().isProgressBar)) {
      this.showError("invalidParent");
    }
  },

  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (nextProps.controlled) {
      this.setState({
        progress: nextProps.progress,
        content: nextProps.content,
        striped: nextProps.striped,
        animated: nextProps.animated
      });
    }
    return this;
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  getProgress: function getProgress() {
    return this.state.progress;
  },

  getProgressContent: function getProgressContent() {
    return this.state.content;
  },

  isStriped: function isStriped() {
    return this.state.striped;
  },

  isAnimated: function isAnimated() {
    return this.state.animated;
  },

  setProgress: function setProgress(params, setStateCallback) {
    var newProgress = this._checkProgress(params.value);

    var newState = {
      progress: newProgress
    };

    params.content !== undefined && (newState.content = params.content);
    params.striped !== undefined && (newState.striped = params.striped);
    params.animated !== undefined && (newState.animated = params.animated);
    params.colorSchema !== undefined && (newState.colorSchema = params.colorSchema);

    this.setState(newState, setStateCallback);
    return this;
  },

  increase: function increase(params, setStateCallback) {
    var progressItem = this;
    this.setState(function (state) {
      var newProgress = progressItem._checkProgress(state.progress + params.value);

      var newState = {
        progress: newProgress
      };

      params.content !== undefined && (newState.content = params.content);
      params.striped !== undefined && (newState.striped = params.striped);
      params.animated !== undefined && (newState.animated = params.animated);
      params.colorSchema !== undefined && (newState.colorSchema = params.colorSchema);

      return newState;
    }, setStateCallback);
    return this;
  },

  decrease: function decrease(params, setStateCallback) {
    if (typeof params === 'number') {
      params = -params;
    } else {
      params.value = -params.value;
    }

    this.increase(params, setStateCallback);
    return this;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _checkProgress: function _checkProgress(newProgress) {
    var result = newProgress;
    if (newProgress > 100) {
      result = 100;
    } else if (newProgress < 0) {
      result = 0;
    }
    return result;
  },

  _getMainAttrs: function _getMainAttrs() {
    var mainAttrs = this.getMainAttrs();

    if (this.isAnimated()) {
      mainAttrs.className += ' ' + this.getClassName().striped + ' ' + this.getClassName().active;
    } else if (this.isStriped()) {
      mainAttrs.className += ' ' + this.getClassName().striped;
    }

    mainAttrs.style = { width: this.getProgress() + '%' };

    return mainAttrs;
  },

  _getChildren: function _getChildren() {
    var children;

    if (this.getProgress() > 0 && this.getProgressContent() !== '') {
      children = this.getProgressContent() || this.getProgress() + ' %';
    }

    return children;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? _react2.default.createElement(
      'div',
      this._getMainAttrs(),
      this._getChildren()
    ) : null;
  }
  //@@viewOff:render
});
