// uu5g04/dist-node/bricks/youtube-video.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.YoutubeVideo = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

require('./youtube-video.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var YoutubeVideo = exports.YoutubeVideo = (0, _createReactClass2.default)({
  displayName: 'YoutubeVideo',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("YoutubeVideo"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBox', 'box'),
    classNames: {
      main: _bricksNs2.default.css("youtube-video"),
      size: _bricksNs2.default.css("youtube-video-")
    },
    defaults: {
      source: 'https://www.youtube.com/embed/'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    autoplay: _propTypes2.default.bool,
    disableControls: _propTypes2.default.bool,
    loop: _propTypes2.default.bool,
    disableInfo: _propTypes2.default.bool,
    src: _propTypes2.default.string.isRequired,
    disableRelatedVideos: _propTypes2.default.bool,
    size: _propTypes2.default.oneOf(['s', 'm', 'l', 'xl']),
    disableFullscreen: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      autoplay: false,
      disableControls: false,
      loop: false,
      disableInfo: false,
      src: '',
      disableRelatedVideos: false,
      size: 'm',
      disableFullscreen: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _buildLink: function _buildLink() {
    var origin = this.props.src;
    var source = this.getDefault().source;
    var params = [];

    source += origin.slice(32);

    this.props.autoplay && params.push('autoplay=1');
    this.props.disableControls && params.push('controls=0');
    this.props.loop && params.push('loop=1');
    this.props.disableInfo && params.push('showinfo=0');
    this.props.disableRelatedVideos && params.push('rel=0');
    this.props.disableFullscreen && params.push('fs=0');

    if (params.length) source = source + '?' + params.join('&amp;');

    return source;
  },

  _buildMainAttrs: function _buildMainAttrs() {
    var mainProps = this.getMainAttrs();

    mainProps.className += ' ' + this.getClassName().size + this.props.size;

    mainProps.src = this._buildLink();

    return mainProps;
  },

  //@@viewOff:componentSpecificHelpers

  // Render

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? _react2.default.createElement('iframe', this._buildMainAttrs()) : null;
  }
  //@@viewOff:render
});

exports.default = YoutubeVideo;
