// uu5g04/dist-node/core/common/base-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BaseMixin = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _tools = require('./tools.js');

var _tools2 = _interopRequireDefault(_tools);

var _environment = require('../environment/environment.js');

var _environment2 = _interopRequireDefault(_environment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BaseMixin = exports.BaseMixin = {

  //@@viewOn:statics
  statics: {
    "UU5.Common.BaseMixin": {
      warnings: {
        colWidthNotUsed: 'Property colWidth was not used. Component was not covered itself by column.',
        overflowNotUsed: 'Property overflow was not used. Component was not covered itself by flc.'
      },
      errors: {
        idMissing: 'Id %s was not set.',
        parentMissing: 'Parent was not found in props.',
        childrenAreNotIndexedByParent: 'Children are not indexed by parent %s.',
        tagIsWrong: 'Wrong tag %s - element was not found.',
        notRequiredMixin: 'Registered mixin %s needs to require mixin %s in component %s!',
        invalidParentTagName: 'Parent %s is not %s.',
        invalidParentType: 'Parent %s has not function %s.'
      },
      defaults: {
        regexpChars: /([ '"{}])/g
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    id: _propTypes2.default.string,
    name: _propTypes2.default.string,
    tooltip: _propTypes2.default.any, // content - for now just string is valid
    className: _propTypes2.default.string,
    style: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.string]),
    mainAttrs: _propTypes2.default.object,
    parent: _propTypes2.default.object,
    ref_: _propTypes2.default.func,
    noIndex: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      id: null,
      name: null,
      tooltip: null,
      className: null,
      style: null,
      mainAttrs: null,
      parent: null,
      ref_: null,
      noIndex: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    // initialize
    this._isMounted = true;
    this.registerMixin("UU5.Common.BaseMixin");
    this.id = this.props.id === null ? _tools2.default.generateUUID() : this.props.id;
    // state
    return null;
  },

  componentDidMount: function componentDidMount() {
    if (typeof this.props.ref_ === 'function' && (!this.constructor.opt || !this.constructor.opt.hoc)) {
      this.props.ref_(this);
    }
  },

  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    // set id only if it was set in nextProps
    if (nextProps.id !== undefined && nextProps.id !== null && nextProps.id !== this.getId()) {
      var parent = this.getParent();
      // if parent support rendered children register - methods unregisterRenderedChild, registerRenderedChild
      if (parent && parent.unregisterRenderedChild) {
        var index = this.getIndex();
        parent.unregisterRenderedChild(this);
        this.id = nextProps.id;
        parent.registerRenderedChild && parent.registerRenderedChild(this, index);
      } else {
        this.id = nextProps.id;
      }
    }
  },

  componentWillUnmount: function componentWillUnmount() {
    // if parent support rendered children register - unregisterRenderedChild
    var parent = this.getParent();
    parent && parent.unregisterRenderedChild && parent.unregisterRenderedChild(this);
    this._isMounted = false;
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonBaseMixin: function hasUU5CommonBaseMixin() {
    return this.hasMixin("UU5.Common.BaseMixin");
  },

  getTagName: function getTagName() {
    return this.constructor.tagName || 'unknownTagName';
  },

  getMixinRegister: function getMixinRegister() {
    return this.mixinRegister = this.mixinRegister || [];
  },

  registerMixin: function registerMixin(mixinName) {
    this.getMixinRegister().push(mixinName);

    var component = this;
    var requiredMixins = this.constructor[mixinName].requiredMixins;
    if (requiredMixins && requiredMixins.length) {
      requiredMixins.forEach(function (mixin) {
        if (!component.hasMixin(mixin)) {
          component.showError('notRequiredMixin', [mixinName, mixin, component.getTagName()], {
            mixinName: "UU5.Common.BaseMixin",
            context: { requiredMixins: requiredMixins }
          });
        }
      });
    }

    return this;
  },

  hasMixin: function hasMixin(mixinName) {
    return this.getMixinRegister().indexOf(mixinName) !== -1;
  },

  getClassName: function getClassName(item, mixinName) {
    var classNames = mixinName ? this.constructor[mixinName] ? this.constructor[mixinName].classNames : null : this.constructor.classNames;
    // var resultValue = classNames && item ? classNames[item] || (classNames['main'] && classNames['main'] + '-' + Tools.getDashCase(item)) : null;
    var resultValue = classNames && item ? classNames[item] : null;
    if (typeof this[resultValue] === 'function') {
      resultValue = this[resultValue]();
    }
    return classNames && item ? resultValue : classNames;
  },

  getDefault: function getDefault(item, mixinName) {
    var defaults = mixinName ? this.constructor[mixinName] ? this.constructor[mixinName].defaults : null : this.constructor.defaults;
    return defaults && item ? defaults[item] : defaults;
  },

  getOpt: function getOpt(item, mixinName) {
    var opt = mixinName ? this.constructor[mixinName] ? this.constructor[mixinName].opt : null : this.constructor.opt;
    return opt && item ? opt[item] : opt;
  },

  getLsi: function getLsi(item, mixinName) {
    var lsi = void 0;
    if (mixinName) {
      lsi = this.constructor[mixinName] ? this.constructor[mixinName].lsi : null;
      typeof lsi === 'function' && (lsi = lsi());
    } else {
      lsi = typeof this.constructor.lsi === 'function' ? this.constructor.lsi() : this.constructor.lsi;
    }
    return lsi && item ? lsi[item] : lsi;
  },


  // TODO: deprecated
  getLSI: function getLSI(item, mixinName) {
    _tools2.default.warning("Method this.getLSI is deprecated. Use this.getLsi instead.");
    return this.getLsi(item, mixinName);
  },
  getLsiComponent: function getLsiComponent(item, mixin, params) {
    return _tools2.default.findComponent('UU5.Bricks.Lsi', { lsi: this.getLsi(item, mixin), params: params });
  },


  // TODO: deprecated
  getLSIComponent: function getLSIComponent(item, mixin, params) {
    _tools2.default.warning("Method this.getLSIComponent is deprecated. Use this.getLsiComponent instead.");
    return this.getLsiComponent(item, mixin, params);
  },
  getLsiValue: function getLsiValue(item, mixin, params) {
    return _tools2.default.getLsiItemByLanguage(this.getLsi(item, mixin), params);
  },


  // TODO: deprecated
  getLSIValue: function getLSIValue(item, mixin, params) {
    _tools2.default.warning("Method this.getLSIValue is deprecated. Use this.getLsiValue instead.");
    return this.getLsiValue(item, mixin, params);
  },
  getLsiItem: function getLsiItem(value, params) {
    return (typeof value === 'undefined' ? 'undefined' : _typeof(value)) === 'object' ? _tools2.default.getLsiItemByLanguage(value, params) : value;
  },


  // TODO: deprecated
  getLSIItem: function getLSIItem(value, params) {
    _tools2.default.warning("Method this.getLSIItem is deprecated. Use this.getLsiItem instead.");
    return this.getLsiItem(value, params);
  },


  isRendered: function isRendered() {
    return this._isMounted;
  },

  setAsyncState: function setAsyncState(newState, setStateCallback) {
    if (this.isRendered()) {
      this.setState(newState, setStateCallback);
    }
    return this;
  },


  getLimit: function getLimit(item, mixinName) {
    var limits = mixinName ? _environment2.default.limits[mixinName] : this.constructor[mixinName] ? this.constructor[mixinName].limits : this.constructor.limits;
    return limits && item ? limits[item] : limits;
  },

  getError: function getError(item, mixinName) {
    var errors = mixinName ? this.constructor[mixinName] ? this.constructor[mixinName].errors || {} : {} : this.constructor.errors || {};
    return item ? errors[item] || item : errors;
  },

  getWarning: function getWarning(item, mixinName) {
    var warnings = mixinName ? this.constructor[mixinName] ? this.constructor[mixinName].warnings || {} : {} : this.constructor.warnings || {};
    return item ? warnings[item] || item : warnings;
  },

  getCallName: function getCallName(item, mixinName) {
    var calls = mixinName ? this.constructor[mixinName] ? this.constructor[mixinName].calls || {} : {} : this.constructor.calls || {};
    return item ? calls[item] || null : calls;
  },

  getId: function getId() {
    var id = this.id;
    !id && this.showError('idMissing', id, {
      mixinName: "UU5.Common.BaseMixin"
    });
    return id;
  },

  getName: function getName() {
    return this.props.name;
  },

  getTooltip: function getTooltip() {
    return this.props.tooltip;
  },

  getProps: function getProps(prop) {
    return this.props ? prop ? this.props[prop] : this.props : null;
  },

  getUU5CommonBaseMixinProps: function getUU5CommonBaseMixinProps() {
    return {
      id: this.getId(),
      name: this.getName(),
      tooltip: this.getTooltip(),
      className: this.props.className,
      style: this.props.style,
      mainAttrs: this.props.mainAttrs,
      parent: this.getParent()
    };
  },

  getUU5CommonBaseMixinPropsToPass: function getUU5CommonBaseMixinPropsToPass(suffix) {
    suffix = suffix || 'inner';

    var props = {
      id: this.getId() + '-' + suffix,
      name: this.getName() ? this.getName() + '-' + suffix : null,
      tooltip: this.getTooltip(),
      className: this.getFullClassName(),
      style: this.props.style && this.props.mainAttrs && this.props.mainAttrs.style ? _tools2.default.mergeDeep(this.props.mainAttrs.style, this.props.style) : this.props.style || this.props.mainAttrs && this.props.mainAttrs.style,
      parent: this
    };

    var mainAttrs = this.props.mainAttrs || {};
    var extend = true;
    if (mainAttrs && mainAttrs.className) {
      extend && (mainAttrs = _tools2.default.mergeDeep({}, mainAttrs)) && (extend = false);
      delete mainAttrs.className;
    }
    if (mainAttrs && mainAttrs.style) {
      extend && (mainAttrs = _tools2.default.mergeDeep({}, mainAttrs)) && (extend = false);
      delete mainAttrs.style;
    }
    if (this.getTooltip()) {
      extend && (mainAttrs = _tools2.default.mergeDeep({}, mainAttrs));
      mainAttrs.title = this.getTooltip() || mainAttrs.title;
    }

    props.mainAttrs = mainAttrs;

    return props;
  },

  // mixinNames = mixins to choose from
  getMainPropsToPass: function getMainPropsToPass(mixinNames) {
    var params = [{}];

    var component = this;
    mixinNames = mixinNames || this.getMixinRegister();
    mixinNames.forEach(function (mixinName) {
      var propsToPassFunction = component['get' + mixinName.replace(/\./g, "") + 'PropsToPass'];
      typeof propsToPassFunction === 'function' && params.push(propsToPassFunction());
    });

    return _tools2.default.mergeDeep.apply(null, params);
  },

  getMainAttrs: function getMainAttrs() {
    var newMainAttrs = _tools2.default.mergeDeep({}, this.getUU5CommonBaseMixinProps().mainAttrs, {
      className: this.getFullClassName()
    });

    if (this.getTooltip()) newMainAttrs.title = this.getTooltip();
    if (this.props.style) {
      var style = typeof this.props.style === 'string' ? this.__styleStringToObject(this.props.style) : this.props.style;

      newMainAttrs.style = _tools2.default.mergeDeep({}, newMainAttrs.style, style);
    }
    if (this.props.id) newMainAttrs.id = this.props.id;
    if (this.getName()) newMainAttrs.name = this.getName();

    if (this.buildMainAttrs) newMainAttrs = this.buildMainAttrs(newMainAttrs);
    return newMainAttrs;
  },

  getParent: function getParent() {
    return this.props.parent;
  },

  getParentByType: function getParentByType(typeFunction) {
    var parent = this.getParent && this.getParent();
    while (parent && (typeof parent[typeFunction] !== 'function' || !parent[typeFunction]())) {
      parent = parent.getParent && parent.getParent();
    }
    return parent;
  },

  checkParentTagName: function checkParentTagName(parentTagNames) {
    parentTagNames = Array.isArray(parentTagNames) ? parentTagNames : [parentTagNames];
    var parent = this.getParent();
    var currentParentTagName = parent && parent.getTagName();
    var result = !currentParentTagName || parentTagNames.indexOf(currentParentTagName) == -1;
    if (result) {
      this.showError('invalidParentTagName', [currentParentTagName, parentTagNames.join(' or ')], {
        mixinName: "UU5.Common.BaseMixin"
      });
    }
    return result;
  },

  checkParentType: function checkParentType(typeFunction) {
    var parent = this.getParent();
    var parentTypeFunction = parent && parent[typeFunction];
    var result = parentTypeFunction !== 'function' || !parentTypeFunction();
    if (result) {
      this.showError('invalidParentType', [parent ? parent.getTagName() : null, typeFunction], {
        mixinName: "UU5.Common.BaseMixin"
      });
    }
    return result;
  },

  getIndex: function getIndex() {
    var parent = this.getParent();
    var index = null;
    // parent should support children index - method getChildIndexById
    if (parent && parent.getChildIndexById) {
      index = parent.getChildIndexById(this.getId());
    } else {
      this.showError('childrenAreNotIndexedByParent', parent ? parent.getTagName() : null, {
        mixinName: "UU5.Common.BaseMixin",
        context: { parent: parent }
      });
    }
    return index;
  },

  // Component helpers for React
  getFullClassName: function getFullClassName(className) {
    var component = this;
    var myClassName = className || 'main';
    var classArray = [];
    var classes = {
      main: this.getClassName(myClassName),
      props: this.props.className,
      mainAttrs: this.props.mainAttrs && this.props.mainAttrs.className
    };

    if (this.props.noIndex) {
      classes.main += " uu5-noindex";
    }

    this.getMixinRegister().forEach(function (v) {
      var mixinClassName = v + '_' + myClassName;
      classes[mixinClassName] = component.getClassName(myClassName, v);
      classArray.push(mixinClassName);
    });
    classArray.push(myClassName);
    classArray.push('props');
    classArray.push('mainAttrs');

    return _tools2.default.buildClasses(classes, classArray);
  },

  stringToObjectType: function stringToObjectType(string, type, prefix) {
    var calculated = prefix || window;
    if (typeof string === 'string') {
      var sArray = string.split('.');
      while (calculated && sArray.length > 0) {
        calculated = calculated[sArray.shift()];
      }
    }
    if ((typeof calculated === 'undefined' ? 'undefined' : _typeof(calculated)) !== type) {
      this.showError('tagIsWrong', string, {
        mixinName: "UU5.Common.BaseMixin",
        context: {
          notFoundObject: calculated,
          notFoundObjectType: typeof calculated === 'undefined' ? 'undefined' : _typeof(calculated),
          checkedType: type,
          prefix: prefix
        }
      });
      calculated = null;
    }
    return calculated;
  },

  switchChildrenToBody: function switchChildrenToBody(props) {
    var newProps = _tools2.default.mergeDeep({}, props || this.props);
    var children = newProps.children;
    newProps.children = null;
    newProps.body = _react2.default.Children.map(children, _tools2.default.childToBodyItem());
    return newProps;
  },

  findDOMNode: function findDOMNode() {
    return _reactDom2.default.findDOMNode(this);
  },

  exportToObject: function exportToObject() {
    return {
      tag: this.getTagName(),
      props: _tools2.default.mergeDeep({}, this.props),
      state: _tools2.default.mergeDeep({}, this.state)
    };
  },

  // msgParams could be array of params or just one param
  showError: function showError(msgKey, msgParams, opt) {
    opt = opt || {};
    var msg = this.getError(msgKey, opt.mixinName);
    _tools2.default.error.apply(null, this.__getLogParams(msg, msgParams, opt.context));
    return this;
  },

  // msgParams could be array of params or just one param
  showWarning: function showWarning(msgKey, msgParams, opt) {
    opt = opt || {};
    var msg = this.getWarning(msgKey, opt.mixinName);
    _tools2.default.warning.apply(null, this.__getLogParams(msg, msgParams, opt.context));
    return this;
  },

  replaceByHardSpace: function replaceByHardSpace(text, language) {
    var replacer = language ? _environment2.default.hardSpace.lsiReplacer[language] : null;
    replacer = replacer || _tools2.default.getLsiItemByLanguage(_environment2.default.hardSpace.lsiReplacer);
    return replacer ? replacer(text) : text;
  },

  shouldRender: function shouldRender(nextProps, nextState, thisProps, thisState) {
    var props = thisProps || this.props;
    var state = thisState || this.state;
    var oldProps = _tools2.default.merge({}, props);
    var newProps = _tools2.default.merge({}, nextProps);
    return !_tools2.default.deepEqual(oldProps, newProps) || !_tools2.default.deepEqual(state, nextState);
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  __getLogParams: function __getLogParams(msg, msgParams, context) {
    var title = this.getTagName() + ' [' + this.getId() + ']: ';
    var msgWithParams = msg ? _tools2.default.formatString(msg, msgParams) : msgParams || null;

    context = context || {};
    context.thisTagName = this.getTagName();
    context.this = this;
    context.thisProps = this.props;
    context.thisState = this.state;

    return [title + msgWithParams, context];
  },

  __styleStringToObject: function __styleStringToObject(style) {
    if (typeof style !== 'string') return;

    var obj = {};
    var formattedString = style.replace(this.getDefault('regexpChars', "UU5.Common.BaseMixin"), '');
    var properties = formattedString.split(';');

    properties.forEach(function (item) {
      var keysAndValues = item.split(':');
      if (keysAndValues.length > 1) {
        obj[keysAndValues[0]] = keysAndValues[1];
      }
    });

    return obj;
  }
  //@@viewOff:componentSpecificHelpers

};

exports.default = BaseMixin;
