// uu5g04/dist-node/core/common/calls-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CallsMixin = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _environment = require('../environment/environment.js');

var _environment2 = _interopRequireDefault(_environment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CallsMixin = exports.CallsMixin = {

  //@@viewOn:statics
  statics: {
    "UU5.Common.CallsMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"],
      errors: {
        callsNotFound: 'Property calls was not set.',
        staticsCallsNotFound: 'Variable calls was not found in statics.',
        callNameNotFound: 'Call key %s was not found in calls.',
        callNotFound: 'Call %s was not found in calls.'
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    calls: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.object])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      calls: null
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    // initialize
    this.registerMixin("UU5.Common.CallsMixin");
    // state
    return {
      calls: null
    };
  },
  componentWillMount: function componentWillMount() {
    this._setCalls(this.props.calls);
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    this._setCalls(nextProps.calls);
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonCallsMixin: function hasUU5CommonCallsMixin() {
    return this.hasMixin("UU5.Common.CallsMixin");
  },
  getUU5CommonCallsMixinProps: function getUU5CommonCallsMixinProps() {
    return {
      calls: this.getCalls()
    };
  },
  getUU5CommonCallsMixinPropsToPass: function getUU5CommonCallsMixinPropsToPass() {
    return this.getUU5CommonCallsMixinProps();
  },
  getCalls: function getCalls() {
    if (!this.state.calls) {
      this.showError('callsNotFound', null, { mixinName: "UU5.Common.CallsMixin" });
    }
    return this.state.calls;
  },
  setCalls: function setCalls(calls) {
    this._setCalls(calls);
    return this;
  },
  getCall: function getCall(item, mixinName) {
    var callNames = mixinName ? this.constructor[mixinName] ? this.constructor[mixinName].calls : null : this.constructor.calls;

    var callName = callNames && callNames[item];
    var calls = this.getCalls();
    var call = null;

    if (!callNames) {
      this.showError('staticsCallsNotFound', null, {
        mixinName: "UU5.Common.CallsMixin",
        context: {
          constructor: this.constructor
        }
      });
    } else if (!callName) {
      this.showError('callNameNotFound', item, {
        mixinName: "UU5.Common.CallsMixin",
        context: {
          calls: calls
        }
      });
    } else {
      call = calls[callName];

      if (!call) {
        this.showError('callNotFound', callName, {
          mixinName: "UU5.Common.CallsMixin",
          context: {
            calls: calls
          }
        });
      }
    }

    return call;
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _setCalls: function _setCalls(calls) {
    if (calls) {
      typeof calls === 'string' && (calls = this.stringToObjectType(calls, 'object', _environment2.default.calls));

      this.setState({ calls: calls });
    }
    return this;
  }
  //@@viewOff:componentSpecificHelpers

};

exports.default = CallsMixin;
