// uu5g04/dist-node/core/common/help.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Help = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _commonNs = require('./common-ns.js');

var _commonNs2 = _interopRequireDefault(_commonNs);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _baseMixin = require('./base-mixin.js');

var _baseMixin2 = _interopRequireDefault(_baseMixin);

var _elementaryMixin = require('./elementary-mixin.js');

var _elementaryMixin2 = _interopRequireDefault(_elementaryMixin);

var _tools = require('./tools.js');

var _tools2 = _interopRequireDefault(_tools);

var _pureRenderMixin = require('./pure-render-mixin');

var _pureRenderMixin2 = _interopRequireDefault(_pureRenderMixin);

var _contentMixin = require('./content-mixin.js');

var _contentMixin2 = _interopRequireDefault(_contentMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Help = exports.Help = (0, _createReactClass2.default)({
  displayName: 'Help',


  //@@viewOn:mixins
  mixins: [_baseMixin2.default, _elementaryMixin2.default, _pureRenderMixin2.default, _contentMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _commonNs2.default.name("Help"),
    classNames: {
      main: _commonNs2.default.css("help")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    tagName: _propTypes2.default.string,
    target: _propTypes2.default.oneOf(['_blank', '_parent', '_top', '_self'])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      tagName: "UU5.Common.Help",
      target: "_blank"
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      errorDetails: false,
      responseLink: null
    };
  },
  componentWillMount: function componentWillMount() {
    var _this = this;

    var tagNameArr = this.props.tagName.split('.');
    var libraryName = tagNameArr[0] + '.' + tagNameArr[1];
    var pageName = "/page?code=" + tagNameArr[0].toLowerCase() + tagNameArr[1] + tagNameArr[2];
    _tools2.default.loadLibrary(libraryName, function (response, error) {
      if (!error && response && response.doc) {
        _this.setState({ responseLink: response.doc + pageName });
      }
    });
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getLink: function _getLink() {
    if (this.state.responseLink) {
      return _tools2.default.findComponent('UU5.Bricks.Link', Object.assign(this.getMainPropsToPass(), {
        href: this.state.responseLink,
        content: this.props.content || this.props.children || this.props.tagName,
        target: this.props.target
      }));
    } else {
      return this.props.tagName;
    }
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this._getLink();
  }
  //@@viewOff:render
});

exports.default = Help;
