// uu5g04/dist-node/core/common/nesting-level-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NestingLevelMixin = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _environment = require('../environment/environment.js');

var _environment2 = _interopRequireDefault(_environment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NestingLevelMixin = exports.NestingLevelMixin = {

  //@@viewOn:statics
  statics: {
    "UU5.Common.NestingLevelMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"],
      defaults: {
        nestingLevel: 'boxCollection'
      },
      errors: {
        incorrectRequestedNestingLevel: 'Component requested nesting level %s is bigger than the biggest component available nesting level %s.',
        nestingLevelMismatch: 'Component with nesting level %s can not be nested into parent component with nesting level %s.'
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    nestingLevel: _propTypes2.default.oneOf(_environment2.default.nestingLevelList)
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      nestingLevel: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    // initialize
    this.registerMixin("UU5.Common.NestingLevelMixin");
    // state
    return {
      nestingLevel: this.checkNestingLevel()
    };
  },

  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    this.getNestingLevel() !== nextProps.nestingLevel && this.setState({ nestingLevel: this.checkNestingLevel() });
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonNestingLevelMixin: function hasUU5CommonNestingLevelMixin() {
    return this.hasMixin("UU5.Common.NestingLevelMixin");
  },

  getNestingLevel: function getNestingLevel() {
    return this.state.nestingLevel;
  },

  getUU5CommonNestingLevelMixinProps: function getUU5CommonNestingLevelMixinProps() {
    return {
      nestingLevel: this.props.nestingLevel
    };
  },

  getUU5CommonNestingLevelMixinPropsToPass: function getUU5CommonNestingLevelMixinPropsToPass() {
    return this.getUU5CommonNestingLevelMixinProps();
  },

  getParentNestingLevel: function getParentNestingLevel(parentNestingLevelComponent) {
    var component = parentNestingLevelComponent || this.getParentByType('hasUU5CommonNestingLevelMixin');
    return component ? component.getNestingLevel() : _environment2.default.nestingLevelList[0];
  },

  getNestingLevelList: function getNestingLevelList() {
    return this.constructor.nestingLevel && [this.constructor.nestingLevel] || this.constructor.nestingLevelList || _environment2.default.nestingLevelList;
  },

  checkNestingLevel: function checkNestingLevel() {
    //step 1 - check prop nestingLevel
    var nestingLevelEnvIndex = _environment2.default.nestingLevelList.indexOf(this.props.nestingLevel);
    var nestingLevel = nestingLevelEnvIndex > -1 ? this.props.nestingLevel : null;
    //console.log("step 1 requestedNestingLevel:",nestingLevel);

    //step 2 - check nestingLevelList
    var nestingLevelList = this.getNestingLevelList();
    //console.log("step 2 requestedNestingLevel:",nestingLevel,"nestingLevelList:",nestingLevelList,"parent::",parentNestingLevelComponent);

    //step 3 - check requested nestingLevel with build in nestingLevelList
    if (nestingLevel) {
      nestingLevelEnvIndex = _environment2.default.nestingLevelList.indexOf(nestingLevel);

      //because ie11 dont know .find()
      var nestingLevelLocIndex = void 0;
      if (typeof nestingLevelList.find === 'function') {
        nestingLevelLocIndex = nestingLevelList.indexOf(nestingLevelList.find(function (nestingLevel) {
          return _environment2.default.nestingLevelList.indexOf(nestingLevel) >= nestingLevelEnvIndex;
        }));
      } else {
        for (var i = 0; i < nestingLevelList.lenght; i++) {
          var level = nestingLevelList[i];
          if (_environment2.default.nestingLevelList.indexOf(level) >= nestingLevelEnvIndex) {
            nestingLevelLocIndex = nestingLevelList.indexOf(level);
          }
        }
      }

      nestingLevelList = nestingLevelLocIndex > -1 ? nestingLevelList.slice(nestingLevelLocIndex) : nestingLevelList;
    }

    //check nestingLevel rule incorrectRequestedNestingLevel
    if (_environment2.default.nestingLevelList.indexOf(nestingLevel) > _environment2.default.nestingLevelList.indexOf(nestingLevelList[nestingLevelList.length - 1])) {
      this.showError('incorrectRequestedNestingLevel', [nestingLevel, nestingLevelList[nestingLevelList.length - 1]], {
        mixinName: "UU5.Common.NestingLevelMixin",
        context: {
          nestingLevelList: nestingLevelList
        }
      });
      nestingLevelList = [];
    }

    //console.log("step 3 requestedNestingLevel:",nestingLevel,"requestedNestingLevelList:",nestingLevelList);

    //step 4 - build calculatedNestingLevelList from nestingLevel parent if exist
    var calculatedNestingLevelIndex = 0;
    var calculatedNestingLevelList = _environment2.default.nestingLevelList;
    var parentNestingLevelComponent = void 0;
    if (!this.getOpt('nestingLevelRoot')) {
      parentNestingLevelComponent = this.getParentByType('hasUU5CommonNestingLevelMixin');
      var parentNestingLevel = this.getParentNestingLevel(parentNestingLevelComponent);
      //console.log("step 4a",parentNestingLevel);
      calculatedNestingLevelIndex = _environment2.default.nestingLevelList.indexOf(parentNestingLevel);
      if (parentNestingLevelComponent && !parentNestingLevelComponent.getOpt('nestingLevelWrapper') && parentNestingLevel.search(/Collection$/) === -1) {
        calculatedNestingLevelIndex = Math.min(_environment2.default.nestingLevelList.length - 1, calculatedNestingLevelIndex + 1);
      }
      calculatedNestingLevelList = _environment2.default.nestingLevelList.slice(calculatedNestingLevelIndex);
    }

    //console.log("step 4",calculatedNestingLevelList);

    //step 5 - find the lowest requested nestingLevel as possible if is not requested or is not requested correctly
    if (!nestingLevel) {

      //because ie11 dont know .find()
      if (typeof nestingLevelList.find === 'function') {
        nestingLevel = nestingLevelList.find(function (nestingLevel) {
          return _environment2.default.nestingLevelList.indexOf(nestingLevel) >= calculatedNestingLevelIndex;
        });
      } else {
        for (var _i = 0; _i < nestingLevelList.length; _i++) {
          var _level = nestingLevelList[_i];
          if (_environment2.default.nestingLevelList.indexOf(_level) >= calculatedNestingLevelIndex) {
            nestingLevel = _level;
            break;
          }
        }
      }

      nestingLevelEnvIndex = _environment2.default.nestingLevelList.indexOf(nestingLevel);
      //console.log("step 5",nestingLevel)
    }

    //check nestingLevel rule nestingLevelMismatch
    if (nestingLevelEnvIndex < calculatedNestingLevelIndex) {
      this.showError('nestingLevelMismatch', [nestingLevel, _environment2.default.nestingLevelList[calculatedNestingLevelIndex]], {
        mixinName: "UU5.Common.NestingLevelMixin",
        context: {
          parent: {
            tagName: parentNestingLevelComponent ? parentNestingLevelComponent.getTagName() : null,
            component: parentNestingLevelComponent
          }
        }
      });
    }

    //step 6 - find the lowest nestingLevel as possible for render into parent

    //because ie11 dont know .find()
    if (typeof nestingLevelList.find === 'function') {
      nestingLevel = nestingLevelList.find(function (nestingLevel) {
        return calculatedNestingLevelList.indexOf(nestingLevel) > -1;
      });
    } else {
      for (var _i2 = 0; _i2 < nestingLevelList.length; _i2++) {
        var _level2 = nestingLevelList[_i2];
        if (calculatedNestingLevelList.indexOf(_level2) > -1) {
          nestingLevel = _level2;
          break;
        }
      }
    }

    //console.log("step 6",this.getTagName(),nestingLevel || null);

    return nestingLevel || null;
  }
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

};

exports.default = NestingLevelMixin;
