// uu5g04/dist-node/core/common/not-found-tag.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NotFoundTag = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _commonNs = require('./common-ns.js');

var _commonNs2 = _interopRequireDefault(_commonNs);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _baseMixin = require('./base-mixin.js');

var _baseMixin2 = _interopRequireDefault(_baseMixin);

var _elementaryMixin = require('./elementary-mixin.js');

var _elementaryMixin2 = _interopRequireDefault(_elementaryMixin);

var _error = require('./error.js');

var _error2 = _interopRequireDefault(_error);

var _pureRenderMixin = require('./pure-render-mixin');

var _pureRenderMixin2 = _interopRequireDefault(_pureRenderMixin);

require('./not-found-tag.less');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NotFoundTag = exports.NotFoundTag = (0, _createReactClass2.default)({
  displayName: 'NotFoundTag',


  //@@viewOn:mixins
  mixins: [_baseMixin2.default, _elementaryMixin2.default, _pureRenderMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _commonNs2.default.name("NotFoundTag"),
    defaults: {
      body: 'Tag not found: '
    },
    classNames: {
      main: _commonNs2.default.css("not-found-tag")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    tagName: _propTypes2.default.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      tagName: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var value = this.getDefault().body;
    this.props.tagName && (value += ' ' + this.props.tagName);
    return _react2.default.createElement(
      _error2.default,
      _extends({}, this.getMainPropsToPass(), { id: this.props.id }),
      value
    );
  }
  //@@viewOff:render
});

exports.default = NotFoundTag;
