// uu5g04/dist-node/core/common/route-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RouteMixin = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RouteMixin = exports.RouteMixin = {

  //@@viewOn:mixins
  mixins: [],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    "UU5.Common.RouteMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"]
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    params: _propTypes2.default.object
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      params: null
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    // initialize
    this.registerMixin("UU5.Common.RouteMixin");

    return {};
  },
  componentDidMount: function componentDidMount() {
    if (typeof this.onRouteChanged_ === 'function') {
      this.onRouteChanged_();
    }
  },
  componentDidUpdate: function componentDidUpdate(prevProps, prevState) {
    if (typeof this.onRouteChanged_ === 'function') {
      this.onRouteChanged_(prevProps, prevState);
    }
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonRouteMixin: function hasUU5CommonRouteMixin() {
    return this.hasMixin("UU5.Common.RouteMixin");
  },

  getUU5CommonRouteMixinProps: function getUU5CommonRouteMixinProps() {
    return {};
  },

  getUU5CommonRouteMixinPropsToPass: function getUU5CommonRouteMixinPropsToPass() {
    return this.getUU5CommonRouteMixinProps();
  }
  //@@viewOff:interface
};

exports.default = RouteMixin;
