// uu5g04/dist-node/core/common/section-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SectionMixin = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _contentMixin = require('./content-mixin.js');

var _contentMixin2 = _interopRequireDefault(_contentMixin);

var _levelMixin = require('./level-mixin.js');

var _levelMixin2 = _interopRequireDefault(_levelMixin);

var _tools = require('./tools.js');

var _tools2 = _interopRequireDefault(_tools);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SectionMixin = exports.SectionMixin = {

  //@@viewOn:mixins
  mixins: [_contentMixin2.default, _levelMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    "UU5.Common.SectionMixin": {
      requiredMixins: ["UU5.Common.BaseMixin", "UU5.Common.ContentMixin", "UU5.Common.LevelMixin"],
      defaults: {
        headerTag: 'UU5.Bricks.Header',
        footerTag: 'UU5.Bricks.Footer',
        regexpUu5Json: /(^<uu5json\/>)/
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    header: _propTypes2.default.oneOfType([
    // content body:[{tag:'',props:{}},{tag:'',props:{}},...]
    _propTypes2.default.arrayOf(_propTypes2.default.shape({
      tag: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.element]),
      props: _propTypes2.default.object
    })),
    // content bodyItem:{tag:'',props{}}
    _propTypes2.default.shape({
      tag: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.element]),
      props: _propTypes2.default.arrayOf(_propTypes2.default.object)
    }),
    // content items:{tag:'',propsArray:[{},{},{},...]}
    _propTypes2.default.shape({
      tag: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.element]),
      propsArray: _propTypes2.default.arrayOf(_propTypes2.default.object)
    }),
    // content node
    _propTypes2.default.node,
    // element
    _propTypes2.default.shape({
      element: _propTypes2.default.oneOfType([_propTypes2.default.element, _propTypes2.default.shape({
        tag: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.element]),
        props: _propTypes2.default.object
      })])
    })]),
    footer: _propTypes2.default.oneOfType([
    // content body:[{tag:'',props:{}},{tag:'',props:{}},...]
    _propTypes2.default.arrayOf(_propTypes2.default.shape({
      tag: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.element]),
      props: _propTypes2.default.object
    })),
    // content bodyItem:{tag:'',props{}}
    _propTypes2.default.shape({
      tag: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.element]),
      props: _propTypes2.default.arrayOf(_propTypes2.default.object)
    }),
    // content items:{tag:'',propsArray:[{},{},{},...]}
    _propTypes2.default.shape({
      tag: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.element]),
      propsArray: _propTypes2.default.arrayOf(_propTypes2.default.object)
    }),
    // content node
    _propTypes2.default.node,
    // element
    _propTypes2.default.shape({
      element: _propTypes2.default.oneOfType([_propTypes2.default.element, _propTypes2.default.shape({
        tag: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.element]),
        props: _propTypes2.default.object
      })])
    })]),
    underline: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      header: null,
      footer: null,
      underline: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    // initialize
    this.registerMixin("UU5.Common.SectionMixin");
    this.renderedHeaderChild = null; // renderedChild
    this.renderedFooterChild = null; // renderedChild
    // state
    return null;
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonSectionMixin: function hasUU5CommonSectionMixin() {
    return this.hasMixin("UU5.Common.SectionMixin");
  },

  getHeader: function getHeader() {
    return typeof this.getHeader_ === 'function' ? this.getHeader_() : this.props.header;
  },

  getFooter: function getFooter() {
    return typeof this.getFooter_ === 'function' ? this.getFooter_() : this.props.footer;
  },

  getHeaderType: function getHeaderType(headerContent) {
    var type = null; // one of ['contentOfStandardHeader','headerElement','headerBodyItem']
    if (headerContent || headerContent === "") {
      if (_typeof(headerContent.element) === 'object') {
        type = headerContent.element.tag ? 'headerBodyItem' : 'headerElement';
      } else if (typeof headerContent === 'string') {
        var match = this.getDefault('regexpUu5Json', "UU5.Common.SectionMixin").exec(headerContent);
        type = match ? 'uu5JSON' : 'contentOfStandardHeader';
      } else {
        type = 'contentOfStandardHeader';
      }
    }
    return type;
  },

  getFooterType: function getFooterType(footerContent) {
    var type = null; // one of ['contentOfStandardFooter','footerElement','footerBodyItem']
    if (footerContent || footerContent === "") {
      if (_typeof(footerContent.element) === 'object') {
        type = footerContent.element.tag ? 'footerBodyItem' : 'footerElement';
      } else if (typeof footerContent === 'string') {
        var match = this.getDefault('regexpUu5Json', "UU5.Common.SectionMixin").exec(footerContent);
        type = match ? 'uu5JSON' : 'contentOfStandardFooter';
      } else {
        type = 'contentOfStandardFooter';
      }
    }
    return type;
  },

  getUU5CommonSectionMixinProps: function getUU5CommonSectionMixinProps() {
    return {
      header: this.getHeader(),
      footer: this.getFooter()
    };
  },

  getUU5CommonSectionMixinPropsToPass: function getUU5CommonSectionMixinPropsToPass() {
    return _tools2.default.mergeDeep({}, this.getUU5CommonSectionMixinProps(), this.getUU5CommonContentMixinPropsToPass(), this.getUU5CommonLevelMixinPropsToPass());
  },

  registerRenderedHeaderChild: function registerRenderedHeaderChild(renderedHeaderChild) {
    if (renderedHeaderChild.hasUU5CommonBaseMixin) {
      this.renderedHeaderChild = renderedHeaderChild;
    }
    return this;
  },

  registerRenderedFooterChild: function registerRenderedFooterChild(renderedFooterChild) {
    if (renderedFooterChild.hasUU5CommonBaseMixin) {
      this.renderedFooterChild = renderedFooterChild;
    }
    return this;
  },

  expandHeaderProps: function expandHeaderProps(prevHeaderChild) {
    var newHeaderChildProps = prevHeaderChild.props;
    newHeaderChildProps = _tools2.default.mergeDeep({}, newHeaderChildProps);
    newHeaderChildProps.parent = this;
    newHeaderChildProps.id = newHeaderChildProps.id || this.getId() + '-header';
    newHeaderChildProps.underline = newHeaderChildProps.underline || this.props.underline;

    if (typeof this.expandHeaderProps_ === 'function') {
      var tempHeaderChild = this.cloneChild(prevHeaderChild, newHeaderChildProps);
      newHeaderChildProps = this.expandHeaderProps_(tempHeaderChild);
    }

    newHeaderChildProps.key = newHeaderChildProps.id;

    newHeaderChildProps.ref = function (renderedChild) {
      renderedChild && this.registerRenderedHeaderChild(renderedChild);
    }.bind(this);

    return newHeaderChildProps;
  },

  expandFooterProps: function expandFooterProps(prevFooterChild) {
    var newFooterChildProps = prevFooterChild.props;
    newFooterChildProps = _tools2.default.mergeDeep({}, newFooterChildProps);
    newFooterChildProps.parent = this;
    newFooterChildProps.id = newFooterChildProps.id || this.getId() + '-footer';

    if (typeof this.expandFooterProps_ === 'function') {
      var tempFooterChild = this.cloneChild(prevFooterChild, newFooterChildProps);
      newFooterChildProps = this.expandFooterProps_(tempFooterChild);
    }

    newFooterChildProps.key = newFooterChildProps.id;

    newFooterChildProps.ref = function (renderedChild) {
      renderedChild && this.registerRenderedFooterChild(renderedChild);
    }.bind(this);

    return newFooterChildProps;
  },

  buildHeaderChild: function buildHeaderChild(header) {
    header = header || this.getHeader();

    var headerChild = null;

    if (typeof this.buildHeaderChild_ === 'function') {
      headerChild = this.buildHeaderChild_(header);
    } else {

      var headerValue = header;
      var headerType = this.getHeaderType(headerValue);

      if (headerType === 'uu5JSON') {
        headerValue = _tools2.default.parseFromUu5JSON(headerValue);
        headerType = this.getHeaderType(headerValue);
      }

      switch (headerType) {
        case 'headerBodyItem':
          headerChild = this.buildChild(headerValue.element.tag, headerValue.element.props);
          headerChild = this.cloneChild(headerChild, this.expandHeaderProps(headerChild));
          break;
        case 'headerElement':
          headerChild = headerValue.element;
          headerChild = this.cloneChild(headerChild, this.expandHeaderProps(headerChild));
          break;
        case 'contentOfStandardHeader':
          headerChild = this.buildChild(this.getDefault('headerTag', "UU5.Common.SectionMixin"), { content: headerValue });
          headerChild = this.cloneChild(headerChild, this.expandHeaderProps(headerChild));
          break;
        default:
        // there is no header
      }
    }
    return headerChild;
  },

  buildFooterChild: function buildFooterChild(footer) {
    footer = footer || this.getFooter();

    var footerChild = null;

    if (typeof this.buildFooterChild_ === 'function') {
      footerChild = this.buildFooterChild_(footer);
    } else {

      var footerValue = footer;
      var footerType = this.getFooterType(footerValue);

      if (footerType === 'uu5JSON') {
        footerValue = _tools2.default.parseFromUu5JSON(footerValue);
        footerType = this.getFooterType(footerValue);
      }

      switch (footerType) {
        case 'footerBodyItem':
          footerChild = this.buildChild(footerValue.element.tag, footerValue.element.props);
          footerChild = this.cloneChild(footerChild, this.expandFooterProps(footerChild));
          break;
        case 'footerElement':
          footerChild = footerValue.element;
          footerChild = this.cloneChild(footerChild, this.expandFooterProps(footerChild));
          break;
        case 'contentOfStandardFooter':
          footerChild = this.buildChild(this.getDefault('footerTag', "UU5.Common.SectionMixin"), { content: footerValue });
          footerChild = this.cloneChild(footerChild, this.expandFooterProps(footerChild));
          break;
        default:
        // there is no footer
      }
    }
    return footerChild;
  },

  getHeaderChild: function getHeaderChild() {
    return this.buildHeaderChild();
  },

  getFooterChild: function getFooterChild() {
    return this.buildFooterChild();
  },

  getRenderedHeaderChild: function getRenderedHeaderChild() {
    return this.renderedHeaderChild;
  },

  getRenderedFooterChild: function getRenderedFooterChild() {
    return this.renderedFooterChild;
  }
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

};

exports.default = SectionMixin;
