// uu5g04/dist-node/core/common/session-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SessionMixin = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _environment = require('../environment/environment.js');

var _environment2 = _interopRequireDefault(_environment);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SessionMixin = exports.SessionMixin = {

  //@@viewOn:statics
  statics: {
    "UU5.Common.SessionMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"],
      errors: {
        sessionNotFound: 'Session has to be set.'
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    session: _propTypes2.default.object
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      session: null
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    // initialize
    this.registerMixin("UU5.Common.SessionMixin");
    return null;
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonSessionMixin: function hasUU5CommonSessionMixin() {
    return this.hasMixin("UU5.Common.SessionMixin");
  },
  getUU5CommonSessionMixinProps: function getUU5CommonSessionMixinProps() {
    return {
      session: this.props.session
    };
  },
  getUU5CommonSessionMixinPropsToPass: function getUU5CommonSessionMixinPropsToPass() {
    return this.getUU5CommonSessionMixinProps();
  },
  getSession: function getSession() {
    var session = this.props.session || _environment2.default.getSession();
    !session && this.showError('sessionNotFound', null, { mixinName: "UU5.Common.SessionMixin" });
    return session;
  },
  login: function login() {
    var session = this.getSession();
    session && session.login();
    return this;
  },
  logout: function logout() {
    var session = this.getSession();
    session && session.logout().catch(function (e) {
      // TODO
    });
    return this;
  }
};

exports.default = SessionMixin;
