// uu5g04/dist-node/core/common/swipe-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SwipeMixin = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _tools = require('./tools.js');

var _tools2 = _interopRequireDefault(_tools);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SwipeMixin = exports.SwipeMixin = {

  //@@viewOn:statics
  statics: {
    "UU5.Common.SwipeMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"],
      defaults: {
        maxLengthOnShortSwipe: 70,
        minLengthOnSwipe: 10,
        maxSpeedOnSlowSwipe: 0.3,
        maxAngleDivergence: 10
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  // React lifecycle
  propTypes: {
    swiped: _propTypes2.default.oneOf(['up', 'right', 'down', 'left', 'upRight', 'downRight', 'upLeft', 'downLeft']),
    // null is not false - null is not swiped
    swipedLong: _propTypes2.default.bool,
    swipedFast: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      swiped: null, // up|right|down|left|upRight|downRight|upLeft|downLeft
      swipedLong: null, // true|false
      swipedFast: null // true|false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    // initialize
    this.registerMixin("UU5.Common.SwipeMixin");
    this.lastSwipe = {
      x: 0,
      y: 0,
      dx: 0,
      dy: 0,
      ex: 0,
      ey: 0,
      left: false,
      right: false,
      up: false,
      down: false,
      long: this.props.swipedLong,
      fast: this.props.swipedFast,
      length: 0,
      speed: 0,
      startTime: null,
      endTime: null,
      angle: 0
    };
    // state
    return null;
  },

  componentWillMount: function componentWillMount() {
    this.props.swiped && (this.lastSwipe = this._getUpdatedSwipe(this.props));
  },

  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    nextProps.swiped && (this.lastSwipe = this._getUpdatedSwipe(nextProps));
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonSwipeMixin: function hasUU5CommonSwipeMixin() {
    return this.hasMixin("UU5.Common.SwipeMixin");
  },

  getUU5CommonSwipeMixinProps: function getUU5CommonSwipeMixinProps() {
    return {
      swiped: this.props.swiped,
      swipedLong: this.props.swipedLong,
      swipedFast: this.props.swipedFast
    };
  },

  getUU5CommonSwipeMixinPropsToPass: function getUU5CommonSwipeMixinPropsToPass() {
    return this.getUU5CommonSwipeMixinProps();
  },

  getSwipeLength: function getSwipeLength() {
    var dx = this.lastSwipe.dx;
    var dy = this.lastSwipe.dy;
    return Math.sqrt(Math.pow(dx, 2) + Math.pow(dy, 2));
  },

  getAngle: function getAngle() {
    return this.lastSwipe.angle;
  },

  isSwipedHorizontal: function isSwipedHorizontal() {
    return Math.abs(this.getAngle()) <= this.getDefault('maxAngleDivergence', "UU5.Common.SwipeMixin") || Math.abs(this.getAngle()) >= 180 - this.getDefault('maxLengthOnShortSwipe', "UU5.Common.SwipeMixin");
  },

  isSwipedVertical: function isSwipedVertical() {
    return Math.abs(this.getAngle()) <= 90 + this.getDefault('maxLengthOnShortSwipe', "UU5.Common.SwipeMixin") && Math.abs(this.getAngle()) >= 90 - this.getDefault('maxLengthOnShortSwipe', "UU5.Common.SwipeMixin");
  },

  isSwiped: function isSwiped() {
    return this.isSwipedUp() || this.isSwipedRight() || this.isSwipedDown() || this.isSwipedLeft();
  },

  isSwipedShort: function isSwipedShort() {
    return this.lastSwipe.long !== null && !this.lastSwipe.long;
  },

  isSwipedLong: function isSwipedLong() {
    return this.lastSwipe.long !== null && this.lastSwipe.long;
  },

  isSwipedSlow: function isSwipedSlow() {
    return this.lastSwipe.fast !== null && !this.lastSwipe.fast;
  },

  isSwipedFast: function isSwipedFast() {
    return this.lastSwipe.fast !== null && this.lastSwipe.fast;
  },

  isSwipedUp: function isSwipedUp() {
    return this.lastSwipe.up && this.isSwipedVertical();
  },

  isSwipedRight: function isSwipedRight() {
    return this.lastSwipe.right && this.isSwipedHorizontal();
  },

  isSwipedDown: function isSwipedDown() {
    return this.lastSwipe.down && this.isSwipedVertical();
  },

  isSwipedLeft: function isSwipedLeft() {
    return this.lastSwipe.left && this.isSwipedHorizontal();
  },

  // Two directions
  isSwipedUpRight: function isSwipedUpRight() {
    return this.lastSwipe.up && this.lastSwipe.right;
  },

  isSwipedDownRight: function isSwipedDownRight() {
    return this.lastSwipe.down && this.lastSwipe.right;
  },

  isSwipedUpLeft: function isSwipedUpLeft() {
    return this.lastSwipe.up && this.lastSwipe.left;
  },

  isSwipedDownLeft: function isSwipedDownLeft() {
    return this.lastSwipe.down && this.lastSwipe.left;
  },

  // Short
  isSwipedUpShort: function isSwipedUpShort() {
    return this.isSwipedUp() && this.isSwipedShort();
  },

  isSwipedRightShort: function isSwipedRightShort() {
    return this.isSwipedRight() && this.isSwipedShort();
  },

  isSwipedDownShort: function isSwipedDownShort() {
    return this.isSwipedDown() && this.isSwipedShort();
  },

  isSwipedLeftShort: function isSwipedLeftShort() {
    return this.isSwipedLeft() && this.isSwipedShort();
  },

  isSwipedUpRightShort: function isSwipedUpRightShort() {
    return this.isSwipedUpRight() && this.isSwipedShort();
  },

  isSwipedDownRightShort: function isSwipedDownRightShort() {
    return this.isSwipedDownRight() && this.isSwipedShort();
  },

  isSwipedUpLeftShort: function isSwipedUpLeftShort() {
    return this.isSwipedUpLeft() && this.isSwipedShort();
  },

  isSwipedDownLeftShort: function isSwipedDownLeftShort() {
    return this.isSwipedDownLeft() && this.isSwipedShort();
  },

  // Long
  isSwipedUpLong: function isSwipedUpLong() {
    return this.isSwipedUp() && this.isSwipedLong();
  },

  isSwipedRightLong: function isSwipedRightLong() {
    return this.isSwipedRight() && this.isSwipedLong();
  },

  isSwipedDownLong: function isSwipedDownLong() {
    return this.isSwipedDown() && this.isSwipedLong();
  },

  isSwipedLeftLong: function isSwipedLeftLong() {
    return this.isSwipedLeft() && this.isSwipedLong();
  },

  isSwipedUpRightLong: function isSwipedUpRightLong() {
    return this.isSwipedUpRight() && this.isSwipedLong();
  },

  isSwipedDownRightLong: function isSwipedDownRightLong() {
    return this.isSwipedDownRight() && this.isSwipedLong();
  },

  isSwipedUpLeftLong: function isSwipedUpLeftLong() {
    return this.isSwipedUpLeft() && this.isSwipedLong();
  },

  isSwipedDownLeftLong: function isSwipedDownLeftLong() {
    return this.isSwipedDownLeft() && this.isSwipedLong();
  },

  swipeOnTouchStart: function swipeOnTouchStart(handler, e) {
    if ((typeof handler === 'undefined' ? 'undefined' : _typeof(handler)) === 'object') {
      e = handler;
      handler = null;
    }
    e = e || handler;
    this.lastSwipe = _tools2.default.merge({}, this.lastSwipe, {
      x: e.touches[0].clientX,
      y: e.touches[0].clientY,
      dx: 0,
      dy: 0,
      length: 0,
      speed: null,
      startTime: new Date().getTime()
    });
    typeof handler === 'function' && handler(arguments);
    return this;
  },

  swipeOnTouchMove: function swipeOnTouchMove(handler, e) {
    if ((typeof handler === 'undefined' ? 'undefined' : _typeof(handler)) === 'object') {
      e = handler;
      handler = null;
    }
    e = e || handler;
    // e.preventDefault();
    this.lastSwipe = _tools2.default.merge({}, this.lastSwipe, {
      dx: e.touches[0].clientX - this.lastSwipe.x,
      dy: e.touches[0].clientY - this.lastSwipe.y,
      ex: e.touches[0].clientX,
      ey: e.touches[0].clientY,
      endTime: new Date().getTime()
    });

    (this.isSwipedLeft() || this.isSwipedRight()) && e.preventDefault();

    typeof handler === 'function' && handler(e);
    return this;
  },

  swipeOnTouchEnd: function swipeOnTouchEnd(handler, e) {
    if ((typeof handler === 'undefined' ? 'undefined' : _typeof(handler)) === 'object') {
      e = handler;
      handler = null;
    }
    e = e || handler;
    var length = this.getSwipeLength();

    if (length > this.getDefault('maxLengthOnShortSwipe', "UU5.Common.SwipeMixin")) {
      var duration = this.lastSwipe.endTime - this.lastSwipe.startTime;
      var speed = length / duration;

      this.lastSwipe = _tools2.default.merge({}, this.lastSwipe, {
        left: this.lastSwipe.dx < -this.getDefault('minLengthOnSwipe', "UU5.Common.SwipeMixin"),
        right: this.lastSwipe.dx > this.getDefault('minLengthOnSwipe', "UU5.Common.SwipeMixin"),
        down: this.lastSwipe.dy > -this.getDefault('minLengthOnSwipe', "UU5.Common.SwipeMixin") && this.lastSwipe.dy > 0,
        up: this.lastSwipe.dy < this.getDefault('minLengthOnSwipe', "UU5.Common.SwipeMixin") && this.lastSwipe.dy < 0,
        long: length > this.getDefault('maxLengthOnShortSwipe', "UU5.Common.SwipeMixin"),
        fast: speed > this.getDefault('maxLengthOnShortSwipe', "UU5.Common.SwipeMixin"),
        length: length,
        speed: length / duration,
        angle: this.angle(this.lastSwipe.x, this.lastSwipe.y, this.lastSwipe.ex, this.lastSwipe.ey)
      });

      typeof handler === 'function' && handler(arguments);
    }
    return this;
  },

  angle: function angle(x, y, ex, ey) {
    var dy = y - ey;
    var dx = ex - x;
    var theta = Math.atan2(dy, dx);
    theta *= 180 / Math.PI;
    //if (theta < 0) theta = 360 + theta;
    return theta;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getUpdatedSwipe: function _getUpdatedSwipe(props) {
    return {
      up: props.swiped === 'up' || props.swiped === 'upRight' || props.swiped === 'upLeft',
      right: props.swiped === 'right' || props.swiped === 'upRight' || props.swiped === 'downRight',
      down: props.swiped === 'down' || props.swiped === 'downRight' || props.swiped === 'downLeft',
      left: props.swiped === 'left' || props.swiped === 'upLeft' || props.swiped === 'downLeft',
      long: props.swipedLong,
      fast: props.swipedFast
    };
  }
  //@@viewOff:componentSpecificHelpers
};

exports.default = SwipeMixin;
