// uu5g04/dist-esm/forms/controls.js
import React from 'react';
import createReactClass from '../create-react-class.js';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import ns from "./forms-ns.js";
import "uu5g04-bricks";

import ControlsMixin from './mixins/controls-mixin.js';
import './controls.less';

export const Controls = createReactClass({

  //@@viewOn:mixins
  mixins: [ControlsMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Controls"),
    classNames: {
      main: ns.css("controls")
    },
    lsi: () => UU5.Environment.Lsi.Forms.form
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    buttonReset: PropTypes.bool,
    buttonValidate: PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      buttonReset: false,
      buttonValidate: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _save(e) {
    if (this.getForm().isValid()) {
      this.getForm().save();
    }

    return this;
  },

  _getMainPropsToPass() {
    let props = this.getMainPropsToPass();
    props.disabled = false;

    return props;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return <UU5.Common.Div {...this._getMainPropsToPass()}>
        {this.props.buttonReset && <UU5.Bricks.Button bgStyle="transparent" onClick={() => this.getForm().reset()} content={this.getLsiComponent('reset')} disabled={this.isDisabled()} />}
        {this.props.buttonValidate && <UU5.Bricks.Button bgStyle="transparent" onClick={() => this.getForm().validate()} content={this.getLsiComponent('validate')} disabled={this.isDisabled()} />}
        <UU5.Bricks.Button onClick={() => this.getForm().cancel()} content={this.getLsiComponent('cancel')} disabled={this.isDisabled()} />
        <UU5.Bricks.Button colorSchema="primary" onClick={this._save} content={this.getLsiComponent('ok')} disabled={this.isDisabled()} />
      </UU5.Common.Div>;
  }
  //@@viewOff:render
});

export default Controls;
