// uu5g04/dist-esm/forms/form.js
import React from 'react';
import createReactClass from '../create-react-class.js';
import * as UU5 from "uu5g04";
import ns from "./forms-ns.js";
import 'uu5g04-bricks';
import FormMixin from './mixins/form-mixin.js';

import './form.less';

export const Form = createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin, FormMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Form"),
    classNames: {
      main: ns.css("form")
    },
    lsi: () => UU5.Environment.Lsi.Forms.form
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  disable_(setStateCallback) {
    let inputs = this.getInputs();
    Object.keys(inputs).forEach(key => {
      inputs[key].disable();
    });
    this.eachFormControls(formControls => formControls.disable(setStateCallback));
    return this;
  },

  enable_(setStateCallback) {
    let inputs = this.getInputs();
    Object.keys(inputs).forEach(key => {
      inputs[key].enable();
    });
    this.eachFormControls(formControls => formControls.enable(setStateCallback));
    return this;
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getFormChildren(() => {
      return [this.getHeaderChild(), this.getChildren(), this.getFooterChild()];
    });
  }
  //@@viewOff:render
});

export default Form;
