// uu_plus4u5g01/dist-node/core/environment/environment.js
import mod from 'module';
import * as UU5 from 'uu5g04';
import {Session} from 'uu_oidcg01';

var uri = ((mod ? mod.uri : (document.currentScript || Array.prototype.slice.call(document.getElementsByTagName('script'), -1)[0] || {}).src) || '').toString();
let plus4u5BaseUrl = uri ? uri.replace(/^(.*\/).*/, '$1') : './';

const Environment = {
  name: process.env.NAME,
  mode: process.env.NODE_ENV,
  version: process.env.VERSION,
  licence: 'uu_plus4u5g01-' + process.env.VERSION + ' ©Unicorn 2017',
  basePath: plus4u5BaseUrl,
  UAFProducts: [
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-86b4d5a4a030400784764ebdb972bbda" // UAF uvod
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-25d3b166760a44b7be70e5c2eb2abaaa" // uuApp
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-ed11ec379073476db0aa295ad6c00178" // uu5
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-b858ae7d7f8041249f4830277b674990" // plus4u5
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-6f64aebc07184e9088d6e0542a8f9682" // uubmldraw
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-34df77ebe0a04adda6dcd62d32c4f513" // uuAppServer
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-34df77ebe0a04adda6dcd62d32c4f513" // uuAppServer for Ruby
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-99c939a08e0849c68df5ee339c94054b" // uuAppServer for Java
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-2590bf997d264d959b9d6a88ee1d0ff5" // uuAppServer for Node.js
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-289fcd2e11d34f3e9b2184bedb236ded" // uuCloud C3
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-3b9da1e91e854406af308a2f59113fc9" // uuCloud Operation Registry
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-8e029f520c1747d3a6b5fa270fe04f15" // uuCloud Logstore
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-e3f5c648e85f4319bd8fc25ea5be6c2c" // dockit
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-f2142743693e4b22b1753c9fb761e945" // codekit
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-eaef66592e2a40b09d97837df52e0bd5" // uu5DesignKit
    },
    {
      appCode: "uu-dockitg01-main",
      tidAwid: "78462435-215c1674232d4634b367f20ad94349c0" // component registry
    }
  ],

  uuUrlLink: 'https://plus4u.net/ues/sesm?SessFree=%s',
  registrationUrl: {
    cs: 'https://www.plus4u.net/registrace/',
    en: 'https://www.plus4u.net/en/registration/'
  },

  isDevelopment: () => (Environment.mode === 'development'),
  isProduction: () => (Environment.mode === 'production'),

  getToken: () => (UU5.Environment.getSession().getCallToken() ? UU5.Environment.getSession().getCallToken().token : null)
};

console.log(Environment.licence);

Environment.Lsi = {};
Environment.App = {};

UU5.Environment.setSession(Session);

export default Environment;
