// uu5g04/dist-node/forms/form.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Form = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('../create-react-class.js');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _formsNs = require('./forms-ns.js');

var _formsNs2 = _interopRequireDefault(_formsNs);

require('uu5g04-bricks');

var _formMixin = require('./mixins/form-mixin.js');

var _formMixin2 = _interopRequireDefault(_formMixin);

require('./form.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Form = exports.Form = (0, _createReactClass2.default)({
  displayName: 'Form',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin, _formMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _formsNs2.default.name("Form"),
    classNames: {
      main: _formsNs2.default.css("form")
    },
    lsi: function lsi() {
      return UU5.Environment.Lsi.Forms.form;
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  disable_: function disable_(setStateCallback) {
    var inputs = this.getInputs();
    Object.keys(inputs).forEach(function (key) {
      inputs[key].disable();
    });
    this.eachFormControls(function (formControls) {
      return formControls.disable(setStateCallback);
    });
    return this;
  },
  enable_: function enable_(setStateCallback) {
    var inputs = this.getInputs();
    Object.keys(inputs).forEach(function (key) {
      inputs[key].enable();
    });
    this.eachFormControls(function (formControls) {
      return formControls.enable(setStateCallback);
    });
    return this;
  },

  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var _this = this;

    return this.getFormChildren(function () {
      return [_this.getHeaderChild(), _this.getChildren(), _this.getFooterChild()];
    });
  }
  //@@viewOff:render

});

exports.default = Form;
