// uu5g04/dist-node/forms/radios.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Radios = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _formsNs = require('./forms-ns.js');

var _formsNs2 = _interopRequireDefault(_formsNs);

var _inputMixin = require('./mixins/input-mixin.js');

var _inputMixin2 = _interopRequireDefault(_inputMixin);

var _groupMixin = require('./mixins/group-mixin.js');

var _groupMixin2 = _interopRequireDefault(_groupMixin);

var _checkbox = require('./checkbox.js');

var _checkbox2 = _interopRequireDefault(_checkbox);

require('./radios.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Radios = exports.Radios = (0, _createReactClass2.default)({
  displayName: 'Radios',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, _inputMixin2.default, _groupMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _formsNs2.default.name("Radios"),
    classNames: {
      main: _formsNs2.default.css("radios"),
      inline: _formsNs2.default.css("inputs-inline")
    },
    lsi: function lsi() {
      return UU5.Environment.Lsi.Forms.message;
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function componentWillMount() {
    var value = this._getInitialValue();
    if (this.props.onValidate && typeof this.props.onValidate === 'function') {
      this._validateOnChange({ value: value, event: null, component: this });
    } else {
      this.setState({ value: value });
    }
    return this;
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    var value = this._getInitialValue(nextProps);
    if (this.props.controlled) {
      if (nextProps.value !== undefined) {
        if (this.props.onValidate && typeof this.props.onValidate === 'function') {
          this._validateOnChange({ value: value, event: null, component: this });
        } else {
          this.setState({ value: value });
        }
      }
    }
    return this;
  },


  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  reset_: function reset_(setStateCallback) {
    this.setState({
      message: this.props.message,
      feedback: this.props.feedback,
      value: this._getInitialValue(),
      readOnly: this.props.readOnly
    }, setStateCallback);
    return this;
  },

  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getInitialValue: function _getInitialValue(props) {
    props = props || this.props;
    var value = null;

    props.value.forEach(function (radio, i) {
      radio.value && (value = radio.name || i.toString());
    });

    return value;
  },

  _validateOnChange: function _validateOnChange(opt) {
    var result = typeof this.props.onValidate === 'function' ? this.props.onValidate(opt) : null;
    if (result) {
      if ((typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object') {
        if (result.feedback) {
          this.setFeedback(result.feedback, result.message, result.value);
        } else {
          this.setState({ value: opt.value });
        }
      } else {
        this.showError('validateError', null, { context: { event: e, func: this.props.onValidate, result: result } });
      }
    }
    return this;
  },
  _onChange: function _onChange(opt) {
    if (!this.isDisabled() && !this.isReadOnly()) {
      if (typeof this.props.onChange === 'function') {
        var result = this._checkRequired(opt.value); // return null || object (error)

        var radioName = this._getRadioName(opt);
        if (result && (typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object' && result.feedback === 'error') {
          //this.setError(result.message, opt.value);
        } else if (typeof this.props.onValidate === 'function') {
          result = this.props.onValidate({ value: radioName, event: opt.event, component: opt.component });
        } else {
          result = this.props.onChange({ value: radioName, event: opt.event, component: opt.component });
        }

        if (result && (typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object') {
          this.setFeedback(result.feedback, result.message, radioName);
        } else {
          this._setValue(opt); // opt.value === true
        }
      } else {
        this._setValue(opt); // opt.value === true || false
      }
    }

    return this;
  },
  _setValue: function _setValue(opt) {
    var result = this._checkRequired(opt.value);
    if (result && (typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object' && result.feedback === 'error') {
      this.setError(result.message, opt.value);
    } else if (typeof this.props.onValidate === 'function') {
      result = this.props.onValidate(opt);
      this.setFeedback(opt.feedback, opt.message, opt.value);
    } else {
      var radioName = this._getRadioName(opt);
      this.setInitial('', radioName);
    }
    return this;
  },
  _getRadioName: function _getRadioName(opt) {
    return opt.component.getName();
  },


  _checkRequired: function _checkRequired(value) {
    var options = null;

    if (this.props.required) {
      if (value) {
        if (typeof this.props.onChange !== 'function') {
          options = { feedback: 'initial', message: '', value: value };
        }
      } else {
        options = { feedback: 'error', message: this.props.requiredMessage || this.getLsiComponent('requiredMessageChoice'), value: value };
      }
    }
    return options;
  },

  // _isSelected: function (newValue) {
  //   var value = newValue ? UU5.Common.Tools.merge({}, this.getValue(), newValue) : this.getValue();
  //   return Object.keys(value).map((name) => {
  //       return value[name];
  //     }).indexOf(true) > -1;
  // },

  _getCheckboxes: function _getCheckboxes() {
    var _this = this;

    return this.props.value.map(function (box, key) {
      var checkboxProps = UU5.Common.Tools.merge({}, box);
      var disabled = _this.props.disabled ? true : !!box.disabled;
      return _react2.default.createElement(_checkbox2.default, { size: _this.props.size, key: key, value: _this.getValue() === checkboxProps.name, label: box.label, onChange: _this._onChange, onIcon: box.onIcon || _this.props.onIcon || null, offIcon: box.offIcon || _this.props.offIcon || null, name: box.name, _radio: true, labelPosition: _this.props.labelPosition, disabled: disabled });
    });
  },
  _getMainAttrs: function _getMainAttrs() {
    var attrs = this._getInputAttrs();
    if (this.props.inline) {
      attrs.className += ' ' + this.getClassName().inline;
    }
    return attrs;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var inputId = this.getId() + '-input';
    return _react2.default.createElement(
      'div',
      this._getMainAttrs(),
      this.getLabel(inputId),
      this.getInputWrapper([this._getCheckboxes()])
    );
  }
  //@@viewOn:render

});

exports.default = Radios;
