// uu5g04/dist-node/forms/text-button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TextButton = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

require('uu5g04-bricks');

var _formsNs = require('./forms-ns.js');

var _formsNs2 = _interopRequireDefault(_formsNs);

var _inputWrapper = require('./internal/input-wrapper.js');

var _inputWrapper2 = _interopRequireDefault(_inputWrapper);

var _textInput = require('./internal/text-input.js');

var _textInput2 = _interopRequireDefault(_textInput);

var _textInputMixin = require('./mixins/text-input-mixin.js');

var _textInputMixin2 = _interopRequireDefault(_textInputMixin);

var _itemList = require('./internal/item-list.js');

var _itemList2 = _interopRequireDefault(_itemList);

require('./text-button.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TextButton = exports.TextButton = (0, _createReactClass2.default)({
  displayName: 'TextButton',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, _textInputMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _formsNs2.default.name("TextButton"),
    classNames: {
      main: _formsNs2.default.css("text-button")
    },
    lsi: function lsi() {
      return UU5.Environment.Lsi.Forms.message;
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    value: _propTypes2.default.string,
    buttons: _propTypes2.default.arrayOf(_propTypes2.default.shape({
      icon: _propTypes2.default.string,
      onClick: _propTypes2.default.func,
      colorSchema: _propTypes2.default.string
    })),
    pattern: _propTypes2.default.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      value: '',
      buttons: null,
      pattern: null
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function componentWillMount() {
    if (this.props.onValidate && typeof this.props.onValidate === 'function') {
      this._validateOnChange({ value: this.state.value, event: null, component: this });
    }

    return this;
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (this.props.controlled) {
      if (nextProps.required && (nextProps.value === '' || nextProps.value === null)) {
        this.setError(nextProps.requiredMessage || this.getLsiComponent('requiredMessage'));
      } else {
        this.setFeedback(nextProps.feedback, nextProps.message, nextProps.value);
      }
    }
    return this;
  },


  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  // TODO: tohle je ještě otázka - je potřeba nastavit hodnotu z jiné komponenty (musí být validace) a z onChange (neměla by být validace)
  setValue_: function setValue_(value, setStateCallback) {
    if (this._checkRequired({ value: value })) {
      if (typeof this.props.onValidate === 'function') {
        this._validateOnChange({ value: value, event: null, component: this });
      } else {
        this.setInitial(null, value, setStateCallback);
      }
    }

    return this;
  },

  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onFocus: function _onFocus(e) {
    var opt = { value: e.target.value, event: e, component: this };

    this._onFocusButton(opt);

    return this;
  },
  _onFocusButton: function _onFocusButton(opt) {
    if (typeof this.props.onFocus === 'function') {
      this.props.onFocus(opt);
    } else {
      var result = this.getFocusFeedback(opt);
      result && this.setFeedback(result.feedback, result.message, result.value);
    }

    return this;
  },
  _validateOnChange: function _validateOnChange(opt) {
    var result = this.onValidate(opt);
    if (result) {
      if ((typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object') {
        if (result.feedback) {
          this.setFeedback(result.feedback, result.message, result.value);
        } else {
          this.setState({ value: opt.value });
        }
      } else {
        this.showError('validateError', null, { context: { event: e, func: this.props.onValidate, result: result } });
      }
    }
    return this;
  },


  /*_getFeedbackIcon(){
    let icon = this.props.required ? this.props.successIcon : null;
    switch (this.getFeedback()) {
      case 'success':
        icon = this.props.successIcon;
        break;
      case 'warning':
        icon = this.props.warningIcon;
        break;
      case 'error':
        icon = this.props.errorIcon;
        break;
    }
    return icon;
  },*/

  _getButtons: function _getButtons() {
    var _this = this;

    var result = [];
    if (!this.isReadOnly()) {
      this.props.buttons && this.props.buttons.map(function (button, key) {
        var newButton = UU5.Common.Tools.merge({
          size: _this.props.size
        }, button);
        if (typeof button.onClick === 'function') {
          newButton.onClick = function () {
            _this._onFocusButton({ value: _this.state.value, component: _this });
            button.onClick({ value: _this.state.value, component: _this });
          };
        }
        if (_this.isDisabled()) {
          newButton.disabled = true;
        }
        result.push(newButton);
      });
    }

    return result;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var _this2 = this;

    var inputId = this.getId() + '-input';

    return _react2.default.createElement(
      'div',
      this._getInputAttrs(),
      this.getLabel(inputId),
      this.getInputWrapper([_react2.default.createElement(_textInput2.default, { id: inputId, name: this.props.name || inputId, value: this.state.value, placeholder: this.props.placeholder, type: 'text', onChange: this.onChange, onBlur: this.onBlur, onFocus: this._onFocus, onKeyDown: this.onKeyDown, mainAttrs: this.props.inputAttrs, disabled: this.isDisabled() || this.isLoading(), readonly: this.isReadOnly(), loading: this.isLoading(), feedback: this.getFeedback(), ref_: function ref_(item) {
          return _this2._textInput = item;
        } }), this.state.autocompleteItems && _react2.default.createElement(
        _itemList2.default,
        this._getItemListProps(),
        this._getChildren()
      ), this.state.autocompleteItems && _react2.default.createElement(UU5.Bricks.Backdrop, this._getBackdropProps())], this._getButtons())
    );
  }
  //@@viewOn:render

});

exports.default = TextButton;
