// uu5g04/dist-node/forms/timepicker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Timepicker = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _formsNs = require('./forms-ns.js');

var _formsNs2 = _interopRequireDefault(_formsNs);

var _textInput = require('./internal/text-input.js');

var _textInput2 = _interopRequireDefault(_textInput);

var _time = require('./time.js');

var _time2 = _interopRequireDefault(_time);

var _textInputMixin = require('./mixins/text-input-mixin.js');

var _textInputMixin2 = _interopRequireDefault(_textInputMixin);

require('./timepicker.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FORMAT_AM = 'AM';
var FORMAT_PM = 'PM';
var FORMAT_12 = '12';
var FORMAT_24 = '24';
var Timepicker = exports.Timepicker = (0, _createReactClass2.default)({
  displayName: 'Timepicker',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, _textInputMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _formsNs2.default.name("Timepicker"),
    classNames: {
      main: _formsNs2.default.css("timepicker"),
      open: _formsNs2.default.css("timepicker-open"),
      menu: _formsNs2.default.css("input-menu")
    },
    defaults: {
      regexpFormat1: /^\d{1,2}:?\d{0,2} ?[PpAa]?\.?[Mm]?\.?$/,
      regexpFormat2: /^\d{1,2}:?\d{0,2}$/,
      regexpPm: /(PM|pm|Pm)/,
      regexpAm: /(AM|am|Am)/
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    value: _propTypes2.default.string,
    iconOpen: _propTypes2.default.string,
    iconClosed: _propTypes2.default.string,
    format: _propTypes2.default.oneOf([FORMAT_24, FORMAT_12]),
    nanMessage: _propTypes2.default.any
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      value: '',
      iconOpen: 'mdi-clock',
      iconClosed: 'mdi-clock',
      format: FORMAT_24,
      nanMessage: 'Please insert a valid time.'
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      open: false
    };
  },
  componentWillMount: function componentWillMount() {
    if (typeof this.props.onValidate === 'function') {
      var result = this._isValidTimeResult({ value: this._formatTime(this._parseTime(this.state.value), true) });
      if (result) {
        if ((typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object') {
          if (result.feedback) {
            this.setFeedback(result.feedback, result.message, this._formatTime(this._parseTime(result.value), true));
          } else {
            this._validateOnChange({
              value: this._formatTime(this._parseTime(this.state.value), true),
              event: null,
              component: this
            });
          }
        }
      }
    }
    return this;
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (nextProps.controlled) {
      var result = this._isValidTimeResult({ value: nextProps.value });
      if (result) {
        if ((typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object') {
          if (result.feedback) {
            this.setFeedback(result.feedback, result.message, result.value);
          } else {
            this.setFeedback(nextProps.feedback, nextProps.message, nextProps.value);
          }
        }
      }
    }
    return this;
  },
  componentWillUnmount: function componentWillUnmount() {
    this._removeEvent();
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  close: function close(setStateCallback, e) {
    var _this = this;

    this._removeEvent();
    var opt = { value: this.getValue(), event: e, component: this };
    setStateCallback = typeof setStateCallback === 'function' ? setStateCallback : function () {
      return _this._onBlurTime(opt);
    };
    this.setState({ open: false }, setStateCallback);
    return this;
  },
  toggle: function toggle(setStateCallback, e) {
    var _this2 = this;

    var opt = { value: this.getValue(), event: e, component: this };
    //setStateCallback = setStateCallback || this.isOpen() ? () => this._onFocusTime(opt) : () => this._onBlurTime(opt);
    // this.setState((state) => ({ open: !state.open }), () => this._onFocusTime(opt));
    this.setState(function (state) {
      state.open ? this._removeEvent() : this._addEvent();
      return { open: !state.open };
    }, function () {
      return _this2._onFocusTime(opt);
    });
    return this;
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  setValue_: function setValue_(value, setStateCallback) {
    if (this._checkRequired({ value: value })) {
      if (typeof this.props.onValidate === 'function') {
        this._validateOnChange({ value: value, event: null, component: this });
      } else {
        this.setInitial(null, value, setStateCallback);
      }
    }
    return this;
  },

  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _findTarget: function _findTarget(item) {
    var result = false;
    var id = this.getId();

    if (item.id === id) {
      result = true;
    } else if (item.parentElement) {
      result = this._findTarget(item.parentElement);
    }
    return result;
  },
  _addEvent: function _addEvent() {
    var _this3 = this;

    UU5.Environment.EventListener.addWindowEvent('click', this.getId(), function (e) {
      var isTimepicker = _this3._findTarget(e.target);

      if (!_this3._stopPropagation && !isTimepicker && _this3.isOpen()) {
        _this3.close();
      } else {
        _this3._stopPropagation = false;
      }
    });
    return this;
  },
  _removeEvent: function _removeEvent() {
    UU5.Environment.EventListener.removeWindowEvent('click', this.getId());
    return this;
  },
  _onChange: function _onChange(e, opt) {
    opt = opt || { value: e.target.value, event: e, component: this };

    if (opt.value === '' || this._parseTime(opt.value)) {
      if (!this.isDisabled() && !this.isReadOnly()) {
        if (typeof this.props.onChange === 'function') {
          this.props.onChange(opt);
        } else {
          if (this.props.validateOnChange) {
            this._validateOnChange(opt);
          } else {
            if (opt.value === '') {
              this.setState({ value: opt.value });
            } else if (this._checkRequired({ value: opt.value })) {
              opt.required = this.props.required;
              var result = this.getChangeFeedback(opt);

              if (!result.value || this.props.format === FORMAT_12 && result.value.match(this.getDefault().regexpFormat1) || this.props.format === FORMAT_24 && result.value.match(this.getDefault().regexpFormat2)) {
                this.setFeedback(result.feedback, result.message, result.value);
              }
            }
          }
        }
      }
    } else {
      this.setState({ value: opt.value });
    }

    return this;
  },
  _onFocus: function _onFocus(e) {
    var opt = { value: e.target.value, event: e, component: this };

    this._onFocusTime(opt);

    return this;
  },
  _onFocusTime: function _onFocusTime(opt) {
    if (typeof this.props.onFocus === 'function') {
      this.props.onFocus(opt);
    } else {
      var result = this.getFocusFeedback(opt);
      result && this.setFeedback(result.feedback, result.message, result.value);
    }

    return this;
  },
  _onBlur: function _onBlur(e) {
    var opt = { value: e.target.value, event: e, component: this };
    //if (this.isOpen()) {
    //this.close();
    //} else {
    this._onBlurTime(opt);
    //}
    return this;
  },
  _onBlurTime: function _onBlurTime(opt) {
    if (typeof this.props.onBlur === 'function') {
      this.props.onBlur(opt);
    } else {
      if (this._checkRequired({ value: opt.value }) && !this.props.validateOnChange) {
        opt.required = this.props.required;
        var blurResult = this.getBlurFeedback(opt);
        var result = this._isValidTimeResult(blurResult);
        this.setFeedback(result.feedback, result.message, this._formatTime(this._parseTime(blurResult.value), true));
      }
    }
    return this;
  },
  _isValidTimeResult: function _isValidTimeResult(opt) {
    var result = opt;
    var time = this._parseTime(opt.value);
    if (!time && opt.value) {
      result.feedback = 'error';
      result.message = this.props.nanMessage;
    }
    return result;
  },
  _validateOnChange: function _validateOnChange(opt) {
    var result = typeof this.props.onValidate === 'function' ? this.props.onValidate(opt) : null;
    if (result) {
      if ((typeof result === 'undefined' ? 'undefined' : _typeof(result)) === 'object') {
        if (result.feedback) {
          this.setFeedback(result.feedback, result.message, result.value);
        } else {
          this.setState({ value: opt.value });
        }
      } else {
        this.showError('validateError', null, { context: { event: e, func: this.props.onValidate, result: result } });
      }
    }
    return this;
  },
  _getFeedbackIcon: function _getFeedbackIcon() {
    return this.isOpen() ? this.props.iconOpen : this.props.iconClosed;
  },
  _onTimeChange: function _onTimeChange(opt) {
    var _this4 = this;

    if (!this.isDisabled() && !this.isReadOnly()) {
      if (typeof this.props.onChange === 'function') {
        this.props.onChange({
          value: this._formatTime(opt.value, true),
          event: opt.event,
          component: this
        });
      } else {
        this.setValue_(this._formatTime(opt.value, true), function () {
          return _this4._onBlurTime({
            value: _this4._formatTime(opt.value, true),
            component: _this4
          });
        });
      }
    }
    return this;
  },
  _getTimeProps: function _getTimeProps(value) {
    if (!this.state.value || this.state.value.trim() === "") {
      value = null;
    }
    return {
      className: this.getClassName().menu,
      value: value,
      ref_: this._refCalendar,
      hidden: !this.isOpen(),
      onChange: this._onTimeChange,
      format: this.props.format,
      controlled: true
    };
  },
  _formatTime: function _formatTime(value, fill0) {
    var time = '';
    if (value) {
      if (fill0) {
        time = UU5.Common.Tools.rjust(value.hours, 2, '0') + ':' + UU5.Common.Tools.rjust(value.minutes, 2, '0');
      } else {
        time = value.hours + ':' + value.minutes;
      }
      if (this.props.format === FORMAT_12) {
        time += ' ' + (value.dayPart || FORMAT_AM);
      }
    }
    return time;
  },
  _parseTime: function _parseTime(stringTime) {
    stringTime = stringTime ? stringTime.trim() : '';
    var value = null;

    if (typeof stringTime === 'string' && stringTime.trim() !== '') {
      stringTime = stringTime.trim();

      value = {
        hours: 0,
        minutes: 0
      };

      if (stringTime.indexOf(':') !== -1) {
        var dateArray = stringTime.split(':');
        value.hours = parseInt(dateArray[0].trim()) || 0;
        value.minutes = parseInt(dateArray[1].trim()) || 0;
      } else {
        value.hours = parseInt(stringTime) || 0;
      }

      if (value.hours < 0 || value.hours > 23 || value.minutes < 0 || value.minutes > 59) {
        value = null;
      } else if (this.props.format === FORMAT_12) {
        if (value.hours > 12) {
          value.hours -= 12;
        } else if (value.hours == 0) {
          value.hours = 12;
        }

        if (stringTime.match(this.getDefault().regexpPm)) {
          value.dayPart = FORMAT_PM;
        } else if (stringTime.match(this.getDefault().regexpAm)) {
          value.dayPart = FORMAT_AM;
        } else {
          value = null;
        }
      }
    }
    return value;
  },
  _getTextInputAttrs: function _getTextInputAttrs() {
    var _this5 = this;

    var props = {};
    if (!this.isReadOnly() && !this.isDisabled()) {
      props.onClick = function () {
        _this5._touchStart === undefined ? _this5.open() : _this5.toggle();
      };

      props.onTouchStart = function () {
        _this5._touchStart = new Date();
      };

      props.onTouchEnd = function (e) {
        e.preventDefault();
        if (new Date() - _this5._touchStart < UU5.Environment.holdTimeout) {
          _this5._touchStart = 0;
          e.target.click();
          e.target.focus();
        }
      };
    }

    props = UU5.Common.Tools.merge(this.props.inputAttrs, props);

    return props;
  },
  _getMainAttrs: function _getMainAttrs() {
    var attrs = this._getInputAttrs();
    attrs.id = this.getId();

    if (this.isOpen()) {
      attrs.className += ' ' + this.getClassName().open;
    }
    return attrs;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var _this6 = this;

    var inputId = this.getId() + '-input';
    var value = this._parseTime(this.state.value);

    return _react2.default.createElement(
      'div',
      this._getMainAttrs(),
      this.getLabel(inputId),
      this.getInputWrapper([_react2.default.createElement(_textInput2.default, { id: inputId, name: this.props.name || inputId, value: this.state.value || '', placeholder: this.props.placeholder, type: 'text', onChange: this._onChange, onBlur: this._onBlur, onFocus: this._onFocus, onKeyDown: this.onKeyDown, mainAttrs: this._getTextInputAttrs() || this.props.inputAttrs, disabled: this.isDisabled() || this.isLoading(), readonly: this.isReadOnly(), icon: this._getFeedbackIcon(), iconClickable: false, loading: this.isLoading(), ref_: function ref_(item) {
          return _this6._textInput = item;
        }, feedback: this.getFeedback() }), _react2.default.createElement(_time2.default, this._getTimeProps(value))])
    );
  }
  //@@viewOn:render

});

exports.default = Timepicker;
