// uu_plus4u5g01/dist-node/bricks/business-card.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from "prop-types";
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import Calls from './calls.js';

import './business-card.less';

export const BusinessCard = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.PureRenderMixin,
    UU5.Common.IdentityMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.Bricks.BusinessCard',
    classNames: {
      main: 'plus4u5-bricks-business-card'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    uuIdentity: PropTypes.string,
    fullName: PropTypes.string,
    email: PropTypes.string,
    phone: PropTypes.string,
    address: PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      uuIdentity: null,
      fullName: null,
      email: null,
      phone: null,
      address: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getUuIdentity() {
    return this.props.uuIdentity || (this.getIdentity() ? this.getIdentity().uuIdentity : null);
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {
    return <BusinessCardBody {...this.getMainPropsToPass()} uuIdentity={this._getUuIdentity()} />;
  }
  //@@viewOff:render
});

const BusinessCardBody = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.PureRenderMixin,
    UU5.Common.LoadMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.Bricks.BusinessCardBody',
    classNames: {
      main: 'plus4u5-bricks-business-card-body uu5-elevation-1',
      photo: 'plus4u5-bricks-business-card-photo',
      info: 'plus4u5-bricks-business-card-info',
      name: 'plus4u5-bricks-business-card-name',
      identity: 'plus4u5-bricks-business-card-identity'
    },
    calls: {
      onLoad: 'loadAccessRole'
    },
    lsi: () => (Plus4U5.Environment.Lsi.Bricks.bussinessCard)
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    uuIdentity: PropTypes.string,
    fullName: PropTypes.string,
    email: PropTypes.string,
    phone: PropTypes.string,
    address: PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      uuIdentity: null,
      fullName: null,
      email: null,
      phone: null,
      address: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount() {
    this.setCalls(Calls);
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  getOnLoadData_() {
    return {
      uuIdentity: this.props.uuIdentity
    }
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getName(){
    return this.state.dtoOut && this.state.dtoOut.pageEntries ? (this.state.dtoOut.pageEntries[0].name) : null;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function () {

    switch (this.getLoadFeedback()) {
      case "loading":
        return <UU5.Bricks.Loading />;
        break;
      case "ready":
        return (
          <div {...this.getMainAttrs()} >
            <Plus4U5.Bricks.UserPhoto
              uuIdentity={this.props.uuIdentity}
              className={this.getClassName("photo")}
              width={100}
            />
            <div className={this.getClassName("info")}>
              <div className={this.getClassName("name")}>
                {this._getName()}
              </div>
              <div className={this.getClassName("identity")}>
                {this.props.uuIdentity}
              </div>
            </div>
          </div>
        );
    }
  }
  //@@viewOff:render
});

export default BusinessCard;
