// uu_plus4u5g01/dist-node/bricks/calls.js
import {Uri} from "uu_appg01_core";
import {Client} from "uu_appg01";

let Calls = {

  personalRoleUris: {},

  call(method, url, params, dtoIn, clientOptions) {
    Client[method](url, params || dtoIn.data, clientOptions).then(
      response => {
        if (typeof dtoIn.done === 'function') {
          // add logged user identity information
          response = response || {};
          response.data = response.data || {};

          dtoIn.done(response.data);
        }
      },
      response => {
        if (typeof dtoIn.fail === 'function') {
          // add logged user identity information
          response = response || {};
          response.data = response.data || {};

          dtoIn.fail(response);
        }
      }
    );
  },

  getData(dtoIn) {
    if(dtoIn && dtoIn.data && dtoIn.data.src) {
      Calls.call("get", dtoIn.data.src, null, dtoIn, {
        headers: {
          'response-type': 'blob'
        }
      });
    }
  },

  getPersonalRoleUri(dtoIn) {
    if (Calls.personalRoleUris[dtoIn.data.uuIdentity]) {
      dtoIn.done(Calls.personalRoleUris[dtoIn.data.uuIdentity]);
    } else {
      let baseUri = 'https://api.plus4u.net/ues/wcp/ues/core/security/session/UESSession/getPersonalRole';
      let uriBuilder = Uri.UriBuilder.parse(baseUri);

      let commandUri = uriBuilder.toUri();
      Calls.call('get', commandUri.toString(), null, {
        done: (uri) => {
          Calls.personalRoleUris[dtoIn.data.uuIdentity] = uri;
          dtoIn.done(uri);
        },
        fail: dtoIn.fail
      });
    }
  },

  loadAccessRole: (dtoIn) => {
    Calls.getPersonalRoleUri({
      data: dtoIn.data,
      done: uri => {
        let baseUri = 'https://api.plus4u.net/ues/wcp/ues/core/artifact/UESArtifactSearch/query';
        let uriBuilder = Uri.UriBuilder.parse(baseUri);
        let commandUri = uriBuilder.toUri();
        Calls.call('get', commandUri.toString(), {uesuri: "ues:VPH-BT:VPH-BT", query: `code='${dtoIn.data.uuIdentity}'`}, dtoIn);
      },
      fail: dtoIn.fail
    });
  }
};

export default Calls;
