// uu_plus4u5g01/dist-node/bricks/image.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from 'uu_plus4u5g01';

import './image.less';

export const Image = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.Bricks.Image',
    classNames: {
      main: 'plus4u5-bricks-image'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    src: PropTypes.string,
    type: PropTypes.oneOf(['rounded', 'circle', 'thumbnail'])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      src: null,
      type: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getSrc() {
    let src = this.props.src;
    let token = Plus4U5.Environment.getToken();

    let divider = /\?/.test(src) ? '&' : '?';
    if (token) {
      let url = src + divider + 'access_token=%s';
      src = UU5.Common.Tools.formatString(url, token);
    }

    return src;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.props.src ? (
      <UU5.Bricks.Image
        {...this.props}
        {...this.getMainPropsToPass()}
        src={this._getSrc()}
      />
    ) : null;
  }
  //@@viewOff:render
});

export default Image;
