// uu_plus4u5g01/dist-node/bricks/link.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from 'uu_plus4u5g01';

import './link.less';

export const Link = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.Bricks.Link',
    classNames: {
      main: 'plus4u5-bricks-link'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    href: PropTypes.string,
    onClick: PropTypes.func,
    onCtrlClick: PropTypes.func,
    onWheelClick: PropTypes.func,
    smoothScroll: PropTypes.number,
    offset: PropTypes.number,
    target: PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getHref() {
    let href = this.props.href;
    let token = Plus4U5.Environment.getToken();

    let divider = /\?/.test(href) ? '&' : '?';
    if (token) {
      let url = href + divider + 'access_token=%s';
      href = UU5.Common.Tools.formatString(url, token);
    }

    return href;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Link
        {...this.getMainPropsToPass()}
        {...this.props}
        href={this._getHref()}
      />
    );
  }
  //@@viewOff:render
});

export default Link;
