// uu_plus4u5g01/dist-node/bricks/plus4u-link.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from 'uu_plus4u5g01';

import './plus4u-link.less';

export const Plus4ULink = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.Bricks.Plus4ULink',
    classNames: {
      main: 'plus4u5-bricks-plus4u-link'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    uri: PropTypes.string,
    target: PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      uri: null,
      target: "_blank"
    }
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onClick(target) {
    window.open(Plus4U5.Common.Tools.getLegacyUrl(this.props.uri), target || this.props.target);
    return this;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Link
        {...this.getMainPropsToPass()}
        href={Plus4U5.Common.Tools.getLegacyUrl(this.props.uri)}
        onClick={() => this._onClick()}
        onCtrlClick={() => this._onClick("_blank")}
        onWheelClick={() => this._onClick("_blank")}
        content={this.props.content}
      >
        {this.props.children}
      </UU5.Bricks.Link>
    );
  }
  //@@viewOff:render
});

export default Plus4ULink;
