// uu_plus4u5g01/dist-node/bricks/user-photo.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from 'uu_plus4u5g01';
import {Uri} from "uu_appg01_core"

import './user-photo.less';
import RichImage from './rich-image.js';

export const UserPhoto = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.IdentityMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.Bricks.UserPhoto',
    classNames: {
      main: 'plus4u5-bricks-user-photo'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    uuIdentity: PropTypes.string,
    src: PropTypes.string,
    type: PropTypes.oneOf(['rounded', 'circle', 'thumbnail']),
    width: PropTypes.number,
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      uuIdentity: null,
      src: null,
      type: null,
      width: 50,
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getUserPhoto(uuIdentity) {
    let uri;

    if (this.isAuthenticated()) {
      let token = Plus4U5.Environment.getToken();

      if (token) {
        if (this.props.src) {
          try {
            let url = Uri.UriBuilder.parse(this.props.src);
            url.setParameter('access_token', token);
            uri = url.toUri().toString();
          } catch (e) {
            uri = this.props.src;
          }
        } else {
          let url = 'https://cmd.plus4u.net/PLUS4U-BT/uu-plus4uppl/PersonalCard/getPhotoByUuId/exec?uuUri=ues:PLUS4U-BT:PLUS4U-BT&uuIdentity=%s&uuAT=Bearer%20%s';
          uri = UU5.Common.Tools.formatString(url, [uuIdentity || this.getIdentity().uuIdentity, token]);
        }
      }
    }

    uri = uri || this.props.src || (Plus4U5.Environment.basePath + 'core/assets/img/anonymous.png');

    return uri;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    const props = this.getMainPropsToPass();
    props.mainAttrs = props.mainAttrs || {};
    props.mainAttrs.style = props.mainAttrs.style || {};
    props.mainAttrs.style.width = this.props.width;

    return (
      <RichImage
        {...props}
        name={this.props.uuIdentity}
        src={this._getUserPhoto(this.props.uuIdentity)}
        shape={this.props.type}
        effect="none"
      />
    );
  }
  //@@viewOff:render
});

export default UserPhoto;
