// uu_plus4u5g01/dist-node/core/common/calls.js
import {Client} from "uu_appg01";
import {Session} from "uu_oidcg01";

export const Calls = {
  personalRoleUris: {},
  peopleCardUris: {},
  peopleCardAttributes: {},
  accessRoleUris: {},

  call(method, url, params, dtoIn, clientOptions) {
    Client[method](url, params || dtoIn.data, clientOptions).then(
      response => {
        if (typeof dtoIn.done === 'function') {
          response = response || {};
          response.data = response.data || {};
          dtoIn.done(response.data);
        }
      },
      response => {
        if (typeof dtoIn.fail === 'function') {
          response = response || {};
          response.data = response.data || {};
          dtoIn.fail(response);
        }
      }
    );
  },

  getPersonalRoleUri(done, fail, uuIdentity) {
    if (!uuIdentity) {
      let identity = Session.currentSession.getIdentity();
      uuIdentity = identity ? identity.uuIdentity : null;
    }

    if (uuIdentity) {
      if (Calls.personalRoleUris[uuIdentity]) {
        done(Calls.personalRoleUris[uuIdentity]);
      } else {
        let commandUri = "https://api.plus4u.net/ues/wcp/ues/core/security/session/UESSession/getPersonalRole";
        Calls.call('get', commandUri, null, {
          done: (uri) => {
            Calls.personalRoleUris[uuIdentity] = uri;
            typeof done === "function" && done(uri);
          },
          fail: fail
        });
      }
    }
  },

  getPeopleCardUri(uuIdentity, done, fail) {
    if (!uuIdentity) {
      let identity = Session.currentSession.getIdentity();
      uuIdentity = identity ? identity.uuIdentity : null;
    }

    if (uuIdentity) {
      if (Calls.peopleCardUris[uuIdentity]) {
        done(Calls.personalRoleUris[uuIdentity]);
      } else {
        let commandUri = "https://cmd.plus4u.net/PLUS4U-BT/uu-plus4uppl/PersonalCard/getCardList/exec";
        Calls.call('get', commandUri, { p4u_id: uuIdentity }, {
          done: (uris) => {
            if (uris && uris[0]) {
              const uri = uris[0].uri;
              Calls.peopleCardUris[uuIdentity] = uri;
              typeof done === "function" && done(uri);
            }
          },
          fail
        });
      }
    }
  },

  getPeopleCardAttributes(uuIdentity, done, fail) {
    if (uuIdentity) {
      if (Calls.peopleCardAttributes[uuIdentity]) {
        done(Calls.peopleCardAttributes[uuIdentity]);
      } else {
        Calls.getPeopleCardUri(
          uuIdentity,
          uri => {
            let commandUri = "https://cmd.plus4u.net/uu-plus4uppl/PersonalCard/getAttributes/exec";
            Calls.call('get', commandUri, { uuUri: uri }, {
              done: (attrs) => {
                Calls.peopleCardAttributes[uuIdentity] = attrs;
                typeof done === "function" && done(attrs);
              },
              fail
            });
          },
          fail
        );
      }
    }
  },

  getPropertyValue: (uri, done, fail) => {
    let commandUri = "https://api.plus4u.net/ues/wcp/ues/core/property/UESProperty/getValue";
    Calls.call('get', commandUri, { uesuri: uri }, { done, fail });
  }
};

export default Calls;
