// uu_plus4u5g01/dist-node/app/authors.js
import React from "react";
import createReactClass from 'create-react-class';
import PropTypes from "prop-types";
import * as UU5 from "uu5g04";
import * as Plus4U5 from 'uu_plus4u5g01';
import 'uu_plus4u5g01-bricks';

import "uu5g04-bricks";

import './authors.less';

export const Authors = createReactClass({
  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.SectionMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.App.Authors',
    classNames: {
      main: 'plus4u5-app-authors',
      header: 'plus4u5-app-authors-header',
      authors: 'plus4u5-app-authors-authors',
    },
    opt: {
      dummyLevel: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    leadingAuthors: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string.isRequired,
      content: PropTypes.string,
      uuIdentity: PropTypes.string,
      signature: PropTypes.string,
      role: PropTypes.any.isRequired,
    })),
    otherAuthors: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
      content: PropTypes.string,
      uuIdentity: PropTypes.string,
      role: PropTypes.any,
    })),
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      leadingAuthors: [],
      otherAuthors: null,
    }
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getAuthors(authors, size) {
    return authors.map((author, i) => <Author key={i} {...author} size={size} />)
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Section {...this.getMainPropsToPass()} content={null} level={this.props.level || 3}>
        {
          this.props.leadingAuthors &&
          <UU5.Bricks.Div className={this.getClassName().authors}
                          content={this._getAuthors(this.props.leadingAuthors, 'm')} />
        }
        {
          this.props.otherAuthors &&
          <UU5.Bricks.Div className={this.getClassName().authors}
                          content={this._getAuthors(this.props.otherAuthors, 's')} />
        }
      </UU5.Bricks.Section>
    );
  }
  //@@viewOff:render
});

export default Authors;

const Author = createReactClass({
  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.ContentMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.App.Author',
    classNames: {
      main: 'plus4u5-app-author',
      size: 'plus4u5-app-author-size',
      image: 'plus4u5-app-author-image',
      signature: 'plus4u5-app-author-signature',
      name: 'plus4u5-app-author-name',
      position: 'plus4u5-app-author-position',
      content: 'plus4u5-app-author-content',
    },
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    signature: PropTypes.string,
    src: PropTypes.string,
    uuIdentity: PropTypes.string,
    role: PropTypes.any,
    size: PropTypes.oneOf(['s', 'm'])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      signature: null,
      src: null,
      uuIdentity: null,
      role: null,
      size: 'm'
    }
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainAttrs() {
    const attrs = this.getMainPropsToPass();
    attrs.content = null;
    attrs.className += ` ${this.getClassName().size}-${(this.props.content ? 'l' : this.props.size)}`;
    return attrs;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Column {...this._getMainAttrs()}>
        <Plus4U5.Bricks.UserPhoto
          className={this.getClassName().image}
          type='circle'
          width={null}
          uuIdentity={this.props.uuIdentity}
          src={this.props.src}
        />

        {this.props.signature &&
        <UU5.Bricks.Image className={this.getClassName().signature} src={this.props.signature} />}

        <UU5.Bricks.Span className={this.getClassName().name} content={this.props.name} />
        <UU5.Bricks.Span className={this.getClassName().position} content={this.props.role} />

        {this.props.content && <UU5.Bricks.P className={this.getClassName().content} content={this.props.content} />}
      </UU5.Bricks.Column>
    );
  }
  //@@viewOff:render
});
