// uu_plus4u5g01/dist-node/app/calls.js
import {Uri} from "uu_appg01_core";
import {Client} from "uu_appg01";

let Calls = {

  personalRoleUris: {},

  call(method, url, params, dtoIn, clientOptions) {
    Client[method](url, params || dtoIn.data, clientOptions).then(
      response => {
        if (typeof dtoIn.done === 'function') {
          // add logged user identity information
          response = response || {};
          response.data = response.data || {};

          dtoIn.done(response.data);
        }
      },
      response => {
        if (typeof dtoIn.fail === 'function') {
          // add logged user identity information
          response = response || {};
          response.data = response.data || {};

          dtoIn.fail(response);
        }
      }
    );
  },

  getCommandUri: function (useCase, tidAwid) {
    tidAwid = tidAwid || '84723967990075193-c3f2af65f4884ae9bc31206fe5600fc9'; //TODO getUuMTTidAwid()
    useCase = useCase.match(/^\//) ? useCase : "/" + useCase;

    let baseUri = `https://uuos9.plus4u.net/uu-mt-main/${tidAwid}/`; //TODO uu-mt-main
    let uriBuilder = Uri.UriBuilder.parse(baseUri).setUseCase(useCase);

    return uriBuilder.toUri();
  },

  getPersonalRoleUri(dtoIn) {
    if (Calls.personalRoleUris[dtoIn.data.uuIdentity]) {
      dtoIn.done(Calls.personalRoleUris[dtoIn.data.uuIdentity]);
    } else {
      let baseUri = 'https://api.plus4u.net/ues/wcp/ues/core/security/session/UESSession/getPersonalRole';
      let uriBuilder = Uri.UriBuilder.parse(baseUri);

      let commandUri = uriBuilder.toUri();
      Calls.call('get', commandUri.toString(), null, {
        done: (uri) => {
          Calls.personalRoleUris[dtoIn.data.uuIdentity] = uri;
          dtoIn.done(uri);
        },
        fail: dtoIn.fail
      });
    }
  },

  loadActivities(dtoIn) {
    Calls.getPersonalRoleUri({
      data: dtoIn.data,
      done: uri => {
        let baseUri = 'https://api.plus4u.net/ues/wcp/ues/digitalworkspace/digitalworkspace/UESDigitalWorkspace/getActiveRecordList';
        let uriBuilder = Uri.UriBuilder.parse(baseUri);

        let commandUri = uriBuilder.toUri();
        Calls.call('get', commandUri.toString(), { uesuri: uri, query: 'isUnread=true AND pending=false' }, dtoIn);
      },
      fail: dtoIn.fail
    });
  },

  loadNavigator: (dtoIn) => {
    Calls.getPersonalRoleUri({
      data: dtoIn.data,
      done: uri => {
        let baseUri = 'https://api.plus4u.net/ues/wcp/ues/core/property/UESProperty/getValue';
        let uriBuilder = Uri.UriBuilder.parse(baseUri);

        let commandUri = uriBuilder.toUri();
        Calls.call('get', commandUri.toString(), { uesuri: uri + 'VAR.UES.COR.SYSTEM_NAVIGATOR' }, dtoIn);
      },
      fail: dtoIn.fail
    });
  }
};

export default Calls;
