// uu_plus4u5g01/dist-node/app/login.js
import React from "react";
import createReactClass from 'create-react-class';
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from 'uu_plus4u5g01';
import LoginButton from './login-button.js';

import './login.less';

export const Login = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.NestingLevelMixin,
    UU5.Common.LsiMixin,
    UU5.Common.IdentityMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.App.Login',
    nestingLevelList: UU5.Environment.getNestingLevelList('box', 'inline'),
    classNames: {
      main: 'plus4u5-app-login',
      text: 'plus4u5-app-login-text',
      buttons: 'plus4u5-app-login-buttons',
      loginText: 'plus4u5-app-login-login-text',
      registerText: 'plus4u5-app-login-register-text',
      loginButton: 'plus4u5-app-login-login-button',
      registerButton: 'plus4u5-app-login-register-button',
    },
    opt: {
      nestingLevelWrapper: true
    },
    lsi: {
      loginText: {
        cs: 'Nejprve se prosím přihlaste',
        en: 'To read the book, please log in…'
      },
      registerText: {
        cs: 'Pokud nemáte založen účet, můžete se zaregistrovat.',
        en: 'If you don’t have an account yet, feel free to make a simple registration.'
      },
      registerButton: {
        cs: 'Vytvořit účet',
        en: 'Create account'
      },
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _register() {
    let url = this.getLsiItem(Plus4U5.Environment.registrationUrl);
    window.open(url, '_blank');
    return this;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getNestingLevel() && this.isNotAuthenticated() ? (
        <UU5.Bricks.Div {...this.getMainPropsToPass()}>
          <UU5.Bricks.Div className={this.getClassName('text')}>
            <UU5.Bricks.P
              className={this.getClassName('loginText')}
              content={this.getLsiValue('loginText')}
            />
            <UU5.Bricks.P
              className={this.getClassName('registerText')}
              content={this.getLsiValue('registerText')}
            />
          </UU5.Bricks.Div>
          <UU5.Bricks.Div className={this.getClassName('buttons')}>
            <LoginButton className={this.getClassName('loginButton')} />
            <UU5.Bricks.Button
              className={this.getClassName('registerButton')}
              onClick={this._register}
              content={this.getLsiValue('registerButton')}
            />
          </UU5.Bricks.Div>
        </UU5.Bricks.Div>
      ) : null;
  }
  //@@viewOff:render
});

export default Login;
