// uu_plus4u5g01/dist-node/app/login-button.js
import React from "react";
import createReactClass from 'create-react-class';
import PropTypes from "prop-types";
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

export const LoginButton = createReactClass({

  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.PureRenderMixin,
    UU5.Common.ColorSchemaMixin,
    UU5.Common.IdentityMixin
  ],

  statics: {
    tagName: 'Plus4U5.App.LoginButton',
    classNames: {
      main: 'plus4u5-app-login-button'
    },
    lsi: {
      login: {
        cs: 'Přihlásit',
        en: 'Login'
      },
      logout: {
        cs: 'Odhlásit',
        en: 'Logout'
      },
    }
  },

  propTypes: {
    size: PropTypes.string,
    displayBlock: PropTypes.bool,
    pressed: PropTypes.bool
  },

  getDefaultProps() {
    return {
      size: 'm',
      displayBlock: false,
      pressed: false
    };
  },

  render() {
    return (
      <UU5.Bricks.ButtonSwitch
        {...this.getMainPropsToPass()}
        colorSchema={this.props.colorSchema}
        size={this.props.size}
        props={{
          displayBlock: this.props.displayBlock,
          pressed: this.props.pressed
        }}
        switchedOn={this.isAuthenticated()}
        onProps={{ content: this.getLsiComponent('logout'), onClick: () => this.logout() }}
        offProps={{ content: this.getLsiComponent('login'), onClick: () => this.login() }}
      >
      </UU5.Bricks.ButtonSwitch>
    );
  },
});

export default LoginButton;
