// uu_plus4u5g01/dist-node/app/page.js
import React from "react";
import createReactClass from 'create-react-class';
import PropTypes from "prop-types";
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import Top from './top.js';
import Bottom from './bottom.js';
import Button from './button.js';
import NotificationBadge from './notification-badge.js';

import './page.less';

export const Page = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.NestingLevelMixin,
    UU5.Common.ContentMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.App.Page',
    nestingLevel: 'page',
    classNames: {
      main: 'plus4u5-app-page',
      topWrapper: 'plus4u5-app-page-top-wrapper',
      bottomWrapper: 'plus4u5-app-page-bottom-wrapper',
      leftWrapper: 'plus4u5-app-page-left-wrapper',
      rightWrapper: 'plus4u5-app-page-right-wrapper',
      contentWrapper: 'plus4u5-app-page-content-wrapper',
      systemLayerWrapper: 'plus4u5-app-page-system-layer-wrapper',
      appLayerWrapper: 'plus4u5-app-page-app-layer-wrapper',
      languageSelector: 'plus4u5-app-page-language-selector'
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    displayedLanguages: PropTypes.oneOfType([
      PropTypes.arrayOf(
        PropTypes.string
      ),
      PropTypes.string,
    ]),
    fullPage: PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      displayedLanguages: null,
      fullPage: true
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount() {
    this.props.userLayerContent && UU5.Common.Tools.warning("Property 'userLayerContent' is deprecated! Use 'appLayerContent' instead.");
    this.props.userLayerWrapperProps && UU5.Common.Tools.warning("Property 'userLayerWrapperProps' is deprecated! Use 'appLayerWrapperProps' instead.");
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getLanguageSelector(){
    return <UU5.Bricks.LanguageSelector
      headerMode='code'
      bodyMode='label-code'
      displayedLanguages={this.props.displayedLanguages}
      className={this.getClassName('languageSelector')}
    />;
  },

  _expandWrapperProps(props, className) {
    const wrapperProps = UU5.Common.Tools.mergeDeep({}, props) || {};
    wrapperProps.className = wrapperProps.className || '';
    wrapperProps.className += ' ' + this.getClassName(className);
    return wrapperProps;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    const topWrapperProps = this._expandWrapperProps(this.props.topWrapperProps, 'topWrapper');
    const bottomWrapperProps = this._expandWrapperProps(this.props.bottomWrapperProps, 'bottomWrapper');
    const leftWrapperProps = this._expandWrapperProps(this.props.leftWrapperProps, 'leftWrapper');
    const rightWrapperProps = this._expandWrapperProps(this.props.rightWrapperProps, 'rightWrapper');
    const contentWrapperProps = this._expandWrapperProps(this.props.contentWrapperProps, 'contentWrapper');
    const appLayerWrapperProps = this._expandWrapperProps(this.props.appLayerWrapperProps || this.props.userLayerWrapperProps, 'appLayerWrapper');
    const systemLayerWrapperProps = this._expandWrapperProps(this.props.systemLayerWrapperProps, 'systemLayerWrapper');

    return (
      (this.getNestingLevel() === 'page')
        ? (
        <UU5.Bricks.Page
          {...this.getMainPropsToPass()}
          type={this.props.type || 1}
          fullPage={this.props.fullPage}
          topWrapperProps={topWrapperProps}
          bottomWrapperProps={bottomWrapperProps}
          leftWrapperProps={leftWrapperProps}
          rightWrapperProps={rightWrapperProps}
          contentWrapperProps={contentWrapperProps}
          appLayerWrapperProps={appLayerWrapperProps}
          systemLayerWrapperProps={systemLayerWrapperProps}
          top={this.props.top === undefined ? <Top /> : this.props.top}
          bottom={this.props.bottom === undefined ? <Bottom /> : this.props.bottom}
          left={this.props.left}
          right={this.props.right}
          leftOpen={this.props.leftOpen}
          leftClosed={this.props.leftClosed}
          rightOpen={this.props.rightOpen}
          rightClosed={this.props.rightClosed}
          leftWidth={this.props.leftWidth}
          rightWidth={this.props.rightWidth}
          alertBus={this.props.alertBus === undefined ? <UU5.Bricks.AlertBus /> : this.props.alertBus}
          modal={this.props.modal === undefined ? <UU5.Bricks.Modal /> : this.props.modal}
          systemLayerContent={this.props.systemLayerContent === undefined ?
            [this._getLanguageSelector(), <Button />, <NotificationBadge />, <UU5.Bricks.ButtonToTop/>] : this.props.systemLayerContent}
          appLayerContent={this.props.appLayerContent || this.props.userLayerContent}
        >
          {this.props.children}
        </UU5.Bricks.Page>
      ) : null
    );
  },
  //@@viewOff:render

});

export default Page;
