// uu_plus4u5g01/dist-node/app/resources.js
import React from "react";
import createReactClass from 'create-react-class';
import PropTypes from "prop-types";
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import './resources.less';

export const Resources = createReactClass({
  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.SectionMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.App.Resources',
    classNames: {
      main: 'plus4u5-app-resources',
      resources: 'plus4u5-app-resources-resources',
      resource: 'plus4u5-app-resources-resource',
    },
    lsi: {
      header: {
        cs: 'Použité zdroje',
        en: 'Resources',
      }
    },
    opt: {
      dummyLevel: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    resources: PropTypes.array
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      resources: []
    }
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getResources() {
    return this.props.resources.map((text, i) =>
      <UU5.Bricks.Div className={this.getClassName().resource} key={i} content={text} />
    );
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    const header = this.props.header ? this.props.header : this.getLsiComponent('header');

    return (
      <UU5.Bricks.Section
        {...this.getMainPropsToPass()}
        level={this.props.level || 4}
        header={header}
        content={this._getResources()}
      />
    );
  }
  //@@viewOff:render
});

export default Resources;
