// uu_plus4u5g01/dist-node/app/top.js
import React from "react";
import createReactClass from 'create-react-class';
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import './top.less';

export const Top = createReactClass({

  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.NestingLevelMixin,
    UU5.Common.CcrReaderMixin,
    UU5.Common.ContentMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: 'Plus4U5.App.Top',
    nestingLevel: 'box',
    classNames: {
      main: 'plus4u5-app-top',
      leftMenu: 'plus4u5-app-top-left-menu',
      content: 'plus4u5-app-top-left-content',
      buttonGroup: 'plus4u5-app-top-left-button-group'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getPropsToPass() {
    return UU5.Common.Tools.merge(this.getMainPropsToPass(),{content:null})
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    //TODO - ContainerCollection
    return (
      this.getNestingLevel()
        ?
        <UU5.Bricks.Div {...this._getPropsToPass()}>
          <UU5.Bricks.Page.MenuButton
            menu='left'
            colorSchema="custom"
            className={this.getClassName('leftMenu')}
          />
          <UU5.Bricks.Span
            className={this.getClassName('content')}
            content={this.props.content || this.props.children}
          />
        </UU5.Bricks.Div>
       : null
    );
  },
  //@@viewOff:render

});

export default Top;
