// uu5g04/dist-node/bricks/alert-bus.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertBus = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _alert = require('./alert.js');

var _alert2 = _interopRequireDefault(_alert);

require('./alert-bus.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AlertBus = exports.AlertBus = (0, _createReactClass2.default)({
  displayName: 'AlertBus',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.CcrWriterMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("AlertBus"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("alert-bus")
    },
    warnings: {
      noMessage: 'Alert "%s" is not set.'
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    // has not color schema because color schema css class is added automatically to all alerts,
    // but we just need to set color schema of added alert
    colorSchema: _propTypes2.default.oneOf(UU5.Environment.colorSchema),
    position: _propTypes2.default.string,
    closeTimer: _propTypes2.default.number,
    closeDisabled: _propTypes2.default.bool,
    block: _propTypes2.default.bool,
    forceRender: _propTypes2.default.bool,
    onClose: _propTypes2.default.func
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      colorSchema: null,
      position: 'center',
      closeTimer: 10000,
      closeDisabled: false,
      block: false,
      forceRender: false,
      onClose: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      alertStack: []
    };
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  addAlert: function addAlert(alertProps, setStateCallback) {
    if (alertProps.content) {
      var messageProps = this._getMessageProps(null, alertProps);
      var page = this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE);
      if (!this.props.forceRender && page && page.getAlertBus() && page.getAlertBus().getId() !== this.getId()) {
        page.getAlertBus().addAlert(messageProps, setStateCallback);
      } else {
        this.setState(function (state) {
          var alertStack = state.alertStack.slice();
          if (alertProps.priority) {
            alertStack.splice(0, 0, messageProps);
          } else {
            alertStack.push(messageProps);
          }
          return { alertStack: alertStack };
        }, setStateCallback);
      }
    } else {
      this.showWarning('noMessage', alertProps.content, { alertProps: alertProps });
    }
    return this;
  },

  addAlertToPosition: function addAlertToPosition(alertIndex, alertProps, setStateCallback) {
    if (alertProps.content) {
      var messageProps = this._getMessageProps(null, alertProps);
      var page = this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE);
      if (!this.props.forceRender && page && page.getAlertBus() && page.getAlertBus().getId() !== this.getId()) {
        page.getAlertBus().addAlertToPosition(alertIndex, messageProps, setStateCallback);
      } else {
        this.setState(function (state) {
          var alertStack = state.alertStack.slice();
          alertStack.splice(alertIndex, 0, messageProps);
          return { alertStack: alertStack };
        }, setStateCallback);
      }
    } else {
      this.showWarning('noMessage', alertProps.content, { alertProps: alertProps });
    }

    return this;
  },

  setAlert: function setAlert(alertProps, setStateCallback) {
    if (alertProps.content) {
      var messageProps = this._getMessageProps(null, alertProps);
      var page = this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE);
      if (!this.props.forceRender && page && page.getAlertBus() && page.getAlertBus().getId() !== this.getId()) {
        page.getAlertBus().setAlert(messageProps, setStateCallback);
      } else {
        this.setState({ alertStack: [messageProps] }, setStateCallback);
      }
    } else {
      this.showWarning('noMessage', alertProps.content, { alertProps: alertProps });
    }
    return this;
  },

  setAlerts: function setAlerts(alertStack, setStateCallback) {
    var stateAlertStack = alertStack.slice();
    var page = this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE);
    if (!this.props.forceRender && page && page.getAlertBus() && page.getAlertBus().getId() !== this.getId()) {
      page.getAlertBus().setAlerts(stateAlertStack, setStateCallback);
    } else {
      this.setState({ alertStack: stateAlertStack }, setStateCallback);
    }
    return this;
  },

  removeAlert: function removeAlert(alertId, setStateCallback) {
    var page = this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE);
    if (!this.props.forceRender && page && page.getAlertBus() && page.getAlertBus().getId() !== this.getId()) {
      page.getAlertBus().removeAlert(alertId, setStateCallback);
    } else {
      this.setState(function (state) {
        var alertStack = state.alertStack.slice();
        var index = null;

        for (var i = 0; i < alertStack.length; i++) {
          if (alertStack[i].id === alertId) {
            index = i;
            break;
          }
        }

        index !== null && alertStack.splice(index, 1);
        return { alertStack: alertStack };
      }, setStateCallback);
    }

    return this;
  },

  clearAlerts: function clearAlerts(setStateCallback) {
    var _this = this;

    var page = this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE);
    if (!this.props.forceRender && page && page.getAlertBus() && page.getAlertBus().getId() !== this.getId()) {
      page.getAlertBus().clearAlerts(setStateCallback);
    } else {
      var props = this.getAlerts()[0];
      if (props) {
        props.hidden = true;

        this.setState({
          alertStack: [props]
        }, function () {
          setTimeout(function () {
            _this.setAsyncState({ alertStack: [] }, setStateCallback);
          }, _alert2.default.defaults.transitionDuration);
        });
      } else {
        typeof setStateCallback === 'function' && setStateCallback();
      }
    }
    return this;
  },

  getAlerts: function getAlerts() {
    var result = this.state.alertStack;

    var page = this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE);
    if (!this.props.forceRender && page && page.getAlertBus() && page.getAlertBus().getId() !== this.getId()) {
      result = page.getAlertBus().getAlerts();
    }

    return result;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMessageProps: function _getMessageProps(message, alertProps) {
    alertProps = alertProps || {};
    var alertBus = this;
    var messageId = alertProps.id || UU5.Common.Tools.generateUUID();

    return {
      id: messageId,
      hidden: false,

      content: alertProps.content,
      colorSchema: typeof alertProps.colorSchema === 'string' ? alertProps.colorSchema : this.props.colorSchema,
      closeTimer: typeof alertProps.closeTimer === 'number' ? alertProps.closeTimer : this.props.closeTimer,
      closeDisabled: alertProps.closeDisabled === undefined ? this.props.closeDisabled : alertProps.closeDisabled,
      onClose: function onClose(alert) {
        var onClose;
        if (typeof alertProps.onClose === 'function') {
          onClose = function onClose() {
            alertProps.onClose(alert);
          };
        } else if (typeof alertBus.props.onClose === 'function') {
          onClose = function onClose() {
            alertBus.props.onClose(alert);
          };
        }

        alertBus.removeAlert(messageId, onClose);
      }
    };
  },
  //@@viewOff:componentSpecificHelpers

  // Render
  _getProps: function _getProps() {
    var alertStack = this.getAlerts();

    var alertProps = this.getMainPropsToPass();
    var hidden = alertStack.length === 0;
    alertProps.hidden = hidden;

    if (!hidden) {
      alertProps = UU5.Common.Tools.mergeDeep(alertProps, alertStack[0]);
    }

    alertProps.id = this.getId() + '-alert';
    alertProps.position = this.props.position;
    alertProps.block = this.props.block;

    return alertProps;
  },

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? _react2.default.createElement(_alert2.default, this._getProps()) : null;
  }
  //@@viewOff:render
});

exports.default = AlertBus;
