// uu5g04/dist-node/bricks/card.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Card = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _section = require('./section.js');

var _section2 = _interopRequireDefault(_section);

require('./card.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Card = exports.Card = (0, _createReactClass2.default)({
  displayName: 'Card',

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.SectionMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Card"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("card"),
      inline: _bricksNs2.default.css("card-inline"),
      spaces: _bricksNs2.default.css("card-spaces")
    },
    opt: {
      dummyLevel: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    elevation: _propTypes2.default.oneOf(['0', '1', '2', '3', '4', '5', 0, 1, 2, 3, 4, 5]),
    elevationHover: _propTypes2.default.oneOf(['0', '1', '2', '3', '4', '5', 0, 1, 2, 3, 4, 5]),
    inline: _propTypes2.default.bool,
    width: _propTypes2.default.number,
    minWidth: _propTypes2.default.number,
    noSpaces: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      elevation: 1,
      elevationHover: 1,
      inline: false,
      width: null,
      minWidth: null,
      noSpaces: false
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainProps: function _getMainProps() {
    var mainAttrs = this.getMainPropsToPass();
    mainAttrs.className += ' uu5-elevation-' + this.props.elevation;
    mainAttrs.className += ' uu5-elevation-hover-' + this.props.elevationHover;
    !this.props.noSpaces && (mainAttrs.className += " " + this.getClassName("spaces"));

    if (this.props.width !== null || this.props.minWidth !== null) {
      mainAttrs.style = mainAttrs.style || {};
      this.props.width !== null && (mainAttrs.style.width = this.props.width);
      this.props.minWidth !== null && (mainAttrs.style.minWidth = this.props.minWidth);
      mainAttrs.className += ' ' + this.getClassName('inline');
    } else if (this.props.inline) {
      mainAttrs.className += ' ' + this.getClassName('inline');
    }
    return mainAttrs;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? _react2.default.createElement(
      _section2.default,
      this._getMainProps(),
      _react2.default.Children.toArray(this.props.children)
    ) : null;
  }
  //@@viewOff:render

});

exports.default = Card;
