// uu5g04/dist-node/bricks/code.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Code = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _factory = require('./factory.js');

var _factory2 = _interopRequireDefault(_factory);

require('./code.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Code_ = _factory2.default.createTag('code', '_code');

var Code = exports.Code = (0, _createReactClass2.default)({
  displayName: 'Code',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Code"),
    nestingLevel: 'inline',
    classNames: {
      main: _bricksNs2.default.css("code")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainProps: function _getMainProps() {
    var mainProps = this.getMainPropsToPass();

    mainProps.id = this.getId() + '-code';
    mainProps.ignoreInnerHTML = true;
    mainProps.language = 'code';

    return mainProps;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? _react2.default.createElement(
      Code_,
      this._getMainProps(),
      this.props.children && _react2.default.Children.toArray(this.props.children)
    ) : null;
  }
  //@@viewOff:render
});

exports.default = Code;
