// uu5g04/dist-node/bricks/cookies-info.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CookiesInfo = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _factory = require('./factory.js');

var _row = require('./row');

var _row2 = _interopRequireDefault(_row);

var _section = require('./section');

var _section2 = _interopRequireDefault(_section);

require('./cookies-info.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CookiesInfo = exports.CookiesInfo = (0, _createReactClass2.default)({
  displayName: 'CookiesInfo',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("CookiesInfo"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'smallBox'),
    classNames: {
      main: _bricksNs2.default.css("cookies-info")
    },
    opt: {
      nestingLevelRoot: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    data: _propTypes2.default.shape({
      header: _propTypes2.default.any,
      chapters: _propTypes2.default.arrayOf(_propTypes2.default.shape({
        header: _propTypes2.default.any,
        // array of contents
        rows: _propTypes2.default.arrayOf(_propTypes2.default.any)
      }))
    }),
    headerLevel: _propTypes2.default.oneOf(['0', '1', '2', '3', '4', '5', '6', 0, 1, 2, 3, 4, 5, 6]),
    chapterLevel: _propTypes2.default.oneOf(['0', '1', '2', '3', '4', '5', '6', 0, 1, 2, 3, 4, 5, 6])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      data: {},
      headerLevel: 1,
      chapterLevel: 2
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getChapters: function _getChapters(chaptersData) {
    var collection = this;
    return chaptersData && chaptersData.map(function (chapter, i) {
      var paragraphs = chapter.rows.map(function (content, j) {
        return _react2.default.createElement(_factory.P, { content: content, key: j });
      });

      return _react2.default.createElement(
        _row2.default,
        { header: chapter.header, key: i, level: collection.props.chapterLevel },
        paragraphs
      );
    });
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var data = this.props.data;
    var chapters = this._getChapters(data.chapters);

    return _react2.default.createElement(
      _section2.default,
      _extends({}, this.getMainPropsToPass(), { header: data.header, level: this.props.headerLevel, nestingLevel: this.getNestingLevel() }),
      chapters
    );
  }
  //@@viewOff:render
});

exports.default = CookiesInfo;
