// uu5g04/dist-node/bricks/cookie-bar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CookieBar = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _factory = require('./factory.js');

var _span = require('./span.js');

var _span2 = _interopRequireDefault(_span);

var _link = require('./link.js');

var _link2 = _interopRequireDefault(_link);

var _button = require('./button.js');

var _button2 = _interopRequireDefault(_button);

var _icon = require('./icon.js');

var _icon2 = _interopRequireDefault(_icon);

require('./cookie-bar.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CookieBar = exports.CookieBar = (0, _createReactClass2.default)({
  displayName: 'CookieBar',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.ColorSchemaMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("CookieBar"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("cookie-bar uu5-common-bg"),
      top: _bricksNs2.default.css("cookie-bar-top"),
      bottom: _bricksNs2.default.css("cookie-bar-bottom"),
      button: _bricksNs2.default.css("cookie-bar-button"),
      link: _bricksNs2.default.css("cookie-bar-link")
    },
    defaults: {
      content: 'Cookies help us to provide, protect and improve our services. By viewing this site, you agree to their use.',
      expireDays: 10 * 365.25
    },
    opt: {
      nestingLevelRoot: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    infoText: _propTypes2.default.any,
    infoHref: _propTypes2.default.string,
    infoTarget: _propTypes2.default.oneOf(['_blank', '_parent', '_top', '_self']),
    fixed: _propTypes2.default.oneOf(['top', 'bottom']),
    onClose: _propTypes2.default.func,
    expireDays: _propTypes2.default.number,
    cookieKey: _propTypes2.default.string,
    cookieValue: _propTypes2.default.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      infoText: null,
      infoHref: null,
      infoTarget: '_blank',
      fixed: 'bottom',
      onClose: null,
      expireDays: null,
      cookieKey: 'uu5-cookies',
      cookieValue: 'yes'
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function componentWillMount() {
    this.props.expireDays ? this._checkCookies() : this._checkLocalStorageItem();
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    this.props.expireDays ? this._checkCookies() : this._checkLocalStorageItem();
  },

  //@@viewOff:standardComponentLifeCycle

  // Interface

  // Overriding Functions

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _checkCookies: function _checkCookies() {
    if (UU5.Common.Tools.getCookie(this.props.cookieKey) === this.props.cookieValue && !this._isAgreed) {
      this._isAgreed = true;
      this.setState({ hidden: true }, this.props.onClose);
    }
    return this;
  },
  _checkLocalStorageItem: function _checkLocalStorageItem() {
    var item = localStorage.getItem(this.props.cookieKey);
    if (item === this.props.cookieValue) {
      this.setState({ hidden: true }, this.props.onClose);
    }
    return this;
  },
  _setLocalStorageItem: function _setLocalStorageItem() {
    localStorage.setItem(this.props.cookieKey, this.props.cookieValue);
    this.hide(this.props.onClose);
  },
  _confirm: function _confirm() {
    UU5.Common.Tools.setCookie(this.props.cookieKey, this.props.cookieValue, this.props.expireDays || this.getDefault().expireDays);
    this.hide(this.props.onClose);
  },
  _getText: function _getText() {
    var content = this.getContent();

    if (!content && !this.props.children) {
      content = this.getDefault().content;
    }

    return _react2.default.createElement(
      _span2.default,
      { content: content },
      this.props.children && _react2.default.Children.toArray(this.props.children)
    );
  },
  _getButton: function _getButton() {
    return _react2.default.createElement(
      _button2.default,
      { className: this.getClassName('button'), bgStyle: 'transparent', colorSchema: 'custom', size: 's', onClick: this.props.expireDays ? this._confirm : this._setLocalStorageItem },
      _react2.default.createElement(_icon2.default, { icon: 'mdi-close' })
    );
  },
  _getLink: function _getLink() {
    var link = null;

    if (this.props.infoText) {
      // href -> linkHref || undefined -> if undefined, default prop of link is set - other way, null is set
      link = _react2.default.createElement(_link2.default, { content: this.props.infoText, href: this.props.infoHref || undefined, className: this.getClassName().link, target: this.props.infoTarget });
    }

    return link;
  },
  _getMainProps: function _getMainProps() {
    var mainProps = this.getMainPropsToPass(["UU5.Common.BaseMixin", "UU5.Common.ElementaryMixin"]);
    this.props.fixed && (mainProps.className += ' ' + this.getClassName(this.props.fixed));

    mainProps.nestingLevel = this.getNestingLevel();

    return mainProps;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return _react2.default.createElement(
      _factory.Div,
      this._getMainProps(),
      this._getText(),
      this._getLink(),
      this._getButton()
    );
  }
  //@@viewOff:render

});

exports.default = CookieBar;
