// uu5g04/dist-node/bricks/data-table.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DataTable = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _table = require('./table.js');

var _table2 = _interopRequireDefault(_table);

var _tableThead = require('./table-thead.js');

var _tableThead2 = _interopRequireDefault(_tableThead);

var _tableTfoot = require('./table-tfoot.js');

var _tableTfoot2 = _interopRequireDefault(_tableTfoot);

var _tableTbody = require('./table-tbody.js');

var _tableTbody2 = _interopRequireDefault(_tableTbody);

var _tableTh = require('./table-th.js');

var _tableTh2 = _interopRequireDefault(_tableTh);

var _tableTd = require('./table-td.js');

var _tableTd2 = _interopRequireDefault(_tableTd);

var _tableTr = require('./table-tr.js');

var _tableTr2 = _interopRequireDefault(_tableTr);

require('./data-table.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DataTable = exports.DataTable = (0, _createReactClass2.default)({
  displayName: 'DataTable',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.LsiMixin, UU5.Common.SectionMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("DataTable"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("data-table")
    },
    lsi: function lsi() {
      return UU5.Environment.Lsi.Bricks.dataTable;
    },
    errors: {
      invalidDataHeader: 'Header row items count (%d) are not equal to data row (0) items length (%d).',
      invalidDataBody: 'Data row (%d) items count (%d) are not equal to data row (%d) items length (%d).',
      invalidDataFooter: 'Footer row items count (%d) are not equal to data row (%d) items length (%d).'
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    striped: _propTypes2.default.bool,
    bordered: _propTypes2.default.bool,
    hover: _propTypes2.default.bool,
    condensed: _propTypes2.default.bool,
    headerRow: _propTypes2.default.arrayOf(_propTypes2.default.any //content props
    ),
    footerRow: _propTypes2.default.arrayOf(_propTypes2.default.any //content props
    ),
    rows: _propTypes2.default.arrayOf(_propTypes2.default.arrayOf(_propTypes2.default.any //content props
    ))
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      striped: false,
      bordered: false,
      hover: false,
      condensed: false,
      headerRow: null,
      footerRow: null,
      rows: null
    };
  },

  getInitialState: function getInitialState() {
    return {
      valid: this._validateData(this.props)
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (this.props.headerRow !== nextProps.headerRow || this.props.rows !== nextProps.rows || this.props.footerRow !== nextProps.footerRow) {
      this.setState({ valid: this._validateData(nextProps) });
    }
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _validateData: function _validateData(props) {
    var dataTable = this;
    var valid = true;
    var referenceRow = props.headerRow;

    for (var i = 0; i < props.rows.length; i++) {
      var row = props.rows[i];

      if (referenceRow && valid) {
        valid = referenceRow.length === row.length;

        if (!valid) {
          if (i === 0) {
            dataTable.showError('invalidDataHeader', [referenceRow.length, row.length]);
          } else {
            dataTable.showError('invalidDataBody', [i - 1, referenceRow.length, i, row.length]);
          }
          break;
        }
      } else {
        referenceRow = row;
      }
    }

    if (valid && props.footerRow && referenceRow) {
      valid = props.footerRow.length === referenceRow.length;
      if (!valid) {
        dataTable.showError('invalidDataFooter', [props.footerRow.length, referenceRow.length - 1, referenceRow.length]);
      }
    }

    return valid;
  },

  _getHeaderRow: function _getHeaderRow() {
    return _react2.default.createElement(
      _tableThead2.default,
      null,
      _react2.default.createElement(_tableTr2.default, { content: this.props.headerRow && this._getRowCells(this.props.headerRow, true) })
    );
  },

  _getFooterRow: function _getFooterRow() {
    return _react2.default.createElement(
      _tableTfoot2.default,
      null,
      _react2.default.createElement(_tableTr2.default, { content: this.props.footerRow && this._getRowCells(this.props.footerRow) })
    );
  },

  _getBodyRows: function _getBodyRows() {
    var dataTable = this;
    var rows = this.props.rows.map(function (row, i) {
      return _react2.default.createElement(
        _tableTr2.default,
        { key: i },
        dataTable._getRowCells(row)
      );
    });

    return _react2.default.createElement(
      _tableTbody2.default,
      null,
      rows
    );
  },

  _getRowCells: function _getRowCells(row, isHeader) {
    var brick = isHeader ? _tableTh2.default : _tableTd2.default;
    var dataTable = this;
    return row.map(function (cell, i) {
      var props;
      if (dataTable._isProps(cell)) {
        props = UU5.Common.Tools.merge({}, cell, { key: i });
      } else {
        props = { content: dataTable._checkLsiContent(cell), key: i };
      }
      return _react2.default.createElement(brick, props);
    });
  },

  _isProps: function _isProps(cell) {
    return !(typeof cell === 'string' || typeof cell === 'number' || // number or string
    (typeof cell === 'undefined' ? 'undefined' : _typeof(cell)) === 'object' && (cell.lsi || cell === null) || // empty string or lsi content
    (typeof cell === 'undefined' ? 'undefined' : _typeof(cell)) === 'object' && typeof cell.type === 'function' // react.element
    );
  },

  _checkLsiContent: function _checkLsiContent(content) {
    return content.lsi ? this.getLsiItem(content.lsi) : content;
  },

  _getMainProps: function _getMainProps() {
    var mainProps = this.getMainPropsToPass(["UU5.Common.BaseMixin", "UU5.Common.ElementaryMixin", "UU5.Common.SectionMixin"]);
    mainProps.striped = this.props.striped;
    mainProps.bordered = this.props.bordered;
    mainProps.hover = this.props.hover;
    mainProps.condensed = this.props.condensed;

    return mainProps;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var result;

    if (this.state.valid) {
      result = _react2.default.createElement(
        _table2.default,
        _extends({}, this._getMainProps(), { header: this.props.header, footer: this.props.footer, nestingLevel: this.getNestingLevel() }),
        this._getHeaderRow(),
        this._getBodyRows(),
        this._getFooterRow()
      );
    } else {
      result = _react2.default.createElement(UU5.Common.Error, { content: this.getLsiComponent('invalidDataLabel') });
    }

    return result;
  }
  //@@viewOff:render
});

exports.default = DataTable;
