// uu5g04/dist-node/bricks/dropdown.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Dropdown = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _button = require('./button.js');

var _button2 = _interopRequireDefault(_button);

var _icon = require('./icon.js');

var _icon2 = _interopRequireDefault(_icon);

var _dropdownItem = require('./dropdown-item.js');

var _dropdownItem2 = _interopRequireDefault(_dropdownItem);

require('./dropdown.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Dropdown = exports.Dropdown = (0, _createReactClass2.default)({
  displayName: 'Dropdown',

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ColorSchemaMixin, UU5.Common.ContentMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Dropdown"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("dropdown"),
      dropdown: _bricksNs2.default.css("dropdown-dropdown"),
      dropup: _bricksNs2.default.css("dropdown-dropup"),
      split: _bricksNs2.default.css("dropdown-group"),
      buttonGroup: _bricksNs2.default.css("button-group", "button-group-horizontal"),
      pullRight: _bricksNs2.default.css("dropdown-right"),
      pullRightMenu: _bricksNs2.default.css("dropdown-menu-right"),
      dropdownBtn: _bricksNs2.default.css("dropdown-button"),
      buttonCover: _bricksNs2.default.css("dropdown-button-cover"),
      menu: _bricksNs2.default.css("dropdown-menu"),
      menuList: _bricksNs2.default.css("dropdown-menu-list"),
      menuWrapper: _bricksNs2.default.css("dropdown-menu-wrapper"),
      open: _bricksNs2.default.css("dropdown-open"),
      autoDropup: _bricksNs2.default.css("dropdown-dropup"),
      autoDropdown: _bricksNs2.default.css("dropdown-dropdown"),
      block: _bricksNs2.default.css("dropdown-block")
    },
    defaults: {
      childTagName: 'UU5.Bricks.Dropdown.Item'
    },
    opt: {
      nestingLevelRoot: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    label: _propTypes2.default.any, // content
    size: _propTypes2.default.string,
    onClick: _propTypes2.default.func,
    bgStyle: _propTypes2.default.string,

    // icon props
    iconOpen: _propTypes2.default.string,
    iconClosed: _propTypes2.default.string,
    iconHidden: _propTypes2.default.bool,

    // dropdown props
    items: _propTypes2.default.arrayOf(_propTypes2.default.object // UU5.Bricks.Dropdown.item props
    ),
    pullRight: _propTypes2.default.bool,
    dropup: _propTypes2.default.bool,
    split: _propTypes2.default.bool,

    // link child props
    smoothScroll: _propTypes2.default.number,
    offset: _propTypes2.default.number,

    closedOnLeave: _propTypes2.default.bool,
    allowTags: _propTypes2.default.arrayOf(_propTypes2.default.string),
    disableBackdrop: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      label: 'Dropdown',
      size: 'm',
      onClick: null,
      bgStyle: undefined,
      iconOpen: 'mdi-menu-down',
      iconClosed: 'mdi-menu-down',
      iconHidden: false,
      pullRight: false,
      dropup: false,
      split: false,
      smoothScroll: null,
      offset: null,
      closedOnLeave: false,
      allowTags: [],
      disableBackdrop: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      open: false,
      dropup: this.props.dropup,
      block: false,
      left: 0
    };
  },

  componentWillUnmount: function componentWillUnmount() {
    this._removeEvent();
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isDropdown: function isDropdown() {
    return true;
  },


  open: function open(setStateCallback) {
    this._stopPropagation = true;
    this._addEvent();
    this.setState({ open: true }, setStateCallback);
    return this;
  },

  close: function close(setStateCallback) {
    this._removeEvent();
    this.setState({ open: false }, setStateCallback);
    return this;
  },

  toggle: function toggle(setStateCallback) {
    this.setState(function (state) {
      this._stopPropagation = !state.open;
      state.open ? this._removeEvent() : this._addEvent();
      return { open: !state.open };
    }, setStateCallback);
    return this;
  },

  isOpen: function isOpen() {
    return this.state.open;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  shouldChildRender_: function shouldChildRender_(child) {
    var childTagName = UU5.Common.Tools.getChildTagName(child);
    var childTagNames = this.props.allowTags.concat(this.getDefault().childTagName);
    return childTagNames.indexOf(childTagName) > -1;
  },

  expandChildProps_: function expandChildProps_(child) {
    var newChildProps = UU5.Common.Tools.mergeDeep({}, child.props);

    newChildProps.smoothScroll = newChildProps.smoothScroll || this.props.smoothScroll;
    newChildProps.offset = newChildProps.offset || this.props.offset;
    newChildProps.dropup = newChildProps.dropup || this.state.dropup;

    return newChildProps;
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _findTarget: function _findTarget(item) {
    var result = false;
    var id = this.getId();

    if (item.id === id) {
      result = true;
    } else if (item.parentElement) {
      result = this._findTarget(item.parentElement);
    }

    return result;
  },
  _addEvent: function _addEvent() {
    var _this = this;

    !this.props.disableBackdrop && UU5.Environment.EventListener.addWindowEvent('click', this.getId(), function (e) {
      var isDropdown = _this._findTarget(e.target);

      if (!_this._stopPropagation && !isDropdown && _this.isOpen()) {
        _this.close();
      } else {
        _this._stopPropagation = false;
      }
    });
    return this;
  },
  _removeEvent: function _removeEvent() {
    !this.props.disableBackdrop && UU5.Environment.EventListener.removeWindowEvent('click', this.getId());
    return this;
  },


  _toggle: function _toggle(newState, setStateCallback) {
    var _this2 = this;

    this.setState(function (state) {
      state.open ? _this2._removeEvent() : _this2._addEvent();
      newState.open = !state.open;
      return newState;
    }, setStateCallback);
    return this;
  },

  _getIcon: function _getIcon() {
    var icon = null;
    if (!this.props.iconHidden) {
      var iconName = this.isOpen() ? this.props.iconOpen : this.props.iconClosed;
      icon = _react2.default.createElement(_icon2.default, { icon: iconName });
    }
    return icon;
  },

  _getChildren: function _getChildren() {
    var contentProps = {};

    if (this.props.items) {
      contentProps = {
        content: {
          tag: this.getDefault().childTagName,
          propsArray: this.props.items
        }
      };
    } else if (this.props.children) {
      contentProps = { children: this.props.children };
    }
    contentProps.nestingLevel = this.getNestingLevel();

    return this.buildChildren(contentProps);
  },

  _getButton: function _getButton() {
    var dropdown = this;
    var onClick = null;
    if (this.props.split) {
      if (typeof this.props.onClick === 'function') {
        onClick = function onClick(button, event) {
          dropdown.props.onClick(dropdown, event);
        };
      }
    } else {
      onClick = this._onClickHandler;
    }

    var buttonProps = {
      colorSchema: this.getColorSchema(),
      size: this.props.size,
      disabled: this.isDisabled(),
      onClick: onClick,
      bgStyle: this.props.bgStyle
    };

    var icon = null;
    if (!this.props.split) {
      buttonProps.className = this.getClassName().dropdownBtn;
      icon = this._getIcon();
    }
    /*
     return (
     <Button {...buttonProps}>
     <Span content={this.props.label} /> {icon}
     </Button>
     );*/
    var content = void 0;
    if (Array.isArray(this.props.label)) {
      content = this.props.label.concat([icon]);
    } else {
      content = [this.props.label, icon];
    }
    return _react2.default.createElement(_button2.default, _extends({}, buttonProps, { content: content }));
  },

  _getIconButton: function _getIconButton() {
    var iconButton = null;

    if (this.props.split) {
      iconButton = _react2.default.createElement(
        _button2.default,
        { colorSchema: this.getColorSchema(), size: this.props.size, disabled: this.isDisabled(), className: this.getClassName().dropdownBtn, onClick: this._onClickHandler },
        this._getIcon()
      );
    }

    return iconButton;
  },

  _onClickHandler: function _onClickHandler() {
    this._stopPropagation = true;

    var dropdown = _reactDom2.default.findDOMNode(this);
    var dropdownMenu = _reactDom2.default.findDOMNode(this._dropdownMenu);
    var dropdownRect = dropdown.getBoundingClientRect();
    var dropdownMenuRect = dropdownMenu.getBoundingClientRect();

    var parent = UU5.Common.Tools.getElementByComputedStyle(document.getElementById(this.getId()), 'overflow', 'hidden');
    var documentClientHeight = document.getElementsByTagName('html')[0].clientHeight;
    var parentHeight = parent ? UU5.Common.Tools.getInnerHeight(parent) : documentClientHeight;

    var documentClientWidth = document.getElementsByTagName('html')[0].clientWidth;
    var parentWidth = parent ? UU5.Common.Tools.getInnerWidth(parent) : documentClientWidth;

    var width = parentWidth < documentClientWidth ? parentWidth : documentClientWidth;
    var left = 0;
    if (width - dropdownMenuRect.left < dropdownMenuRect.width) {
      left = dropdownRect.width - dropdownMenuRect.width;
      if (dropdownMenuRect.left < left * -1) {
        left = dropdownMenuRect.left;
      }
    }
    if (this.props.pullRight) {
      left = null;
      if (dropdown.clientWidth + dropdown.offsetLeft < dropdownMenu.clientWidth) {
        left = 0;
      }
      if (width < dropdownMenu.clientWidth) {
        left = null;
      }
    }

    var fitParent = dropdownRect.height + dropdownMenuRect.height < parentHeight;
    var fitDown = dropdownRect.height + dropdownMenuRect.height + dropdown.offsetTop - window.pageYOffset < (parentHeight < documentClientHeight ? parentHeight : documentClientHeight);
    var fitUp = dropdown.offsetTop - window.pageYOffset > dropdownMenuRect.height;

    if (!fitParent) {
      this._toggle({ dropup: false, block: true, left: left, marginBottom: null });
    } else if (!fitDown && fitUp || this.state.dropup) {
      this._toggle({ dropup: true, block: false, left: left, marginBottom: dropdownRect.height });
    } else if (!this.props.dropup && fitDown) {
      this._toggle({ dropup: false, block: false, left: left, marginBottom: null });
    } else {
      this.toggle();
    }
    return this;
  },

  _getMainAttrs: function _getMainAttrs() {
    var _this3 = this;

    var mainAttrs = this.getMainAttrs();

    mainAttrs.id = this.getId();

    if (this.isOpen()) {
      mainAttrs.className += ' ' + this.getClassName().open;
    }

    this.props.pullRight && (mainAttrs.className += ' ' + this.getClassName().pullRight);

    if (this.props.split) {
      mainAttrs.className += ' ' + this.getClassName().split;
    } else if (this.props.dropup) {
      mainAttrs.className += ' ' + this.getClassName().dropup;
    } else {
      mainAttrs.className += ' ' + this.getClassName().dropdown;
    }

    if (this.state.dropup) {
      mainAttrs.className += ' ' + this.getClassName().autoDropup;
    } else {
      mainAttrs.className += ' ' + this.getClassName().autoDropdown;
    }

    mainAttrs.onMouseLeave = this.props.closedOnLeave ? function () {
      _this3.close();
    } : null;

    return mainAttrs;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var _this4 = this;

    var menuClass = this.getClassName().menu;
    this.props.pullRight && (menuClass += ' ' + this.getClassName().pullRightMenu);
    this.state.block && (menuClass += ' ' + this.getClassName('block'));

    var style = {
      left: this.state.left + 'px',
      right: 'auto',
      marginBottom: this.state.marginBottom || 0
    };
    if (this.props.pullRight) {
      if (!this.state.open || this.state.left === null) {
        style = {
          right: 0,
          left: 'auto',
          marginBottom: this.state.marginBottom || 0
        };
      }
    }

    var buttonCoverClassName = this.getClassName().buttonCover;
    if (this.props.split) {
      buttonCoverClassName += " " + this.getClassName("buttonGroup");
    }

    return this.getNestingLevel() ? _react2.default.createElement(
      'div',
      this._getMainAttrs(),
      _react2.default.createElement(
        'div',
        { className: buttonCoverClassName, id: this.getId() + '-cover' },
        this._getButton(),
        this._getIconButton()
      ),
      _react2.default.createElement(
        'div',
        { className: this.getClassName('menuWrapper') },
        _react2.default.createElement(
          'div',
          { className: menuClass, ref: function ref(dropdownMenu) {
              return _this4._dropdownMenu = dropdownMenu;
            }, style: style },
          _react2.default.createElement(
            'ul',
            { className: this.getClassName('menuList') },
            this._getChildren()
          )
        )
      ),
      this.getDisabledCover()
    ) : null;
  }
  //@@viewOff:render
});

Dropdown.Item = _dropdownItem2.default;
exports.default = Dropdown;
