// uu5g04/dist-node/bricks/factory.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinkMaterialDesign = exports.LinkMongoDB = exports.LinkMSAzure = exports.LinkDocker = exports.LinkPuma = exports.LinkRuby = exports.LinkReact = exports.LinkJQuery = exports.LinkJavaScript = exports.LinkCSS = exports.LinkHTML5 = exports.LinkW3Schools = exports.LinkBootstrap = exports.LinkPlus4U = exports.LinkUnicornApproach = exports.LinkUUP = exports.LinkUuP = exports.LinkUuKnowledgeBase = exports.LinkUuBmlDraw = exports.LinkUuDocKit = exports.LinkUuCloud = exports.LinkUuOIDC = exports.LinkUuAppServerJava = exports.LinkUuAppServer = exports.LinkUu5CodeKit = exports.LinkUu5ComponentRegistry = exports.LinkUU5 = exports.LinkUuApp = exports.LinkUAF = exports.LinkUnicornCollege = exports.LinkUnicornUniverse = exports.LinkUnicornSystems = exports.LinkUnicorn = exports.Samp = exports.Pre = exports.Kbd = exports.Var = exports.Dt = exports.Dd = exports.Address = exports.Abbr = exports.Em = exports.Strong = exports.U = exports.Ins = exports.S = exports.Del = exports.Mark = exports.Lead = exports.Small = exports.Paragraph = exports.P = exports.Div = exports.Factory = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _link = require('./link');

var _link2 = _interopRequireDefault(_link);

require('./factory.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Factory = exports.Factory = {

  createTag: function createTag(tag, uu5Name, className, defaultContent, levelFrom, levelTo) {
    return (0, _createReactClass2.default)({
      displayName: uu5Name || tag,
      //@@viewOn:mixins
      mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
      //@@viewOff:mixins

      //@@viewOn:statics
      statics: {
        tagName: _bricksNs2.default.name(uu5Name || tag),
        nestingLevelList: UU5.Environment.getNestingLevelList(levelFrom, levelTo),
        classNames: {
          main: _bricksNs2.default.css((uu5Name || tag).toLowerCase()) + (className ? ' ' + className : '')
        },
        defaults: {
          content: defaultContent || null
        }
      },
      //@@viewOff:statics

      //@@viewOn:propTypes
      //@@viewOff:propTypes

      //@@viewOn:getDefaultProps
      //@@viewOff:getDefaultProps

      //@@viewOn:standardComponentLifeCycle
      //@@viewOff:standardComponentLifeCycle

      //@@viewOn:interface
      //@@viewOff:interface

      //@@viewOn:overridingMethods
      //@@viewOff:overridingMethods

      //@@viewOn:componentSpecificHelpers
      //@@viewOff:componentSpecificHelpers

      //@@viewOn:render
      render: function render() {
        return this.getNestingLevel() ? _react2.default.createElement(tag.toLowerCase(), this.getMainAttrs(), this.getChildren() || this.getDefault().content, this.getDisabledCover()) : null;
      }
      //@@viewOff:render
    });
  },
  createLink: function createLink(uu5Name, href, defaultContent, target) {
    return (0, _createReactClass2.default)({

      //@@viewOn:mixins
      mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.NestingLevelMixin, UU5.Common.PureRenderMixin],
      //@@viewOff:mixins

      //@@viewOn:statics
      statics: {
        tagName: 'UU5.Bricks.Link' + uu5Name,
        nestingLevel: 'inline',
        classNames: {
          main: _bricksNs2.default.css("link") + uu5Name.toLowerCase()
        },
        defaults: {
          content: defaultContent || uu5Name
        }
      },
      //@@viewOn:propTypes
      //@@viewOff:propTypes

      //@@viewOn:getDefaultProps
      //@@viewOff:getDefaultProps

      //@@viewOn:standardComponentLifeCycle
      //@@viewOff:standardComponentLifeCycle

      //@@viewOn:interface
      //@@viewOff:interface

      //@@viewOn:overridingMethods
      //@@viewOff:overridingMethods

      //@@viewOn:componentSpecificHelpers
      //@@viewOff:componentSpecificHelpers

      //@@viewOn:render
      render: function render() {
        var mainProps = this.getMainPropsToPass();
        mainProps.href = href;
        mainProps.target = target || '_blank';

        return this.getNestingLevel() ? _react2.default.createElement(
          _link2.default,
          mainProps,
          this.props.children || this.getDefault().content
        ) : null;
      }
      //@@viewOn:render
    });
  }
};

var Div = exports.Div = UU5.Common.Div;

var P = exports.P = Factory.createTag('P', null, null, null, 'smallBox', 'smallBox'); //smallBox,smallBox

var Paragraph = exports.Paragraph = Factory.createTag('p', 'Paragraph', null, 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos hymenaeos. Donec ipsum massa, ullamcorper in, auctor et, scelerisque sed, est. Vestibulum fermentum tortor id mi. Etiam commodo dui eget wisi. Integer malesuada. Fusce consectetuer risus a nunc. Nullam eget nisl. In sem justo, commodo ut, suscipit at, pharetra vitae, orci. Aenean placerat. Etiam neque. Fusce suscipit libero eget elit.', 'smallBox', 'smallBox');

var Small = exports.Small = Factory.createTag('Small', null, null, null, 'inline'); //inline
var Lead = exports.Lead = Factory.createTag('span', 'Lead', 'lead', 'inline'); //inline
var Mark = exports.Mark = Factory.createTag('Mark', null, null, null, 'inline'); //inline
var Del = exports.Del = Factory.createTag('Del', null, null, null, 'inline'); //inline
var S = exports.S = Factory.createTag('S', null, null, null, 'inline'); //inline
var Ins = exports.Ins = Factory.createTag('Ins', null, null, null, 'inline'); //inline
var U = exports.U = Factory.createTag('U', null, null, null, 'inline'); //inline
var Strong = exports.Strong = Factory.createTag('Strong', null, null, null, 'inline'); //inline
var Em = exports.Em = Factory.createTag('Em', null, null, null, 'inline'); //inline
var Abbr = exports.Abbr = Factory.createTag('Abbr', null, null, null, 'inline'); //inline
var Address = exports.Address = Factory.createTag('Address', null, null, null, 'inline'); //inline
var Dd = exports.Dd = Factory.createTag('Dd', null, null, null, 'smallBox', 'smallBox'); //smallBox,smallBox
var Dt = exports.Dt = Factory.createTag('Dt', null, null, null, 'smallBox', 'smallBox'); //smallBox,smallBox

// code
var Var = exports.Var = Factory.createTag('Var', null, null, null, 'inline'); //inline
var Kbd = exports.Kbd = Factory.createTag('Kbd', null, null, null, 'inline'); //inline
var Pre = exports.Pre = Factory.createTag('Pre', null, _bricksNs2.default.css("pre"), null, 'boxCollection', 'inline'); //boxCollection,inline
var Samp = exports.Samp = Factory.createTag('Samp', null, null, null, 'inline'); //inline

var LinkUnicorn = exports.LinkUnicorn = Factory.createLink("Unicorn", "http://unicorn.eu/", "Unicorn");
var LinkUnicornSystems = exports.LinkUnicornSystems = Factory.createLink("UnicornSystems", "http://unicornsystems.eu/", 'Unicorn\xA0Systems');
var LinkUnicornUniverse = exports.LinkUnicornUniverse = Factory.createLink("UnicornUniverse", "https://unicornuniverse.eu/", 'Unicorn\xA0Universe');
var LinkUnicornCollege = exports.LinkUnicornCollege = Factory.createLink("UnicornCollege", "https://unicorncollege.cz/", 'Unicorn\xA0College');

var docKitUrl = 'https://uuos9.plus4u.net/uu-dockitg01-main/';
var LinkUAF = exports.LinkUAF = Factory.createLink("UAF", docKitUrl + "78462435-86b4d5a4a030400784764ebdb972bbda/book", "UAF");
var LinkUuApp = exports.LinkUuApp = Factory.createLink("UuApp", docKitUrl + "78462435-25d3b166760a44b7be70e5c2eb2abaaa/book", "uuApp");
var LinkUU5 = exports.LinkUU5 = Factory.createLink("UU5", docKitUrl + "78462435-ed11ec379073476db0aa295ad6c00178/book", "uu5");
var LinkUu5ComponentRegistry = exports.LinkUu5ComponentRegistry = Factory.createLink("Uu5ComponentRegistry", docKitUrl + "78462435-215c1674232d4634b367f20ad94349c0/book", "uu5ComponentRegistry");
var LinkUu5CodeKit = exports.LinkUu5CodeKit = Factory.createLink("Uu5CodeKit", docKitUrl + "78462435-f2142743693e4b22b1753c9fb761e945/book", "uu5CodeKit");
var LinkUuAppServer = exports.LinkUuAppServer = Factory.createLink("UuAppServer", docKitUrl + "78462435-34df77ebe0a04adda6dcd62d32c4f513/book", "uuAppServer");
var LinkUuAppServerJava = exports.LinkUuAppServerJava = Factory.createLink("UuAppServerJava", docKitUrl + "78462435-99c939a08e0849c68df5ee339c94054b/book", "uuAppServer-Java");
var LinkUuOIDC = exports.LinkUuOIDC = Factory.createLink("UuOIDC", docKitUrl + "78462435-d684156f06004f2781c88777e74834ef/book", "uuOIDC");
var LinkUuCloud = exports.LinkUuCloud = Factory.createLink("UuCloud", docKitUrl + "78462435-289fcd2e11d34f3e9b2184bedb236ded/book", "uuCloud");
var LinkUuDocKit = exports.LinkUuDocKit = Factory.createLink("UuDocKit", docKitUrl + "78462435-e3f5c648e85f4319bd8fc25ea5be6c2c/book", "uuDocKit");
var LinkUuBmlDraw = exports.LinkUuBmlDraw = Factory.createLink("UuBmlDraw", docKitUrl + "78462435-6f64aebc07184e9088d6e0542a8f9682/book", "uuBmlDraw");
var LinkUuKnowledgeBase = exports.LinkUuKnowledgeBase = Factory.createLink("UuKnowledgeBase", docKitUrl + "78462435-58d23c00b3b64ea99b2f7df3274e08ff/book", "uuKnowledgeBase ");
var LinkUuP = exports.LinkUuP = Factory.createLink("LinkUuP", docKitUrl + "78462435-c86acb9189cb421892546005a1099ea7/book", "uuP");
var LinkUUP = exports.LinkUUP = LinkUuP;
var LinkUnicornApproach = exports.LinkUnicornApproach = Factory.createLink("LinkUnicornApproach", docKitUrl + "78462435-c86acb9189cb421892546005a1099ea7/book/page?code=unicornApproach", "Unicorn Approach");

// export const LinkUUIoT = Factory.createLink("UUIoT", docKitUrl + "84723967990075193-18171658587168657/book", "uuIoT");
// export const LinkUUBT = Factory.createLink("UUBT", docKitUrl + "84723967990075193-38780273195993502/book", "uuBT");
// export const LinkUUMT = Factory.createLink("UUMT", docKitUrl + "84723967990075193-12401017776126880/book", "uuMT");

var LinkPlus4U = exports.LinkPlus4U = Factory.createLink("Plus4U", "https://plus4u.net/", "Plus4U");
var LinkBootstrap = exports.LinkBootstrap = Factory.createLink("Bootstrap", "http://getbootstrap.com/", "Bootstrap");
var LinkW3Schools = exports.LinkW3Schools = Factory.createLink("w3schools", "http://www.w3schools.com/", "w3schools");
var LinkHTML5 = exports.LinkHTML5 = Factory.createLink("HTML5", "http://www.w3schools.com/html/default.asp", "Html5");
var LinkCSS = exports.LinkCSS = Factory.createLink("CSS", "http://www.w3schools.com/css/default.asp", "CSS");
var LinkJavaScript = exports.LinkJavaScript = Factory.createLink("JavaScript", "http://www.w3schools.com/js/default.asp", "JavaScript");
var LinkJQuery = exports.LinkJQuery = Factory.createLink("JQuery", "https://jquery.com/", "JQuery");
var LinkReact = exports.LinkReact = Factory.createLink("React", "https://facebook.github.io/react/", "React");
var LinkRuby = exports.LinkRuby = Factory.createLink("Ruby", "https://www.ruby-lang.org/", "Ruby");
var LinkPuma = exports.LinkPuma = Factory.createLink("Puma", "http://puma.io/", "Puma");
var LinkDocker = exports.LinkDocker = Factory.createLink("Docker", "https://www.docker.com/", "Docker");
var LinkMSAzure = exports.LinkMSAzure = Factory.createLink("MSAzure", "https://azure.microsoft.com", 'Microsoft\xA0Azure');
var LinkMongoDB = exports.LinkMongoDB = Factory.createLink("MongoDB", "https://www.mongodb.com/", "MongoDB");
var LinkMaterialDesign = exports.LinkMaterialDesign = Factory.createLink("MaterialDesign", "https://material.io", 'Material\xA0Design');

exports.default = Factory;
