// uu5g04/dist-node/bricks/header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Header = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _jumbotron = require('./jumbotron.js');

var _jumbotron2 = _interopRequireDefault(_jumbotron);

require('./header.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Header = exports.Header = (0, _createReactClass2.default)({
  displayName: 'Header',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.ContentMixin, UU5.Common.ColorSchemaMixin, UU5.Common.LevelMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Header"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("header"),
      text: 'uu5-common-text',
      underline: _bricksNs2.default.css("header-underline"),
      jumbotronContent: _bricksNs2.default.css("header-jumbotron-content")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    underline: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      underline: undefined // default: false
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var level = this.getLevel();
    var result = void 0;

    if (level > 0) {
      var attrs = this.getMainAttrs();
      attrs.className += ' ' + this.getClassName("text");
      this.props.underline && (attrs.className += ' ' + this.getClassName("underline"));

      result = _react2.default.createElement('h' + level, attrs, this.getChildren(), this.getDisabledCover());
    } else {
      var className = this.getClassName().main;
      this.props.className && (className += ' ' + this.props.className);

      result = _react2.default.createElement(
        _jumbotron2.default,
        { disabled: this.isDisabled(), hidden: this.isHidden(), selected: this.isSelected(), colorSchema: this.props.colorSchema, className: className },
        _react2.default.createElement(
          'h1',
          { className: this.getClassName('jumbotronContent') },
          this.getChildren()
        )
      );
    }

    return this.getNestingLevel() ? result : null;
  }
  //@@viewOff:render

});

exports.default = Header;
