// uu5g04/dist-node/bricks/link.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.A = exports.Link = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

require('./link.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Link = exports.Link = (0, _createReactClass2.default)({
  displayName: 'Link',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.ContentMixin, UU5.Common.ColorSchemaMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Link"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'inline'),
    classNames: {
      main: _bricksNs2.default.css("link")
    },
    defaults: {
      content: 'noText',
      regexpHash: /^#/
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    href: _propTypes2.default.string,
    onClick: _propTypes2.default.func,
    onCtrlClick: _propTypes2.default.func,
    onWheelClick: _propTypes2.default.func,
    smoothScroll: _propTypes2.default.number,
    offset: _propTypes2.default.number,
    target: _propTypes2.default.oneOf(['_blank', '_parent', '_top', '_self']),
    download: _propTypes2.default.bool
    // disabledRouter: PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      href: null,
      onClick: null,
      onCtrlClick: null,
      onWheelClick: null,
      smoothScroll: 0,
      offset: 0,
      target: '_self',
      download: false
      // disabledRouter: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _isSmoothScroll: function _isSmoothScroll() {
    return this.props.href && this.props.href.length > 1 && this.props.href.lastIndexOf('#', 0) === 0 && (this.props.smoothScroll > 0 || this.props.offset > 0);
  },
  _onClickHandler: function _onClickHandler(e) {
    typeof this.props.onClick === 'function' && this.props.onClick(this, e);
    return this;
  },
  _onClickSmoothScroll: function _onClickSmoothScroll(e) {
    e.preventDefault();
    UU5.Common.Tools.scrollToTarget(this.props.href, this.props.smoothScroll, this.props.offset);
    this._onClickHandler(e);
    return this;
  },


  // _onClickRouter: function (e) {
  //   e.preventDefault();
  //   this.props.router.push(this.props.href);
  //   this._onClickHandler(e);
  //   return this;
  // },

  _containsHash: function _containsHash(url) {
    return this.getDefault('regexpHash').test(url);
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var _this = this;

    var mainAttrs = this.getMainAttrs();

    if (!this.isDisabled()) {
      mainAttrs.href = this._containsHash(this.props.href) ? location.href.split('#')[0] + this.props.href : this.props.href;

      if (typeof this.props.onClick === 'function' || typeof this.props.onWheelClick === 'function' || typeof this.props.onCtrlClick === 'function' || this._isSmoothScroll()) {
        mainAttrs.onClick = function (e) {
          _this.props.href && e.preventDefault();

          if (e.ctrlKey && typeof _this.props.onCtrlClick === 'function') {
            _this.props.onCtrlClick(_this, e);
          } else {
            if (_this._isSmoothScroll()) {
              _this._onClickSmoothScroll(e);
            } else if (typeof _this.props.onClick === 'function') {
              _this._onClickHandler(e);
            }
          }
        };
        if (typeof this.props.onWheelClick === 'function') {
          var onMouseDown = mainAttrs.onMouseDown;
          mainAttrs.onMouseDown = function (e) {
            e.button === 1 && e.preventDefault();
            typeof onMouseDown === 'function' && onMouseDown(e, _this);
          };

          var onMouseUp = mainAttrs.onMouseUp;
          mainAttrs.onMouseUp = function (e) {
            if (e.button === 1) {
              e.preventDefault();
              _this.props.onWheelClick(_this, e);
            }
            typeof onMouseUp === 'function' && onMouseUp(e, _this);
          };
        }
      } else {
        mainAttrs.target = this.props.target;
        mainAttrs.download = this.props.download;
      }
    }

    return this.getNestingLevel() ? _react2.default.createElement(
      'a',
      mainAttrs,
      this.getChildren() || this.props.href || this.getDefault().content
    ) : null;
  }
  //@@viewOff:render

});

var A = exports.A = Link;

exports.default = Link;
