// uu5g04/dist-node/bricks/lsi.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Lsi = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _lsiItem = require('./lsi-item.js');

var _lsiItem2 = _interopRequireDefault(_lsiItem);

require('./lsi.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Lsi = exports.Lsi = (0, _createReactClass2.default)({
  displayName: 'Lsi',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.LsiMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Lsi"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'inline'),
    classNames: {
      main: _bricksNs2.default.css("lsi")
    },
    defaults: {
      childTagName: 'UU5.Bricks.Lsi.Item'
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    lsi: _propTypes2.default.object,
    allowTags: _propTypes2.default.arrayOf(_propTypes2.default.string),
    params: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.array])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      lsi: null,
      allowTags: [],
      params: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isLsi: function isLsi() {
    return true;
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  shouldChildRender_: function shouldChildRender_(child) {
    var childTagName = UU5.Common.Tools.getChildTagName(child);
    var childTagNames = this.props.allowTags.concat(this.getDefault().childTagName);
    return childTagNames.indexOf(childTagName) > -1;
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _expandChildProps: function _expandChildProps(child) {
    var newChildProps = child ? UU5.Common.Tools.mergeDeep({}, child.props) : {};

    newChildProps.id = this.getId() + '-inner';
    newChildProps.controlled = true;
    newChildProps.hidden = this.isHidden();
    newChildProps.disabled = this.isDisabled();
    newChildProps.selected = this.isSelected();
    newChildProps.nestingLevel = this.getNestingLevel();

    return newChildProps;
  },

  _getChildren: function _getChildren() {
    var child = void 0;
    var children = this.getChildren();

    if (children) {
      var lsi = {};
      children.forEach(function (child) {
        if (child.props.language) {
          lsi[child.props.language] = child;
        }
      });
      child = this.getLsiItem(lsi, this.props.params);
      child = child ? _react2.default.cloneElement(child, this._expandChildProps(child)) : null;
    } else {
      child = _react2.default.createElement(_lsiItem2.default, _extends({}, this.getMainPropsToPass(["UU5.Common.BaseMixin", "UU5.Common.ElementaryMixin"]), { controlled: true, language: this.getLanguage(), content: this.getLsiItem(this.props.lsi, this.props.params), nestingLevel: this.getNestingLevel() }));
    }

    return child;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? this._getChildren() : null;
  }
  //@@viewOff:render
});

Lsi.Item = _lsiItem2.default;

exports.default = Lsi;
