// uu5g04/dist-node/bricks/pager.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Pager = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _row = require('./row.js');

var _row2 = _interopRequireDefault(_row);

var _icon = require('./icon.js');

var _icon2 = _interopRequireDefault(_icon);

var _span = require('./span.js');

var _span2 = _interopRequireDefault(_span);

var _button = require('./button.js');

var _button2 = _interopRequireDefault(_button);

require('./pager.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Pager = exports.Pager = (0, _createReactClass2.default)({
  displayName: 'Pager',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.ColorSchemaMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Pager"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'smallBox'),
    classNames: {
      main: _bricksNs2.default.css("pager uu5-common-text"),
      link: _bricksNs2.default.css("pager-link"),
      text: _bricksNs2.default.css("pager-text"),
      icon: _bricksNs2.default.css("pager-icon"),
      bg: 'uu5-common-bg',
      size: _bricksNs2.default.css("pager-size-")
    },
    defaults: {
      leftLink: 'mdi-chevron-left',
      rightLink: 'mdi-chevron-right',
      upLink: 'mdi-chevron-up',
      downLink: 'mdi-chevron-down'
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    leftLink: _propTypes2.default.shape({
      text: _propTypes2.default.any,
      href: _propTypes2.default.string,
      icon: _propTypes2.default.string,
      onClick: _propTypes2.default.function
    }),
    rightLink: _propTypes2.default.shape({
      text: _propTypes2.default.any,
      href: _propTypes2.default.string,
      icon: _propTypes2.default.string,
      onClick: _propTypes2.default.function
    }),
    upLink: _propTypes2.default.shape({
      text: _propTypes2.default.any,
      href: _propTypes2.default.string,
      icon: _propTypes2.default.string,
      onClick: _propTypes2.default.function
    }),
    downLink: _propTypes2.default.shape({
      text: _propTypes2.default.any,
      href: _propTypes2.default.string,
      icon: _propTypes2.default.string,
      onClick: _propTypes2.default.function
    }),
    size: _propTypes2.default.oneOf(['s', 'm', 'l', 'xl']),
    background: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      leftLink: null,
      rightLink: null,
      upLink: null,
      downLink: null,
      size: 'm',
      background: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getColumn: function _getColumn(key) {
    if (this.props[key]) {
      var buttonProps = UU5.Common.Tools.merge({}, this.props[key]);
      buttonProps.className = buttonProps.className || '';
      buttonProps.className += ' ' + this.getClassName('link');
      buttonProps.bgStyle = this.props.background ? 'filled' : 'transparent';
      buttonProps.size = this.props.size;
      if (!buttonProps.text) {
        buttonProps.style = buttonProps.style || {};
        var widht = void 0;
        switch (this.props.size) {
          case 's':
            widht = 24;
            break;
          case 'm':
            widht = 32;
            break;
          case 'l':
            widht = 40;
            break;
          case 'xl':
            widht = 48;
            break;
          default:
            widht = 32;
            break;
        }
        buttonProps.style.width = widht;
      }

      var content = [_react2.default.createElement(_icon2.default, { icon: buttonProps.icon || this.getDefault(key), className: this.getClassName('icon') }), buttonProps.text ? _react2.default.createElement(_span2.default, { className: this.getClassName('text'), content: buttonProps.text }) : null];

      if (key === 'rightLink' || key === 'downLink') {
        content.push(content.shift());
      }

      return _react2.default.createElement(_button2.default, _extends({}, buttonProps, { content: content }));
    }
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var mainProps = this.getMainPropsToPass();
    mainProps.nestingLevel = this.getNestingLevel();
    mainProps.className += ' ' + this.getClassName('size') + this.props.size;

    return _react2.default.createElement(
      _row2.default,
      mainProps,
      this._getColumn('leftLink'),
      this._getColumn('upLink'),
      this._getColumn('downLink'),
      this._getColumn('rightLink')
    );
  }
  //@@viewOff:render

});

exports.default = Pager;
