// uu5g04/dist-node/bricks/panel-header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _icon = require('./icon.js');

var _icon2 = _interopRequireDefault(_icon);

require('./panel-header.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = (0, _createReactClass2.default)({
  displayName: 'panel-header',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Panel.Header"),
    classNames: {
      main: _bricksNs2.default.css("panel-header"),
      click: _bricksNs2.default.css("panel-header-click"),
      content: _bricksNs2.default.css("panel-header-content")
    },
    defaults: {
      parentTagName: 'UU5.Bricks.Panel'
    },
    errors: {
      invalidParent: 'Parent of this component is not Panel.'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    _icon: _propTypes2.default.string,
    _onClick: _propTypes2.default.func,
    _disableHeaderClick: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      _icon: null,
      _onClick: null,
      _disableHeaderClick: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function componentWillMount() {
    //this.checkParentTagName(this.getDefault().parentTagName);

    if (!(this.getParent() && this.getParent().isPanel)) {
      this.showError("invalidParent");
    }
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onClickHandler: function _onClickHandler(e) {
    !this.props._disableHeaderClick && this.props._onClick(e);
    return this;
  },

  _buildMainAttrs: function _buildMainAttrs() {
    var mainAttrs = this.getMainAttrs();
    !this.props._disableHeaderClick && (mainAttrs.className += ' ' + this.getClassName().click);
    return mainAttrs;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return _react2.default.createElement(
      'div',
      _extends({}, this._buildMainAttrs(), { onClick: this._onClickHandler }),
      _react2.default.createElement(
        'span',
        { className: this.getClassName('content') },
        this.getChildren()
      ),
      _react2.default.createElement(_icon2.default, { icon: this.props._icon }),
      this.getDisabledCover()
    );
  }
  //@@viewOff:render
});
