// uu5g04/dist-node/bricks/popover.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Popover = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _factory = require('./factory.js');

require('./popover.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Popover = exports.Popover = (0, _createReactClass2.default)({
  displayName: 'Popover',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin, UU5.Common.NestingLevelMixin, UU5.Common.CcrWriterMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("ContextMenu"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'box'),
    classNames: {
      main: _bricksNs2.default.css("popover"),
      open: _bricksNs2.default.css("popover-shown"),
      header: _bricksNs2.default.css("popover-header"),
      body: _bricksNs2.default.css("popover-body"),
      footer: _bricksNs2.default.css("popover-footer"),
      top: _bricksNs2.default.css("popover-menu-top"),
      bottom: _bricksNs2.default.css("popover-menu-bottom"),
      left: _bricksNs2.default.css("popover-menu-left"),
      right: _bricksNs2.default.css("popover-menu-right")
    },
    defaults: {
      transitionDuration: 100
    },
    opt: {
      nestingLevelRoot: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    shown: _propTypes2.default.bool,
    parentElement: _propTypes2.default.object
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      shown: false,
      parentElement: null
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      header: null,
      content: null,
      footer: null,
      pageX: null,
      pageY: null,
      position: {
        top: false,
        left: false
      }
    };
  },
  componentWillMount: function componentWillMount() {
    this.setState({ hidden: !this.props.shown });
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (nextProps.controlled) {
      this.setState({ hidden: !nextProps.shown });
    }
  },


  componentWillUnmount: function componentWillUnmount() {
    UU5.Environment.EventListener.removeWindowEvent('click', this.getId());
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  open: function open(opt, setStateCallback) {
    var _this = this;

    this._addEvent();
    this._stopPropagation = true;
    var pageX = opt.pageX || 0;
    var pageY = opt.pageY || 0;
    var position = UU5.Common.Tools.mergeDeep({}, this.state.position);

    if (opt.event) {
      // opt.event.preventDefault();
      // opt.event.stopPropagation();

      pageX = pageX || opt.event.pageX;
      pageY = pageY || opt.event.pageY;
    }

    var parent = void 0;
    var parentElement = opt.parentElement === undefined ? this.props.parentElement : opt.parentElement;

    if (parentElement) {
      parent = _reactDom2.default.findDOMNode(parentElement);
    } else {
      var i = 0;
      do {
        i++;
        parent = UU5.Common.Tools.getElementByComputedStyle(document.getElementById(parent ? parent.id : this.getId()), 'position', ['relative', 'absolute', 'fixed']);
        // TODO not good!!!!
      } while (parent.id && (parent.className.split(" ").indexOf(this.getClassName("main")) > -1 || parent.className.split(" ").indexOf("uu5-bricks-context-menu-item") > -1) && i < 100);
    }
    if (parent) {
      var parentX = UU5.Common.Tools.getOffsetLeft(parent);
      var parentY = UU5.Common.Tools.getOffsetTop(parent);
      pageX = pageX - parentX;
      pageY = pageY - parentY;
    }

    this.setState({
      position: position,
      header: opt.header || this.getHeader(),
      content: opt.content ? this.buildChildren({ content: opt.content }) : this.getChildren(),
      footer: opt.footer || this.getFooter(),
      hidden: true,
      pageX: 0,
      pageY: 0
    }, function () {
      var parentHeight = parent ? UU5.Common.Tools.getInnerHeight(parent) : window.innerHeight;
      var parentWidth = parent ? UU5.Common.Tools.getInnerWidth(parent) : window.innerWidth;
      if (parentHeight < pageY + UU5.Common.Tools.getOuterHeight(_this, true) && pageY >= UU5.Common.Tools.getOuterHeight(_this, true)) {
        pageY = pageY - UU5.Common.Tools.getOuterHeight(_this, true);
        position.top = true;
      } else {
        position.top = false;
      }

      if (parentWidth < pageX + UU5.Common.Tools.getOuterWidth(_this, true) && pageX >= UU5.Common.Tools.getOuterWidth(_this, true)) {
        pageX = pageX - UU5.Common.Tools.getOuterWidth(_this, true);
        position.left = true;
      } else {
        position.left = false;
      }
      _this.setState({
        hidden: false,
        pageX: pageX,
        pageY: pageY
      }, function () {
        typeof setStateCallback === 'function' && setStateCallback();
      });
    });
    return this;
  },
  close: function close(setStateCallback) {
    this._removeEvent();

    this.setState({
      // header: null,
      // content: null,
      // footer: null,
      hidden: true
    }, setStateCallback);

    return this;
  },
  isOpen: function isOpen() {
    return !this.state.hidden;
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _findTarget: function _findTarget(item) {
    var result = false;
    var id = this.getId();

    if (item.id === id) {
      result = true;
    } else if (item.parentElement) {
      result = this._findTarget(item.parentElement);
    }

    return result;
  },
  _addEvent: function _addEvent() {
    var _this2 = this;

    UU5.Environment.EventListener.addWindowEvent('click', this.getId(), function (e) {
      var isPopover = _this2._findTarget(e.target);

      if (!_this2._stopPropagation && !isPopover && !_this2.isHidden()) {
        _this2.close();
      } else {
        _this2._stopPropagation = false;
      }
    });
    UU5.Environment.EventListener.addWindowEvent('contextmenu', this.getId(), function (e) {
      var isPopover = _this2._findTarget(e.target);

      if (!_this2._stopPropagation && !isPopover && !_this2.isHidden()) {
        _this2.close();
      } else {
        _this2._stopPropagation = false;
      }
    });
    return this;
  },
  _removeEvent: function _removeEvent() {
    UU5.Environment.EventListener.removeWindowEvent('click', this.getId());
    UU5.Environment.EventListener.removeWindowEvent('contextmenu', this.getId());
    return this;
  },
  _getMainAttrs: function _getMainAttrs() {
    var props = this.getMainAttrs();

    props.id = this.getId();

    !this.isHidden() && (props.className += ' ' + this.getClassName().open);
    props.className += this.state.position.top ? ' ' + this.getClassName('top') : ' ' + this.getClassName('bottom');
    props.className += this.state.position.left ? ' ' + this.getClassName('left') : ' ' + this.getClassName('right');

    props.style = {};
    if (this.state.pageX !== null) {
      props.style = {
        left: this.state.pageX,
        top: this.state.pageY
      };
    }

    //opera mini dont know how to use transitions :(
    if (navigator.userAgent.indexOf('Opera Mini') == -1) {
      var time = this.getDefault().transitionDuration / 1000;
      ['WebkitTransitionDuration', 'MozTransitionDuration', 'MsTransitionDuration', 'OTransitionDuration', 'transitionDuration'].forEach(function (style) {
        // props.style[style] = time + 's';
      });
    }

    return props;
  },
  _getHeader: function _getHeader() {
    var header = this.state.header || this.getHeader();
    if (header) {
      header = _react2.default.createElement(
        'div',
        { className: this.getClassName().header, key: 'header' },
        header
      );
    }
    return header;
  },
  _getFooter: function _getFooter() {
    var footer = this.state.footer || this.getFooter();
    if (footer) {
      footer = _react2.default.createElement(
        'div',
        { className: this.getClassName().footer, key: 'footer' },
        footer
      );
    }
    return footer;
  },
  _getBody: function _getBody() {
    var children = this.buildChildren({ content: this.state.content, children: this.props.children });
    if (children) {
      children = _react2.default.createElement(
        'div',
        { className: this.getClassName().body, key: 'chidren' },
        children
      );
    }
    return children;
  },


  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return _react2.default.createElement(
      'div',
      this._getMainAttrs(),
      this._getHeader(),
      this._getBody(),
      this._getFooter(),
      this.getDisabledCover()
    );
  }
  //@@viewOff:render

});

exports.default = Popover;
