// uu5g04/dist-node/bricks/table.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Table = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _bricksNs = require('./bricks-ns.js');

var _bricksNs2 = _interopRequireDefault(_bricksNs);

var _tableCol = require('./table-col.js');

var _tableCol2 = _interopRequireDefault(_tableCol);

var _tableColGroup = require('./table-col-group.js');

var _tableColGroup2 = _interopRequireDefault(_tableColGroup);

var _tableTbody = require('./table-tbody.js');

var _tableTbody2 = _interopRequireDefault(_tableTbody);

var _tableTd = require('./table-td.js');

var _tableTd2 = _interopRequireDefault(_tableTd);

var _tableTfoot = require('./table-tfoot.js');

var _tableTfoot2 = _interopRequireDefault(_tableTfoot);

var _tableTh = require('./table-th.js');

var _tableTh2 = _interopRequireDefault(_tableTh);

var _tableThead = require('./table-thead.js');

var _tableThead2 = _interopRequireDefault(_tableThead);

var _tableTr = require('./table-tr.js');

var _tableTr2 = _interopRequireDefault(_tableTr);

require('./table.less');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Table = exports.Table = (0, _createReactClass2.default)({
  displayName: 'Table',


  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin, UU5.Common.ColorSchemaMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _bricksNs2.default.name("Table"),
    nestingLevelList: UU5.Environment.getNestingLevelList('bigBoxCollection', 'smallBox'),
    classNames: {
      main: _bricksNs2.default.css("table"),
      table: _bricksNs2.default.css("table-table"),
      striped: _bricksNs2.default.css("table-striped"),
      bordered: _bricksNs2.default.css("table-bordered"),
      hover: _bricksNs2.default.css("table-hover"),
      condensed: _bricksNs2.default.css("table-condensed"),
      responsive: _bricksNs2.default.css("table-responsive")
    },
    defaults: {
      childTagNames: ['UU5.Bricks.Table.Tr', 'UU5.Bricks.Table.THead', 'UU5.Bricks.Table.TBody', 'UU5.Bricks.Table.TFoot', 'UU5.Bricks.Table.ColGroup']
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    striped: _propTypes2.default.bool,
    bordered: _propTypes2.default.bool,
    hover: _propTypes2.default.bool,
    condensed: _propTypes2.default.bool,
    responsive: _propTypes2.default.bool,
    allowTags: _propTypes2.default.arrayOf(_propTypes2.default.string)
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      striped: false,
      bordered: false,
      hover: false,
      condensed: false,
      responsive: false,
      allowTags: []
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isTable: function isTable() {
    return true;
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  shouldChildRender_: function shouldChildRender_(child) {
    var childTagName = UU5.Common.Tools.getChildTagName(child);
    var childTagNames = this.props.allowTags.concat(this.getDefault().childTagNames);
    return childTagNames.indexOf(childTagName) > -1;
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainProps: function _getMainProps() {
    var props = this.getMainAttrs();
    this.props.responsive && (props.className += ' ' + this.getClassName().responsive);
    return props;
  },

  _buildTableAttrs: function _buildTableAttrs() {
    var tableAttrs = { className: this.getClassName().table };
    this.props.striped && (tableAttrs.className += ' ' + this.getClassName().striped);
    this.props.bordered && (tableAttrs.className += ' ' + this.getClassName().bordered);
    this.props.hover && (tableAttrs.className += ' ' + this.getClassName().hover);
    this.props.condensed && (tableAttrs.className += ' ' + this.getClassName().condensed);
    return tableAttrs;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? _react2.default.createElement(
      'div',
      this._getMainProps(),
      this.getHeaderChild(),
      _react2.default.createElement(
        'table',
        this._buildTableAttrs(),
        this.getChildren()
      ),
      this.getFooterChild(),
      this.getDisabledCover()
    ) : null;
  }
  //@@viewOff:render
});

Table.ColGroup = _tableColGroup2.default;
Table.Col = _tableCol2.default;
Table.TBody = _tableTbody2.default;
Table.Td = _tableTd2.default;
Table.TFoot = _tableTfoot2.default;
Table.Th = _tableTh2.default;
Table.THead = _tableThead2.default;
Table.Tr = _tableTr2.default;

exports.default = Table;
