// uu5g04/dist-node/core/common/div.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Div = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _commonNs = require('./common-ns.js');

var _commonNs2 = _interopRequireDefault(_commonNs);

var _baseMixin = require('./base-mixin.js');

var _baseMixin2 = _interopRequireDefault(_baseMixin);

var _elementaryMixin = require('./elementary-mixin.js');

var _elementaryMixin2 = _interopRequireDefault(_elementaryMixin);

var _contentMixin = require('./content-mixin.js');

var _contentMixin2 = _interopRequireDefault(_contentMixin);

var _nestingLevelMixin = require('./nesting-level-mixin.js');

var _nestingLevelMixin2 = _interopRequireDefault(_nestingLevelMixin);

var _environment = require('../environment/environment.js');

var _environment2 = _interopRequireDefault(_environment);

var _pureRenderMixin = require('./pure-render-mixin');

var _pureRenderMixin2 = _interopRequireDefault(_pureRenderMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Div = exports.Div = (0, _createReactClass2.default)({
  displayName: 'Div',


  //@@viewOn:mixins
  mixins: [_baseMixin2.default, _elementaryMixin2.default, _contentMixin2.default, _nestingLevelMixin2.default, _pureRenderMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _commonNs2.default.name("Div"),
    classNames: {
      main: _commonNs2.default.css("div")
    },
    nestingLevelList: _environment2.default.getNestingLevelList('bigBoxCollection', 'smallBox')
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this.getNestingLevel() ? _react2.default.createElement('div', this.getMainAttrs(), this.getChildren(), this.getDisabledCover()) : null;
  }
  //@@viewOff:render
});

exports.default = Div;
