// uu5g04/dist-node/core/common/outline-modal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OutlineModal = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _commonNs = require('./common-ns.js');

var _commonNs2 = _interopRequireDefault(_commonNs);

var _baseMixin = require('./base-mixin.js');

var _baseMixin2 = _interopRequireDefault(_baseMixin);

var _elementaryMixin = require('./elementary-mixin.js');

var _elementaryMixin2 = _interopRequireDefault(_elementaryMixin);

var _tools = require('./tools.js');

var _tools2 = _interopRequireDefault(_tools);

var _pureRenderMixin = require('./pure-render-mixin');

var _pureRenderMixin2 = _interopRequireDefault(_pureRenderMixin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OutlineModal = exports.OutlineModal = (0, _createReactClass2.default)({
  displayName: 'OutlineModal',


  //@@viewOn:mixins
  mixins: [_baseMixin2.default, _elementaryMixin2.default, _pureRenderMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _commonNs2.default.name("Outline.Modal"),
    classNames: {
      main: _commonNs2.default.css("outline-modal")
    },
    errors: {
      invalidParent: 'Parent of this component is not Outline.'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      tag: null,
      props: null
    };
  },

  componentWillMount: function componentWillMount() {
    if (!this.getParentByType('isOutline')) {
      this.showError("invalidParent");
    }
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  open: function open(tag, props, setStateCallback) {
    this.setState({ shown: true, tag: tag, props: props }, setStateCallback);
    return this;
  },

  close: function close(setStateCallback) {
    this.setState({ shown: false, tag: null, props: null }, setStateCallback);
    return this;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    var props = _tools2.default.mergeDeep({}, this.getMainPropsToPass(), {
      shown: this.state.shown,
      header: this.state.tag,
      content: _react2.default.createElement(
        'pre',
        null,
        this.state.props
      )
    });
    return _tools2.default.findComponent('UU5.Bricks.Modal', props);
  }
  //@@viewOff:render
});

exports.default = OutlineModal;
