// uu5g04/dist-node/core/common/pure-render-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PureRenderMixin = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PureRenderMixin = exports.PureRenderMixin = {

  //@@viewOn:statics
  statics: {
    "UU5.Common.PureRenderMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"]
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    pureRender: _propTypes2.default.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      pureRender: false
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  shouldComponentUpdate: function shouldComponentUpdate(nextProps, nextState) {
    var shouldUpdate = void 0;
    if (!this.props.pureRender) {
      shouldUpdate = true;
    } else if (typeof this.shouldComponentUpdate_ === 'function') {
      shouldUpdate = this.shouldComponentUpdate_(nextProps, nextState);
    } else {
      shouldUpdate = this.shouldRender(nextProps, nextState);
    }
    return shouldUpdate;
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonPureRenderMixin: function hasUU5CommonPureRenderMixin() {
    return this.hasMixin("UU5.Common.PureRenderMixin");
  },
  getUU5CommonPureRenderMixinProps: function getUU5CommonPureRenderMixinProps() {
    return {
      pureRender: this.props.pureRender
    };
  },
  getUU5CommonPureRenderMixinPropsToPass: function getUU5CommonPureRenderMixinPropsToPass() {
    return this.getUU5CommonPureRenderMixinProps();
  }
};

exports.default = PureRenderMixin;
