// uu5g04/dist-node/core/common/router.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Router = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _commonNs = require('./common-ns.js');

var _commonNs2 = _interopRequireDefault(_commonNs);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _tools = require('./tools.js');

var _tools2 = _interopRequireDefault(_tools);

var _environment = require('../environment/environment.js');

var _environment2 = _interopRequireDefault(_environment);

var _baseMixin = require('./base-mixin.js');

var _baseMixin2 = _interopRequireDefault(_baseMixin);

var _elementaryMixin = require('./elementary-mixin.js');

var _elementaryMixin2 = _interopRequireDefault(_elementaryMixin);

var _ccrWriterMixin = require('./ccr-writer-mixin.js');

var _ccrWriterMixin2 = _interopRequireDefault(_ccrWriterMixin);

var _url = require('./url.js');

var _url2 = _interopRequireDefault(_url);

var _pureRenderMixin = require('./pure-render-mixin');

var _pureRenderMixin2 = _interopRequireDefault(_pureRenderMixin);

require('./router.less');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Router = exports.Router = (0, _createReactClass2.default)({
  displayName: 'Router',


  //@@viewOn:mixins
  mixins: [_baseMixin2.default, _elementaryMixin2.default, _ccrWriterMixin2.default, _pureRenderMixin2.default],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: _commonNs2.default.name("Router"),
    opt: {
      ccrKey: _environment2.default.CCRKEY_ROUTER
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    basePath: _propTypes2.default.string,
    route: _propTypes2.default.oneOfType([_propTypes2.default.string, // path
    _propTypes2.default.element, _propTypes2.default.shape({
      tag: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.func]),
      props: _propTypes2.default.object,
      source: _propTypes2.default.string
    })]),
    notFoundRoute: _propTypes2.default.oneOfType([_propTypes2.default.string, // path
    _propTypes2.default.element, _propTypes2.default.shape({
      tag: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.func]),
      props: _propTypes2.default.object
    })]),
    routes: _propTypes2.default.object,
    urlBuilder: _propTypes2.default.func
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      basePath: null,
      route: null,
      notFoundRoute: null,
      routes: null,
      urlBuilder: null
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    _environment2.default.router = this;
    this._history = [];
    this._routeIndex = null;

    return {
      route: null
    };
  },
  componentWillMount: function componentWillMount() {
    if (window.location.pathname && this.props.routes && typeof this.props.route === 'string') {
      var path = window.location.pathname;
      var basePath = this._getBasePath(path, this.props);
      basePath && (path = path.replace(basePath, ''));

      var params = null;
      if (window.location.search) {
        params = _tools2.default.decodeQuery(window.location.search);
      }
      this._setRoute(path, params, this.props);
    } else if (!this._shouldImport(this.props.route)) {
      this._setRoute(this.props.route, null, this.props);
    }
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (nextProps.controlled) {
      this._setRoute(nextProps.route, null, nextProps);
    }
  },
  componentDidMount: function componentDidMount() {
    if (this._shouldImport(this.props.route)) {
      this._importRoute(this.props.route);
    }

    window.onpopstate = this._routeByNavigation;
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  // newRoute = {tag: 'Node', props: {}, source: '', notFoundRoute: {} || < />} || <Node {...props} />
  setRoute: function setRoute(newRoute, params, setStateCallback) {
    if (this._shouldImport(newRoute)) {
      this._importRoute(newRoute, params, setStateCallback);
    } else {
      this._setRoute(newRoute, params, null, setStateCallback);
    }
    return this;
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getBasePath: function _getBasePath(path, props) {
    path = path || window.location.pathname;
    return props.basePath === null ? path ? path.replace(/(\/.*?\/.*?)\/.*/, "$1") : '' : props.basePath;
  },
  _shouldImport: function _shouldImport(route) {
    return route && (typeof route === 'undefined' ? 'undefined' : _typeof(route)) === 'object' && route.tag && !_tools2.default.findComponent(route.tag, true);
  },
  _importRoute: function _importRoute(route, params, setStateCallback) {
    var _this = this;

    if (window.System && window.System.import) {
      if (route.source) {
        System.import(route.source).then(function (exports) {
          var tagArray = route.tag.split('.');
          var calculatedTag = window;
          while (calculatedTag && tagArray.length > 1) {
            var moduleName = tagArray.shift();
            calculatedTag[moduleName] = calculatedTag[moduleName] || {};
            calculatedTag = calculatedTag[moduleName];
          }
          var tagName = tagArray.shift();
          var tagExport = exports[tagName];
          tagExport = tagExport || exports.default;
          calculatedTag[tagName] = tagExport;
          _this._setRoute(route, params, null, setStateCallback);
        }, function (error) {
          _tools2.default.error('Loading package "' + route.source + '" failed with error:', {
            error: error,
            tagName: _this.constructor.tagName
          });
        });
      } else {
        _tools2.default.error('Route was not found and has not set any source to import.', {
          route: route,
          tagName: this.constructor.tagName
        });
        this._setRoute(route.notFoundRoute);
      }
    } else {
      _tools2.default.error('System is not defined in window! Cannot import source:', {
        source: route.source,
        tagName: this.constructor.tagName
      });
    }
    return this;
  },
  _setRoute: function _setRoute(newRoute, params, props, setStateCallback) {
    props = props || this.props;
    this.setState({ route: this._buildRoute(newRoute, params, null, props) }, setStateCallback);
    return this;
  },
  _getRouteByPath: function _getRouteByPath(searchedPath) {
    return this.props.routes[searchedPath] || null;
  },
  _routeByNavigation: function _routeByNavigation(event) {
    if (event.state) {
      if (event.state.index < this._routeIndex) {
        // go back
        this._routeIndex = event.state.index;
        // console.log('goBack', this._history[this._routeIndex], event.state, this._routeIndex, this._history);
      } else {
        // go forward
        this._routeIndex = event.state.index;
        // console.log('goForward', this._history[this._routeIndex], event.state, this._routeIndex, this._history);
      }
      var newRoute = this._history[this._routeIndex];
      newRoute && this.setState({ route: newRoute.component ? this._buildComponent(newRoute.component) : this._buildRoute(newRoute.path, newRoute.params, true) });
    }
    // else {
    //   console.log('goInvalid', window.location.pathname, event.state);
    // }
    return this;
  },
  _getUrl: function _getUrl(route, params, props) {
    props = props || this.props;
    var url = document.location.origin + (this._getBasePath(null, props) || '') + route;
    params && (url += _tools2.default.encodeQuery(params));
    return url;
  },
  _buildComponent: function _buildComponent(newRoute, params) {
    var newRouteChild = void 0;

    if (newRoute && (typeof newRoute === 'undefined' ? 'undefined' : _typeof(newRoute)) === 'object') {
      var newProps = {
        parent: this.getParent()
      };

      params && (newProps.params = params);

      if (newRoute.tag) {
        var tag = _tools2.default.checkTag(newRoute.tag, true);
        var props = _tools2.default.merge({}, newRoute.props, newProps);
        newRouteChild = tag ? _react2.default.createElement(tag, props) : _tools2.default.findComponent(newRoute.tag, props);
      } else {
        newRouteChild = _react2.default.cloneElement(newRoute, newProps);
      }
    }

    return newRouteChild;
  },
  _buildRoute: function _buildRoute(route, params, replace, props) {
    var newRouteChild = null;
    props = props || this.props;

    if (route) {
      var newRoute = route;

      if (typeof route === 'string' && this.props.routes) {
        var foundRoute = this._getRouteByPath(route);
        var method = 'replaceState';

        if (foundRoute) {
          if (!foundRoute.noHistory) {
            if (this._routeIndex === null) {
              this._routeIndex = 0;
              this._history.push({ path: route, params: params });
            } else if (!replace) {
              this._routeIndex++;
              this._history.splice(this._routeIndex, this._history.length - 1);
              method = 'pushState';
              this._history.push({ path: route, params: params });
            }
          }

          // console.log(method, route, this._routeIndex, this._history);
          history[method]({ path: route, index: this._routeIndex }, document.title, this._getUrl(route, params, props));

          newRoute = foundRoute.component;
        }
      } else {
        var _foundRoute = route;
        var _method = 'replaceState';

        params = params || {};
        var path = void 0;

        if (params.url && _typeof(params.url) === 'object') {
          var urlBuilder = this.props.urlBuilder || _environment2.default.getUrlBuilder() || _url2.default;
          path = urlBuilder.parse(window.location.href).set(params.url).toString();
        } else {
          path = params.url || window.location.href;
        }

        if (_foundRoute) {
          if (!params.noHistory) {
            if (this._routeIndex === null) {
              this._routeIndex = 0;
              this._history.push({ path: path, component: _foundRoute });
            } else if (!replace) {
              this._routeIndex++;
              this._history.splice(this._routeIndex, this._history.length - 1);
              _method = 'pushState';
              this._history.push({ path: path, component: _foundRoute });
            }
          }

          history[_method]({ path: path, index: this._routeIndex }, params.title || document.title, path);

          newRoute = _foundRoute;
        }
      }

      newRouteChild = this._buildComponent(newRoute, params);
    }

    if (newRouteChild && (!newRouteChild.type || !newRouteChild.type["UU5.Common.VucMixin"] && !newRouteChild.type["UU5.Common.RouteMixin"])) {
      _tools2.default.error('Route component which should be set is not Visual Use Case.', {
        routeParam: route,
        routeChild: newRouteChild,
        tagName: this.constructor.tagName
      });
    }

    return newRouteChild;
  },
  _buildChild: function _buildChild() {
    var child = void 0;

    if (this.state.route) {
      child = this.state.route;
    } else if (this.props.notFoundRoute) {
      child = this._buildRoute(this.props.notFoundRoute);
    } else {
      child = _react2.default.createElement('div', null);
      _tools2.default.error(this.constructor.tagName, 'Router has no content.');
    }

    return child;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function render() {
    return this._buildChild();
  }
  //@@viewOff:render

});

exports.default = Router;
