// uu5g04/dist-esm/forms/select-option.js
import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import ns from "./forms-ns.js";

import './select-option.less';

export default createReactClass({

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("Select.option"),
    classNames: {
      main: ns.css("select-option", "group-item")
    },
    errors: {
      invalidParent: 'Parent of this component is not Select.'
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    value: PropTypes.string.isRequired,
    selectedContent: PropTypes.any,
    onClick: PropTypes.func
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      value: null,
      selectedContent: null,
      onClick: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount() {
    // because of auto-completed items
    if (!this.getParentByType('isSelect') && !this.getParentByType('isTextInput')) {
      this.showError("invalidParent");
    }
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  getValue() {
    return this.props.value;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainProps() {
    var props = this.getMainPropsToPass();

    props.onClick = (link, e) => {
      this.props.onClick({ component: this, event: e, value: this.getIndex() });
    };

    if (!props.content && !this.props.children) {
      props.content = this.props.value;
    }

    props.content = props.content || this.props.children && React.Children.toArray(this.props.children);

    return props;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return <UU5.Bricks.Link {...this._getMainProps()} />;
  }
  //@@viewOff:render
});
