// uu5g04/dist-node/forms/mixins/group-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GroupMixin = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

require('./group-mixin.less');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var GroupMixin = exports.GroupMixin = {

  //@@viewOn:mixins
  //@@viewOff:mixins

  //@@viewOn:statics
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    value: _propTypes2.default.arrayOf(_propTypes2.default.shape({
      label: _propTypes2.default.any.isRequired,
      name: _propTypes2.default.string,
      value: _propTypes2.default.bool,
      disabled: _propTypes2.default.bool,
      hidden: _propTypes2.default.bool
    })),
    required: _propTypes2.default.bool,
    requiredMessage: _propTypes2.default.any,
    onIcon: _propTypes2.default.string,
    offIcon: _propTypes2.default.string,
    labelPosition: _propTypes2.default.oneOf(['left', 'right']),
    inline: _propTypes2.default.bool

  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      value: [],
      required: false,
      requiredMessage: null,
      onIcon: '',
      offIcon: '',
      inline: false,
      labelPosition: 'right'
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isValid: function isValid() {
    var feedback = this.getFeedback();
    var value = this.getValue();
    var result = true;

    if (this.props.required && (value === null || value.length < 1)) {
      this.setError(this.props.requiredMessage || this.getLsiComponent('requiredMessageGroup'));
      result = false;
    } else if (feedback === 'error') {
      result = false;
    } else if (typeof this.isValid_ === 'function') {
      result = this.isValid_();
    }

    if (result && this.props.onValidate) {
      var validation = this.props.onValidate(value, this);
      if (validation && (typeof validation === 'undefined' ? 'undefined' : _typeof(validation)) === 'object') {
        if (validation.feedback === 'error') {
          result = false;
        }
      }
    }

    return result;
  }
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  //@@viewOff:render
};

exports.default = GroupMixin;
