// uu5g04/dist-esm/forms/mixins/choice-mixin.js
import React from 'react';
import PropTypes from 'prop-types';

export const ChoiceMixin = {

  //@@viewOn:mixins
  //@@viewOff:mixins

  //@@viewOn:statics
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    placeholder: PropTypes.any,
    required: PropTypes.bool,
    requiredMessage: PropTypes.any,
    buttonHidden: PropTypes.bool,
    iconOpen: PropTypes.string,
    iconClosed: PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      placeholder: null,
      required: false,
      requiredMessage: null,
      buttonHidden: false,
      iconOpen: 'uu5-arrow-up',
      iconClosed: 'uu5-arrow-down'
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isValid() {
    var feedback = this.getFeedback();
    var value = this.getValue();
    var result = true;

    if (this.props.required && (!value || value.length < 1)) {
      this.setError(this.props.requiredMessage || this.getLsiComponent('requiredMessageChoice'));
      result = false;
    } else if (feedback === 'error') {
      result = false;
    } else if (typeof this.isValid_ === 'function') {
      result = this.isValid_();
    }

    if (result && this.props.onValidate) {
      var validation = this.props.onValidate(value, this);
      if (validation && typeof validation === 'object') {
        if (validation.feedback === 'error') {
          result = false;
        }
      }
    }

    return result;
  }
};

export default ChoiceMixin;
