// uu5g04/dist-esm/forms/mixins/controls-mixin.js
import React from 'react';
import * as UU5 from "uu5g04";

export const ControlsMixin = {

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentDidMount: function () {
    let parentForm = this.getForm();
    parentForm && parentForm.registerFormControls(this.getId(), this);
  },

  componentWillReceiveProps: function (nextProps) {
    if (nextProps.id && nextProps.id !== this.props.id) {
      let parentForm = this.getForm();
      if (parentForm) {
        parentForm.unregisterFormControls(this.props.id);
        parentForm.registerFormControls(nextProps.id, this);
      }
    }
  },

  componentWillUnmount: function () {
    let parentForm = this.getForm();
    parentForm && parentForm.unregisterFormControls(this.getId());
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  getForm() {
    return this.getParentByType('isForm');
  }
};

export default ControlsMixin;
