// uu5g04/dist-node/forms/mixins/choice-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChoiceMixin = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ChoiceMixin = exports.ChoiceMixin = {

  //@@viewOn:mixins
  //@@viewOff:mixins

  //@@viewOn:statics
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    placeholder: _propTypes2.default.any,
    required: _propTypes2.default.bool,
    requiredMessage: _propTypes2.default.any,
    buttonHidden: _propTypes2.default.bool,
    iconOpen: _propTypes2.default.string,
    iconClosed: _propTypes2.default.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      placeholder: null,
      required: false,
      requiredMessage: null,
      buttonHidden: false,
      iconOpen: 'uu5-arrow-up',
      iconClosed: 'uu5-arrow-down'
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  isValid: function isValid() {
    var feedback = this.getFeedback();
    var value = this.getValue();
    var result = true;

    if (this.props.required && (!value || value.length < 1)) {
      this.setError(this.props.requiredMessage || this.getLsiComponent('requiredMessageChoice'));
      result = false;
    } else if (feedback === 'error') {
      result = false;
    } else if (typeof this.isValid_ === 'function') {
      result = this.isValid_();
    }

    if (result && this.props.onValidate) {
      var validation = this.props.onValidate(value, this);
      if (validation && (typeof validation === 'undefined' ? 'undefined' : _typeof(validation)) === 'object') {
        if (validation.feedback === 'error') {
          result = false;
        }
      }
    }

    return result;
  }
};

exports.default = ChoiceMixin;
