// uu5g04/dist-node/forms/mixins/controls-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ControlsMixin = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ControlsMixin = exports.ControlsMixin = {

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentDidMount: function componentDidMount() {
    var parentForm = this.getForm();
    parentForm && parentForm.registerFormControls(this.getId(), this);
  },

  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (nextProps.id && nextProps.id !== this.props.id) {
      var parentForm = this.getForm();
      if (parentForm) {
        parentForm.unregisterFormControls(this.props.id);
        parentForm.registerFormControls(nextProps.id, this);
      }
    }
  },

  componentWillUnmount: function componentWillUnmount() {
    var parentForm = this.getForm();
    parentForm && parentForm.unregisterFormControls(this.getId());
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  getForm: function getForm() {
    return this.getParentByType('isForm');
  }
};

exports.default = ControlsMixin;
