// uu5g04/dist-node/forms/mixins/form-mixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormMixin = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uu5g = require('uu5g04');

var UU5 = _interopRequireWildcard(_uu5g);

var _formsNs = require('../forms-ns.js');

var _formsNs2 = _interopRequireDefault(_formsNs);

require('uu5g04-bricks');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FormMixin = exports.FormMixin = {

  //@@viewOn:mixins
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    "UU5.Forms.FormMixin": {
      warnings: {
        formInForm: 'Form control %s should not be placed in other form control %s.',
        nonRegistered: 'Form control with ID %s cannot be unregistered. Component with the ID is not registered.',
        noName: 'Form control has not any name. It will be used its ID %s.'
      },
      errors: {
        duplicateId: 'Duplicate id \'%s\' of a form control.'
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    values: _propTypes2.default.object,
    progressIndicator: _propTypes2.default.any, // loading
    onInit: _propTypes2.default.func,
    onSave: _propTypes2.default.func,
    onSaveDone: _propTypes2.default.func,
    onSaveFail: _propTypes2.default.func,
    onSaveByKey: _propTypes2.default.func,
    onValidate: _propTypes2.default.func,
    onReset: _propTypes2.default.func,
    onCancel: _propTypes2.default.func
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      values: null,
      progressIndicator: null,
      onInit: null,
      onSave: null,
      onSaveDone: null,
      onSaveFail: null,
      onSaveByKey: null,
      onValidate: null,
      onReset: null,
      onCancel: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount: function componentWillMount() {
    this.formInputs = {};
    this.formControls = {};

    var parentForm = this.getParentByType('isForm');
    if (parentForm) {
      this.showWarning('formInForm', [this.getTagName(), parentForm.getTagName()], {
        mixinName: "UU5.Forms.FormMixin"
      });
    }
  },
  componentDidMount: function componentDidMount() {
    this.props.values && this.setValues(this.props.values);
    typeof this.props.onInit === "function" && this.props.onInit({ component: this });
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    this._willReceiveProps = true;
  },
  componentDidUpdate: function componentDidUpdate() {
    if (this._willReceiveProps) {
      this._willReceiveProps = false;
      if (this.props.controlled) {
        this.props.values && this.setValues(this.props.values);
      }
      typeof this.props.onInit === "function" && this.props.onInit({ component: this });
    }
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  setValues: function setValues(values) {
    this.eachFormInput(function (input) {
      var value = values[input.getName() || input.getId()];
      value !== undefined && input.setValue(value);
    });
  },
  eachFormInput: function eachFormInput(func) {
    for (var id in this.formInputs) {
      var result = func(this.formInputs[id]);
      if (result === false) {
        break;
      }
    }
    return this.formInputs;
  },
  eachFormControls: function eachFormControls(func) {
    for (var id in this.formControls) {
      var result = func(this.formControls[id]);
      if (result === false) {
        break;
      }
    }
    return this.formInputs;
  },
  isValid: function isValid() {
    var result = true;

    if (typeof this.props.onValidate === 'function') {
      result = this.props.onValidate({ component: this });
    } else if (typeof this.isValid_ === 'function') {
      result = this.isValid_();
    } else {
      this.eachFormInput(function (formControl) {
        var newResult = typeof formControl.isValid !== 'function' || formControl.isValid();

        if (result) {
          result = newResult;
        }
      });
    }

    return result;
  },


  // for our parent type checking
  isForm: function isForm() {
    return true;
  },

  registerFormInput: function registerFormInput(id, formInput) {
    var registeredControl = this.formInputs[id];
    if (registeredControl) {
      this.showError('duplicateId', id, {
        mixinName: "UU5.Forms.FormMixin",
        context: {
          registeredFormInput: {
            tagName: registeredControl.getTagName(),
            props: registeredControl.props,
            component: registeredControl
          },
          newFormInput: {
            tagName: formInput.getTagName(),
            props: formInput.props,
            component: formInput
          }
        }
      });
    } else {
      this.formInputs[id] = formInput;
    }
  },
  unregisterFormInput: function unregisterFormInput(id) {
    if (!this.formInputs[id]) {
      this.showWarning('nonRegistered', id, {
        mixinName: "UU5.Forms.FormMixin"
      });
    } else {
      delete this.formInputs[id];
    }
  },
  registerFormControls: function registerFormControls(id, formControl) {
    var registeredControl = this.formControls[id];
    if (registeredControl) {
      this.showError('duplicateId', id, {
        mixinName: "UU5.Forms.FormMixin",
        context: {
          registeredFormControl: {
            tagName: registeredControl.getTagName(),
            props: registeredControl.props,
            component: registeredControl
          },
          newFormControl: {
            tagName: formControl.getTagName(),
            props: formControl.props,
            component: formControl
          }
        }
      });
    } else {
      this.formControls[id] = formControl;
    }
  },
  unregisterFormControls: function unregisterFormControls(id) {
    if (!this.formControls[id]) {
      this.showWarning('nonRegistered', id, {
        mixinName: "UU5.Forms.FormMixin"
      });
    } else {
      delete this.formControls[id];
    }
  },
  getValues: function getValues() {
    var values = {};
    this._eachFormInputWithName(function (name, input) {
      values[name] = input.getValue();
    });
    return values;
  },
  getInputs: function getInputs() {
    var inputs = {};
    this._eachFormInputWithName(function (name, input) {
      inputs[name] = input;
    });
    return inputs;
  },
  getInputByName: function getInputByName(name) {
    var result = null;

    this._eachFormInputWithName(function (k, input) {
      var compare = k === name;
      compare && (result = input);
      return !compare;
    });

    return result;
  },
  getFormChildren: function getFormChildren(fce) {
    var _this = this;

    return _react2.default.createElement(
      'form',
      this._getMainAttrs(),
      _react2.default.createElement(UU5.Bricks.AlertBus, { ref_: function ref_(alertBus) {
          return _this._alertBus = alertBus;
        }, forceRender: true }),
      typeof fce === 'function' ? fce() : this.getChildren()
    );
  },
  getAlertBus: function getAlertBus() {
    return this._alertBus;
  },
  save: function save(values) {
    var _this2 = this;

    values = values || this.getValues();
    if (typeof this.props.onSave === 'function') {
      typeof this.props.onSaveDone === 'function' ? this.setPending(function () {
        return _this2.props.onSave({ component: _this2, values: values });
      }) : this.props.onSave({ component: this, values: values });
    } else if (typeof this.save_ === 'function') {
      this.save_();
    }
    return this;
  },
  saveDone: function saveDone(dtoOut) {
    var _this3 = this;

    if (typeof this.props.onSaveDone === 'function') {
      this.setReady(function () {
        return _this3.props.onSaveDone({
          component: _this3,
          dtoOut: dtoOut
        });
      });
    } else if (typeof this.saveDone_ === 'function') {
      this.saveDone_();
    }
    return this;
  },
  saveFail: function saveFail(dtoOut) {
    var _this4 = this;

    if (typeof this.props.onSaveFail === 'function') {
      this.setReady(function () {
        return _this4.props.onSaveFail({
          component: _this4,
          dtoOut: dtoOut
        });
      });
    } else if (typeof this.saveFail_ === 'function') {
      this.saveFail_();
    }
    return this;
  },
  validate: function validate() {
    var result = false;
    result = this.isValid();
    var alertBus = this.getAlertBus();
    alertBus && alertBus.setAlert(result ? {
      colorSchema: 'success',
      closeTimer: 1000,
      content: this.getLsiValue('validContent')
    } : {
      colorSchema: 'danger',
      closeTimer: 5000,
      content: this.getLsiValue('invalidContent')
    });
    return this;
  },
  setPending: function setPending(setStateCallback) {
    var inputs = this.getInputs();
    Object.keys(inputs).forEach(function (key) {
      inputs[key].disable();
    });

    var alertBus = this.getAlertBus();
    alertBus.addAlertToPosition(0, {
      id: this.getId() + "-pending",
      closeDisabled: true,
      content: this.props.progressIndicator || _react2.default.createElement(UU5.Bricks.Loading, null)
    });
    this.eachFormControls(function (formControls) {
      return formControls.disable(setStateCallback);
    });

    return this;
  },
  setReady: function setReady(setStateCallback) {
    var inputs = this.getInputs();
    Object.keys(inputs).forEach(function (key) {
      inputs[key].enable();
    });

    var alertBus = this.getAlertBus();
    alertBus.removeAlert(this.getId() + "-pending");
    this.eachFormControls(function (formControls) {
      return formControls.enable(setStateCallback);
    });

    return this;
  },
  reset: function reset(setStateCallback) {
    if (typeof this.props.onReset === 'function') {
      this.props.onReset({ component: this });
    } else if (typeof this.reset_ === 'function') {
      this.reset_();
    } else {
      var counter = 0;
      for (var id in this.formInputs) {
        typeof this.formInputs[id].reset === 'function' && counter++;
      }

      var newSetStateCallback = UU5.Common.Tools.buildCounterCallback(setStateCallback, counter);

      for (var _id in this.formInputs) {
        typeof this.formInputs[_id].reset === 'function' && this.formInputs[_id].reset(newSetStateCallback);
      }
    }
    return this;
  },
  cancel: function cancel() {
    if (typeof this.props.onCancel === 'function') {
      this.props.onCancel({ component: this });
    } else if (typeof this.cancel_ === 'function') {
      this.cancel_();
    }
    return this;
  },
  getSaveFormEvents: function getSaveFormEvents(func) {
    var _this5 = this;

    var CTRL_KEY = 17,
        S_KEY = 83;
    var down = false;

    return {
      onKeyDown: function onKeyDown(e) {
        var key = e.which || e.keyCode;
        var isCtrlPressed = e.ctrlKey ? e.ctrlKey : key === CTRL_KEY;

        if (isCtrlPressed && key === S_KEY) {
          if (down) return;
          down = true;

          e.preventDefault();
          e.stopPropagation();
          if (typeof func === 'function') func({ component: _this5, event: e });
        }
      },
      onKeyUp: function onKeyUp(e) {
        var key = e.which || e.keyCode;
        var isCtrlPressed = e.ctrlKey ? e.ctrlKey : key === CTRL_KEY;

        if (isCtrlPressed && key === S_KEY) {
          down = false;
        }
      }
    };
  },

  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _eachFormInputWithName: function _eachFormInputWithName(func) {
    var form = this;
    this.eachFormInput(function (input) {
      var name = input.getName();
      if (!name) {
        form.showWarning('noName', input.getId(), {
          mixinName: "UU5.Forms.FormMixin",
          context: {
            input: {
              tagName: input.getTagName(),
              props: input.props,
              component: input
            }
          }
        });
        name = input.getId();
      }
      func(name, input);
    });
    return this;
  },
  _getMainAttrs: function _getMainAttrs() {
    var mainAttrs = this.getMainAttrs();

    if (typeof this.props.onSaveByKey === "function") {
      var saveEvents = this.getSaveFormEvents(this.props.onSaveByKey);
      Object.keys(saveEvents).forEach(function (key) {
        var fce = mainAttrs[key];
        mainAttrs[key] = function (e) {
          typeof fce === "function" && fce(e);
          saveEvents[key](e);
        };
      });
    }

    return mainAttrs;
  }
};

exports.default = FormMixin;
