// uu5g04/dist-esm/core/common/ccr-reader-mixin.js
import Environment from '../environment/environment.js';

export const CcrReaderMixin = {

  //@@viewOn:statics
  statics: {
    "UU5.Common.CcrReaderMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"],
      errors: {
        keyNotRegistered: 'Component with key %s is not registered.'
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function () {
    // initialize
    this.registerMixin("UU5.Common.CcrReaderMixin");
    // state
    return null;
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonCcrReaderMixin: function () {
    return this.hasMixin("UU5.Common.CcrReaderMixin");
  },

  getUU5CommonCcrReaderMixinProps: function () {
    return {};
  },

  getUU5CommonCcrReaderMixinPropsToPass: function () {
    return this.getUU5CommonCcrReaderMixinProps();
  },

  getCcrComponentByKey: function (key) {
    return this.getCcrByKeyRegister()[key] || null;
  },

  isCcrRegisteredByKey: function (key) {
    return !!this.getCcrByKeyRegister()[key];
  },

  getCcrByKeyRegister: function () {
    return Environment.ccr.byKey;
  }
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers
};

export default CcrReaderMixin;
