// uu5g04/dist-esm/core/common/screen-size-mixin.js
import React from 'react';
import Environment from '../environment/environment.js';
import Tools from './tools.js';
import EventListener from './event-listener.js';

export const ScreenSizeMixin = {

  //@@viewOn:statics
  statics: {
    "UU5.Common.ScreenSizeMixin": {
      requiredMixins: ["UU5.Common.BaseMixin"],
      defaults: {
        screenSizeEvent: 'UU5_Common_screenSize'
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function () {
    // initialize
    this.registerMixin("UU5.Common.ScreenSizeMixin");
    // state
    return {
      screenSize: this.getScreenSize()
    };
  },

  componentDidMount: function (nextProps) {
    window.UU5.Environment.EventListener.registerScreenSize(this.getId(), this._onChangeScreenSize);
  },

  componentWillUnmount: function () {
    window.UU5.Environment.EventListener.unregisterScreenSize(this.getId(), this._onChangeScreenSize);
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonScreenSizeMixinProps: function () {
    return this.hasMixin("UU5.Common.ScreenSizeMixin");
  },

  getUU5CommonScreenSizeMixinProps: function () {
    return {
      screenSize: this.getScreenSize()
    };
  },

  getUU5CommonScreenSizeMixinPropsToPass: function () {
    return this.getUU5CommonScreenSizeMixinProps();
  },

  getScreenSize: function () {
    return this.state && this.state.screenSize || Tools.getScreenSize();
  },

  isXs: function () {
    return this.getScreenSize() === 'xs';
  },

  isS: function () {
    return this.getScreenSize() === 's';
  },

  isM: function () {
    return this.getScreenSize() === 'm';
  },

  isL: function () {
    return this.getScreenSize() === 'l';
  },

  isXl: function () {
    return this.getScreenSize() === 'xl';
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onChangeScreenSize: function (e, actualScreenSize) {
    if (typeof this.onChangeScreenSize_ === 'function') {
      this.onChangeScreenSize_(actualScreenSize, e);
    } else {
      this.setState({ screenSize: actualScreenSize });
    }
    return this;
  }
  //@@viewOff:componentSpecificHelpers
};

export default ScreenSizeMixin;
