// uu5g04/dist-esm/core/common/section-mixin.js
import React from 'react';
import PropTypes from 'prop-types';
import ContentMixin from './content-mixin.js';
import LevelMixin from './level-mixin.js';
import Tools from './tools.js';

export const SectionMixin = {

  //@@viewOn:mixins
  mixins: [ContentMixin, LevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    "UU5.Common.SectionMixin": {
      requiredMixins: ["UU5.Common.BaseMixin", "UU5.Common.ContentMixin", "UU5.Common.LevelMixin"],
      defaults: {
        headerTag: 'UU5.Bricks.Header',
        footerTag: 'UU5.Bricks.Footer',
        regexpUu5Json: /(^<uu5json\/>)/
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    header: PropTypes.oneOfType([
    // content body:[{tag:'',props:{}},{tag:'',props:{}},...]
    PropTypes.arrayOf(PropTypes.shape({
      tag: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
      props: PropTypes.object
    })),
    // content bodyItem:{tag:'',props{}}
    PropTypes.shape({
      tag: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
      props: PropTypes.arrayOf(PropTypes.object)
    }),
    // content items:{tag:'',propsArray:[{},{},{},...]}
    PropTypes.shape({
      tag: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
      propsArray: PropTypes.arrayOf(PropTypes.object)
    }),
    // content node
    PropTypes.node,
    // element
    PropTypes.shape({
      element: PropTypes.oneOfType([PropTypes.element, PropTypes.shape({
        tag: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
        props: PropTypes.object
      })])
    })]),
    footer: PropTypes.oneOfType([
    // content body:[{tag:'',props:{}},{tag:'',props:{}},...]
    PropTypes.arrayOf(PropTypes.shape({
      tag: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
      props: PropTypes.object
    })),
    // content bodyItem:{tag:'',props{}}
    PropTypes.shape({
      tag: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
      props: PropTypes.arrayOf(PropTypes.object)
    }),
    // content items:{tag:'',propsArray:[{},{},{},...]}
    PropTypes.shape({
      tag: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
      propsArray: PropTypes.arrayOf(PropTypes.object)
    }),
    // content node
    PropTypes.node,
    // element
    PropTypes.shape({
      element: PropTypes.oneOfType([PropTypes.element, PropTypes.shape({
        tag: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
        props: PropTypes.object
      })])
    })]),
    underline: PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function () {
    return {
      header: null,
      footer: null,
      underline: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState: function () {
    // initialize
    this.registerMixin("UU5.Common.SectionMixin");
    this.renderedHeaderChild = null; // renderedChild
    this.renderedFooterChild = null; // renderedChild
    // state
    return null;
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  hasUU5CommonSectionMixin: function () {
    return this.hasMixin("UU5.Common.SectionMixin");
  },

  getHeader: function () {
    return typeof this.getHeader_ === 'function' ? this.getHeader_() : this.props.header;
  },

  getFooter: function () {
    return typeof this.getFooter_ === 'function' ? this.getFooter_() : this.props.footer;
  },

  getHeaderType: function (headerContent) {
    var type = null; // one of ['contentOfStandardHeader','headerElement','headerBodyItem']
    if (headerContent || headerContent === "") {
      if (typeof headerContent.element === 'object') {
        type = headerContent.element.tag ? 'headerBodyItem' : 'headerElement';
      } else if (typeof headerContent === 'string') {
        let match = this.getDefault('regexpUu5Json', "UU5.Common.SectionMixin").exec(headerContent);
        type = match ? 'uu5JSON' : 'contentOfStandardHeader';
      } else {
        type = 'contentOfStandardHeader';
      }
    }
    return type;
  },

  getFooterType: function (footerContent) {
    var type = null; // one of ['contentOfStandardFooter','footerElement','footerBodyItem']
    if (footerContent || footerContent === "") {
      if (typeof footerContent.element === 'object') {
        type = footerContent.element.tag ? 'footerBodyItem' : 'footerElement';
      } else if (typeof footerContent === 'string') {
        let match = this.getDefault('regexpUu5Json', "UU5.Common.SectionMixin").exec(footerContent);
        type = match ? 'uu5JSON' : 'contentOfStandardFooter';
      } else {
        type = 'contentOfStandardFooter';
      }
    }
    return type;
  },

  getUU5CommonSectionMixinProps: function () {
    return {
      header: this.getHeader(),
      footer: this.getFooter()
    };
  },

  getUU5CommonSectionMixinPropsToPass: function () {
    return Tools.mergeDeep({}, this.getUU5CommonSectionMixinProps(), this.getUU5CommonContentMixinPropsToPass(), this.getUU5CommonLevelMixinPropsToPass());
  },

  registerRenderedHeaderChild: function (renderedHeaderChild) {
    if (renderedHeaderChild.hasUU5CommonBaseMixin) {
      this.renderedHeaderChild = renderedHeaderChild;
    }
    return this;
  },

  registerRenderedFooterChild: function (renderedFooterChild) {
    if (renderedFooterChild.hasUU5CommonBaseMixin) {
      this.renderedFooterChild = renderedFooterChild;
    }
    return this;
  },

  expandHeaderProps: function (prevHeaderChild) {
    var newHeaderChildProps = prevHeaderChild.props;
    newHeaderChildProps = Tools.mergeDeep({}, newHeaderChildProps);
    newHeaderChildProps.parent = this;
    newHeaderChildProps.id = newHeaderChildProps.id || this.getId() + '-header';
    newHeaderChildProps.underline = newHeaderChildProps.underline || this.props.underline;

    if (typeof this.expandHeaderProps_ === 'function') {
      var tempHeaderChild = this.cloneChild(prevHeaderChild, newHeaderChildProps);
      newHeaderChildProps = this.expandHeaderProps_(tempHeaderChild);
    }

    newHeaderChildProps.key = newHeaderChildProps.id;

    newHeaderChildProps.ref = function (renderedChild) {
      renderedChild && this.registerRenderedHeaderChild(renderedChild);
    }.bind(this);

    return newHeaderChildProps;
  },

  expandFooterProps: function (prevFooterChild) {
    var newFooterChildProps = prevFooterChild.props;
    newFooterChildProps = Tools.mergeDeep({}, newFooterChildProps);
    newFooterChildProps.parent = this;
    newFooterChildProps.id = newFooterChildProps.id || this.getId() + '-footer';

    if (typeof this.expandFooterProps_ === 'function') {
      var tempFooterChild = this.cloneChild(prevFooterChild, newFooterChildProps);
      newFooterChildProps = this.expandFooterProps_(tempFooterChild);
    }

    newFooterChildProps.key = newFooterChildProps.id;

    newFooterChildProps.ref = function (renderedChild) {
      renderedChild && this.registerRenderedFooterChild(renderedChild);
    }.bind(this);

    return newFooterChildProps;
  },

  buildHeaderChild: function (header) {
    header = header || this.getHeader();

    var headerChild = null;

    if (typeof this.buildHeaderChild_ === 'function') {
      headerChild = this.buildHeaderChild_(header);
    } else {

      var headerValue = header;
      var headerType = this.getHeaderType(headerValue);

      if (headerType === 'uu5JSON') {
        headerValue = Tools.parseFromUu5JSON(headerValue);
        headerType = this.getHeaderType(headerValue);
      }

      switch (headerType) {
        case 'headerBodyItem':
          headerChild = this.buildChild(headerValue.element.tag, headerValue.element.props);
          headerChild = this.cloneChild(headerChild, this.expandHeaderProps(headerChild));
          break;
        case 'headerElement':
          headerChild = headerValue.element;
          headerChild = this.cloneChild(headerChild, this.expandHeaderProps(headerChild));
          break;
        case 'contentOfStandardHeader':
          headerChild = this.buildChild(this.getDefault('headerTag', "UU5.Common.SectionMixin"), { content: headerValue });
          headerChild = this.cloneChild(headerChild, this.expandHeaderProps(headerChild));
          break;
        default:
        // there is no header
      }
    }
    return headerChild;
  },

  buildFooterChild: function (footer) {
    footer = footer || this.getFooter();

    var footerChild = null;

    if (typeof this.buildFooterChild_ === 'function') {
      footerChild = this.buildFooterChild_(footer);
    } else {

      var footerValue = footer;
      var footerType = this.getFooterType(footerValue);

      if (footerType === 'uu5JSON') {
        footerValue = Tools.parseFromUu5JSON(footerValue);
        footerType = this.getFooterType(footerValue);
      }

      switch (footerType) {
        case 'footerBodyItem':
          footerChild = this.buildChild(footerValue.element.tag, footerValue.element.props);
          footerChild = this.cloneChild(footerChild, this.expandFooterProps(footerChild));
          break;
        case 'footerElement':
          footerChild = footerValue.element;
          footerChild = this.cloneChild(footerChild, this.expandFooterProps(footerChild));
          break;
        case 'contentOfStandardFooter':
          footerChild = this.buildChild(this.getDefault('footerTag', "UU5.Common.SectionMixin"), { content: footerValue });
          footerChild = this.cloneChild(footerChild, this.expandFooterProps(footerChild));
          break;
        default:
        // there is no footer
      }
    }
    return footerChild;
  },

  getHeaderChild: function () {
    return this.buildHeaderChild();
  },

  getFooterChild: function () {
    return this.buildFooterChild();
  },

  getRenderedHeaderChild: function () {
    return this.renderedHeaderChild;
  },

  getRenderedFooterChild: function () {
    return this.renderedFooterChild;
  }
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

};

export default SectionMixin;
