// uu5g04/dist-esm/core/common/tag-placeholder.js
import React from 'react';
import createReactClass from 'create-react-class';
import ns from "./common-ns.js";
import PropTypes from 'prop-types';
import BaseMixin from './base-mixin.js';
import ElementaryMixin from './elementary-mixin.js';
import Tools from './tools.js';
import Environment from '../environment/environment.js';
import NotFoundTag from './not-found-tag.js';

//import './tag-placeholder.less';

let importLibraryCache = {};

export const TagPlaceholder = createReactClass({

  //@@viewOn:mixins
  mixins: [BaseMixin, ElementaryMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: ns.name("TagPlaceholder"),
    classNames: {
      main: ns.css("tag-placeholder")
    },
    errors: {
      serverError: 'Unexpected error: %s.'
    },
    defaults: {
      regexpVersion: /\/\d*\.\d*\.\d*(?:-[a-z]+\d+(?:\.\d){0,2})?\//g,
      regexpSlash: /\//g,
      regexpDigit: /^\d/g
    },
    opt: {
      hoc: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    tagName: PropTypes.string,
    props: PropTypes.object,
    content: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
  },
  //@@viewOff:propTypes
  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      tagName: null,
      props: null,
      content: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState() {
    return {
      component: null
    };
  },

  componentDidMount() {
    this._findLib();
  },

  componentWillReceiveProps(nextProps) {
    let component = nextProps.tagName && Tools.checkTag(nextProps.tagName);
    component ? this._setComponent(component, nextProps.props) : this._findLib(nextProps);

    return this;
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _findLib(props) {
    props = props || this.props;
    let tagNameArr = props.tagName.split('.');
    let libraryName = tagNameArr[0] + '.' + tagNameArr[1];
    let libraryReady = (result, error) => {
      if (error) {
        let library = Environment.getLibrary(libraryName);
        if (!library || !library.error) {
          this.showError('serverError', error, {
            component: this.getTagName(),
            searchedTag: props.tagName,
            library: libraryName
          });
        }
        this._setLibError(libraryName);
        this._setComponent(props.tagName, { tagName: props.tagName, props: props.props });
      } else {
        this._setComponent(props.tagName, props.props);
      }
    };

    Tools.loadLibrary(libraryName, (libraryJson, error) => {
      if (!error && libraryJson.source) {
        this._importLibrary(libraryJson, libraryReady);
      } else {
        libraryReady(null, error || new Error("Library is missing a source"));
      }
    });

    return this;
  },

  _getSourceUrl(library) {
    let result;
    if (library && library.name && library.version) {
      let baseName = library.name.replace('_basic', '').replace('_forms', '');
      let sourceName = library.name.replace('uu_', '');
      result = Environment.CDN_URL + '/' + baseName + '/' + library.version + '/' + sourceName + '.js';
    }
    return result;
  },

  _setComponent(tagName, props) {
    // props = props || this.props.props;

    let component = tagName && Tools.checkTag(tagName);
    if (!component) {
      component = NotFoundTag;
      props = Tools.merge({}, props);
      props.tagName = props.tagName || tagName;
    }

    component && this.setState({ component: component, props: props });
    return this;
  },

  _setNotFound(tagCalls) {
    tagCalls.forEach(fce => fce());
    return this;
  },

  _setLibError(libCode) {
    let libSettings = Environment.getLibrary(libCode);
    if (libSettings) {
      libSettings.error = true;
    } else {
      Environment.addLibrary(libCode, { error: true });
    }
    return this;
  },

  _getVersionFromUrl(url) {
    let version = null;
    version = url.match(this.getDefault().regexpVersion);
    version && version[0] && (version = version[0].replace(this.getDefault().regexpSlash, ''));
    return version;
  },

  _getVersionObject(version) {
    let ver = version.split('.');
    return {
      major: parseInt(ver[0]),
      minor: parseInt(ver[1]),
      build: parseInt(ver[2])
    };
  },

  _compareDependencies(version, dependency) {
    let result = true;
    let dependencyVersion;

    if (this.getDefault().regexpDigit.test(dependency)) {
      dependencyVersion = dependency;
    } else {
      dependencyVersion = this._getVersionFromUrl(dependency);
    }

    if (dependencyVersion && version !== dependencyVersion) {

      let ver = this._getVersionObject(version);
      let verDep = this._getVersionObject(dependencyVersion);

      if (ver.major !== verDep.major) {
        // 1.0.0 x 2.0.0 || 1.0.0 x 0.1.1
        result = false;
      } else if (ver.minor < verDep.minor) {
        // 1.0.0 x 1.1.0
        result = false;
      } else if (ver.minor === verDep.minor && ver.build < verDep.build) {
        // 1.0.0 x 1.0.1
        result = false;
      }

      if (!result) {
        this._setDependencyError(version, dependencyVersion);
      }
    }

    return true;
  },

  _setDependencyError(version, dependencyVersion) {
    Tools.error('Wrong dependency.', {
      tag: this.props.tagName,
      ver: version,
      dependencyVersion: dependencyVersion
    });

    return this;
  },

  _updateDependencies(dependencies) {
    if (dependencies['uu5g04'] && this._compareDependencies(Environment.version, dependencies['uu5g04'])) {
      delete dependencies['uu5g04'];
      if (window.UU5) {
        window.UU5.Bricks && Object.keys(window.UU5.Bricks).length && delete dependencies['uu5g04-bricks'];
        window.UU5.Forms && Object.keys(window.UU5.Forms).length && delete dependencies['uu5g04-forms'];
      }
    }
    if (window.Plus4U5 && dependencies['uu_plus4u5g01'] && this._compareDependencies(window.Plus4U5.Environment.version, dependencies['uu_plus4u5g01'])) {
      delete dependencies['uu_plus4u5g01'];
      window.Plus4U5.App && Object.keys(window.Plus4U5.App).length && delete dependencies['uu_plus4u5g01-app'];
      window.Plus4U5.Bricks && Object.keys(window.Plus4U5.Bricks).length && delete dependencies['uu_plus4u5g01-bricks'];
    }

    Object.keys(dependencies).forEach(key => {
      if (this.getDefault().regexpDigit.test(dependencies[key])) {
        let library = {
          name: key,
          version: dependencies[key]
        };
        let source = this._getSourceUrl(library);
        source && (dependencies[key] = source);
      }
    });

    return dependencies;
  },

  _importLibrary(library, callback) {
    let dependencies = library.dependencies;
    let source = library.source;
    if (!source) {
      source = this._getSourceUrl(library);
    }

    var query = library.id;
    var cache = importLibraryCache;
    if (!cache[query]) {
      cache[query] = {
        result: null,
        error: null,
        pendingCallback: [callback]
      };
      if (window.System && window.System.import) {
        if (source) {
          System.config({ map: this._updateDependencies(dependencies) });

          System.import(source).then(libraryComponents => {
            cache[query].pendingCallback.forEach(fn => fn(libraryComponents));
          });
        } else {
          Tools.error('System was not found and has not set any source to import.', {
            source: source,
            tagName: this.props.tagName
          });
          this._setLibError(library.code);
          cache[query].pendingCallback.forEach(fn => fn(null, new Error()));
        }
      } else {
        Tools.error('System is not defined in window! Cannot import source:', {
          source: source,
          tagName: this.props.tagName
        });
        this._setLibError(library.code);
        cache[query].pendingCallback.forEach(fn => fn(null, new Error()));
      }
    } else if (cache[query].result || cache[query].error) {
      callback(cache[query].result, cache[query].error);
    } else {
      cache[query].pendingCallback.push(callback);
    }
    return this;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.state.component ? React.createElement(this.state.component, Tools.merge({}, this.state.props, { parent: this.getParent() }), this.props.content) : null;
  }
  //@@viewOff:render
});

export default TagPlaceholder;
